<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>The {traits.build} data standard, R package, and workflow - 32&nbsp; The austraits package</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./help.html" rel="next">
<link href="./spatial_data_example.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./austraits_database.html">Using outputs of <code>traits.build</code></a></li><li class="breadcrumb-item"><a href="./austraits_package.html"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">The `austraits` package</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">The {traits.build} data standard, R package, and workflow</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Introduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">About</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./motivation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Motivation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./usage_examples.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Usage examples</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Data structure and standard</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./long_wide.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Long vs Wide data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_standard.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data standard</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_structure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Data structure</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_metadata.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Overview of traits.build ontology</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dictionary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Trait dictionary</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Creating with <code>traits.build</code></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./traits_build.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">The package</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_compilation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Tutorial: Example compilation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./file_organisation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">File organisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./create_dictionary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Creating a dictionary</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./adding_data_brief.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Adding datasets (brief)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./custom_R_code.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Adding custom_R_code</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./publishing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Publishing</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Guide to adding data</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Tutorial: Adding datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Tutorial 1: Adding a simple dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Tutorial 2: Adding a more complex dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Tutorial 3: Adding contexts and complex units</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Tutorial 4: Additional complexities</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Tutorial 5: Multiple columns for a trait</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_6.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Tutorial 6: Data with repeat measurements</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_7.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Tutorial 7: Adding long format dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./adding_data_long.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">Adding datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./check_dataset_functions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">Check data quality functions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_common_issues.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">Common issues</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Using outputs of <code>traits.build</code></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./austraits_database.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">AusTraits, Australia’s Plant Trait Database</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./AusTraits_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">AusTraits tutorial</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./traits_and_climate_example.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">Combining trait values and climate data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spatial_data_example.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">Tutorial: Using AusTraits with spatial data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./austraits_package.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">The <code>austraits</code> package</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Getting help</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./help.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">Getting help</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./debugging.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">34</span>&nbsp; <span class="chapter-title">Debugging</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./csv.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">CSV files</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./yaml.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Yaml files</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#getting-started" id="toc-getting-started" class="nav-link active" data-scroll-target="#getting-started"><span class="header-section-number">32.1</span> Getting started</a>
  <ul class="collapse">
  <li><a href="#loading-austraits-database" id="toc-loading-austraits-database" class="nav-link" data-scroll-target="#loading-austraits-database">Loading AusTraits database</a></li>
  </ul></li>
  <li><a href="#descriptive-summaries-of-traits-and-taxa" id="toc-descriptive-summaries-of-traits-and-taxa" class="nav-link" data-scroll-target="#descriptive-summaries-of-traits-and-taxa"><span class="header-section-number">32.2</span> Descriptive summaries of traits and taxa</a></li>
  <li><a href="#extracting-data" id="toc-extracting-data" class="nav-link" data-scroll-target="#extracting-data"><span class="header-section-number">32.3</span> Extracting data</a>
  <ul class="collapse">
  <li><a href="#extracting-by-study" id="toc-extracting-by-study" class="nav-link" data-scroll-target="#extracting-by-study">Extracting by study</a></li>
  <li><a href="#extracting-by-taxonomic-level" id="toc-extracting-by-taxonomic-level" class="nav-link" data-scroll-target="#extracting-by-taxonomic-level">Extracting by taxonomic level</a></li>
  <li><a href="#extracting-by-trait" id="toc-extracting-by-trait" class="nav-link" data-scroll-target="#extracting-by-trait">Extracting by trait</a></li>
  </ul></li>
  <li><a href="#join-data-from-other-tables-and-elements" id="toc-join-data-from-other-tables-and-elements" class="nav-link" data-scroll-target="#join-data-from-other-tables-and-elements"><span class="header-section-number">32.4</span> Join data from other tables and elements</a></li>
  <li><a href="#visualising-data-by-site" id="toc-visualising-data-by-site" class="nav-link" data-scroll-target="#visualising-data-by-site"><span class="header-section-number">32.5</span> Visualising data by site</a></li>
  <li><a href="#visualising-data-distribution-and-variance" id="toc-visualising-data-distribution-and-variance" class="nav-link" data-scroll-target="#visualising-data-distribution-and-variance"><span class="header-section-number">32.6</span> Visualising data distribution and variance</a></li>
  <li><a href="#pivotting-from-long-to-wide-format" id="toc-pivotting-from-long-to-wide-format" class="nav-link" data-scroll-target="#pivotting-from-long-to-wide-format"><span class="header-section-number">32.7</span> Pivotting from long to wide format</a>
  <ul class="collapse">
  <li><a href="#pivot-wider-3.0.2" id="toc-pivot-wider-3.0.2" class="nav-link" data-scroll-target="#pivot-wider-3.0.2">Pivot wider &lt;=3.0.2</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">The <code>austraits</code> package</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>The <code>austraits</code> package is designed to aid users in accessing data from <a href="https://zenodo.org/record/5112001">AusTraits</a>, a curated plant trait database for the Australian flora. This package contains several core functions to wrangle and visualise data. Below, we include a tutorial to illustrate how to use these functions in greater detail.</p>
<p><strong>Note that we are using a reduced subset of v.3.0.2, so the number of records and taxa will not be match your values</strong></p>
<section id="getting-started" class="level2" data-number="32.1">
<h2 data-number="32.1" class="anchored" data-anchor-id="getting-started"><span class="header-section-number">32.1</span> Getting started</h2>
<p><code>austraits</code> is still under development. To install the current version from GitHub:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("remotes")</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co">#remotes::install_github("traitecoevo/austraits", dependencies = TRUE, upgrade = "ask")</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Load the austraits package</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(austraits)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="loading-austraits-database" class="level3">
<h3 class="anchored" data-anchor-id="loading-austraits-database">Loading AusTraits database</h3>
<p>By default, <code>load_austraits</code> will download AusTraits to a specified path e.g.&nbsp;<code>data/austraits</code> and will reload it from this location in the future. You can set <code>update = TRUE</code> so the austrait versions are downloaded fresh from <a href="https://zenodo.org/record/3568429">Zenodo</a>. Note that <code>load_austraits</code> will happily accept a DOI of a particular version.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>austraits <span class="ot">&lt;-</span> <span class="fu">load_austraits</span>(<span class="at">version =</span> <span class="st">"3.0.2"</span>, <span class="at">path =</span> <span class="st">"data/austraits"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can check out different versions and their associated DOI of AusTraits by using:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">get_versions</span>(<span class="at">path =</span> <span class="st">"data/austraits"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The AusTraits object is a very <em>long</em> list with various of elements. If you are not familiar with working with lists in R, we recommend having a quick look at this <a href="https://www.tutorialspoint.com/r/r_lists.htm">tutorial</a>. To learn more about the structure of <code>austraits</code>, check out the <a href="https://traitecoevo.github.io/austraits/articles/structure.html">structure of the database</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>austraits</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; This is version 0.0.0.900 of austraits!
#&gt;  
#&gt; This object contains a total of 5859 records for 1245 taxa and 33 traits.
#&gt; 
#&gt; This object is a 'list' with the following components:
#&gt; 
#&gt;  - `traits`: A table containing measurements of plant traits.
#&gt; - `sites`: A table containing observations of site characteristics associated with information in `traits`. Cross referencing between the two dataframes is possible using combinations of the variables `dataset_id`, `site_name`.
#&gt; - `contexts`: A table containing observations of contextual characteristics associated with information in `traits`. Cross referencing between the two dataframes is possible using combinations of the variables `dataset_id`, `context_name`.
#&gt; - `methods`: A table containing details on methods with which data were collected, including time frame and source.
#&gt; - `excluded_data`: A table of data that did not pass quality test and so were excluded from the master dataset.
#&gt; - `taxa`: A table containing details on taxa associated with information in `traits`. This information has been sourced from the APC (Australian Plant Census) and APNI (Australian Plant Names Index) and is released under a CC-BY3 license.
#&gt; - `definitions`: A copy of the definitions for all tables and terms. Information included here was used to process data and generate any documentation for the study.
#&gt; - `sources`: Bibtex entries for all primary and secondary sources in the compilation.
#&gt; - `contributors`: A table of people contributing to each study.
#&gt; - `taxonomic_updates`: A table of all taxonomic changes implemented in the construction of AusTraits. Changes are determined by comapring against the APC (Australian Plant Census) and APNI (Australian Plant Names Index).
#&gt; - `build_info`: A description of the computing environment used to create this version of the dataset, including version number, git commit and R session_info.
#&gt; 
#&gt; To access a component, try using the $ e.g. austraits$traits</code></pre>
</div>
</div>
</section>
</section>
<section id="descriptive-summaries-of-traits-and-taxa" class="level2" data-number="32.2">
<h2 data-number="32.2" class="anchored" data-anchor-id="descriptive-summaries-of-traits-and-taxa"><span class="header-section-number">32.2</span> Descriptive summaries of traits and taxa</h2>
<p>AusTraits contains 33 plant traits. Check out <a href="https://traitecoevo.github.io/austraits/articles/dictionary.html">definitions of the traits</a> to learn more about how each trait is defined.</p>
<p>Have a look at what trait or taxa we have with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summarise_austraits</span>(austraits, <span class="st">"trait_name"</span>) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt;            trait_name n_records n_dataset n_taxa percent_total
#&gt;  branch_mass_fraction        45         1     45      0.007680
#&gt;     fire_cued_seeding         5         1      5      0.000853
#&gt;         fire_response        73         2     72      0.012500
#&gt;        flowering_time        31         1     31      0.005290
#&gt;           huber_value       193         2     57      0.032900
#&gt;   leaf_N_per_dry_mass        62         2     62      0.010600</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summarise_austraits</span>(austraits, <span class="at">var =</span>  <span class="st">"family"</span>) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt;         family n_records n_dataset n_taxa percent_total
#&gt;    Acanthaceae         1         1      1      0.000171
#&gt;     Akaniaceae         3         1      1      0.000512
#&gt;  Anacardiaceae        20         2      7      0.003410
#&gt;     Annonaceae        15         2      5      0.002560
#&gt;       Apiaceae        17         1      2      0.002900
#&gt;    Apocynaceae        32         3      9      0.005460</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summarise_austraits</span>(austraits, <span class="st">"genus"</span>) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt;       genus n_records n_dataset n_taxa percent_total
#&gt;      Acacia       184         4     77      0.031400
#&gt;   Aceratium         5         1      3      0.000854
#&gt;      Ackama        10         1      2      0.001710
#&gt;   Acradenia         3         1      2      0.000512
#&gt;  Acronychia        21         3      7      0.003590
#&gt;   Actinotus        10         1      1      0.001710</code></pre>
</div>
</div>
<p>Interested in a specific trait? Try <code>lookup_trait</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lookup_trait</span>(austraits, <span class="st">"leaf"</span>) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; [1] "leaf_angle"              "leaf_area"              
#&gt; [3] "leaf_compoundness"       "leaf_N_per_dry_mass"    
#&gt; [5] "specific_leaf_area"      "leaf_cell_wall_fraction"</code></pre>
</div>
</div>
</section>
<section id="extracting-data" class="level2" data-number="32.3">
<h2 data-number="32.3" class="anchored" data-anchor-id="extracting-data"><span class="header-section-number">32.3</span> Extracting data</h2>
<p>In most cases, users would like to extract a subset of <code>austraits</code> for their own research purposes.<code>extract_dataset</code> subsets a particular study, whereas <code>extract_trait</code>subsets by certain traits. Note that the other tables and elements of the AusTraits data are extracted too, not just the main trait table. See <code>?extract_dataset</code> and <code>?extract_trait</code> for more details</p>
<section id="extracting-by-study" class="level3">
<h3 class="anchored" data-anchor-id="extracting-by-study">Extracting by study</h3>
<p>Filtering <strong>one particular study</strong> and assigning it to an object</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>subset_data <span class="ot">&lt;-</span> <span class="fu">extract_dataset</span>(austraits, <span class="st">"Falster_2005_2"</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>subset_data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 6 × 12
#&gt;   dataset_id   taxon_name site_name context_name observation_id trait_name value
#&gt;   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt;
#&gt; 1 Falster_200… Acacia lo… Myall_La… &lt;NA&gt;         Falster_2005_… fire_resp… fire…
#&gt; 2 Falster_200… Acacia lo… Myall_La… &lt;NA&gt;         Falster_2005_… huber_val… 0.00…
#&gt; 3 Falster_200… Acacia lo… Myall_La… &lt;NA&gt;         Falster_2005_… huber_val… 0.00…
#&gt; 4 Falster_200… Acacia lo… Myall_La… &lt;NA&gt;         Falster_2005_… huber_val… 0.00…
#&gt; 5 Falster_200… Acacia lo… Myall_La… &lt;NA&gt;         Falster_2005_… huber_val… 0.00…
#&gt; 6 Falster_200… Acacia lo… Myall_La… &lt;NA&gt;         Falster_2005_… leaf_area  1761 
#&gt; # ℹ 5 more variables: unit &lt;chr&gt;, date &lt;chr&gt;, value_type &lt;fct&gt;,
#&gt; #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
</div>
<p>Filtering <strong>multiple studies by two different lead authors</strong> and assigning it to an object</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>subset_multi_studies <span class="ot">&lt;-</span> <span class="fu">extract_dataset</span>(austraits, </span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">dataset_id =</span> <span class="fu">c</span>(<span class="st">"Thompson_2001"</span>,<span class="st">"Ilic_2000"</span>))</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>subset_multi_studies<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 6 × 12
#&gt;   dataset_id taxon_name   site_name context_name observation_id trait_name value
#&gt;   &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt;
#&gt; 1 Ilic_2000  Acacia acra… &lt;NA&gt;      &lt;NA&gt;         Ilic_2000_0001 wood_dens… 0.904
#&gt; 2 Ilic_2000  Acacia acum… &lt;NA&gt;      &lt;NA&gt;         Ilic_2000_0002 wood_dens… 0.895
#&gt; 3 Ilic_2000  Acacia acum… &lt;NA&gt;      &lt;NA&gt;         Ilic_2000_0003 wood_dens… 1.008
#&gt; 4 Ilic_2000  Acacia adsu… &lt;NA&gt;      &lt;NA&gt;         Ilic_2000_0004 wood_dens… 0.887
#&gt; 5 Ilic_2000  Acacia ampl… &lt;NA&gt;      &lt;NA&gt;         Ilic_2000_0005 wood_dens… 0.568
#&gt; 6 Ilic_2000  Acacia aneu… &lt;NA&gt;      &lt;NA&gt;         Ilic_2000_0006 wood_dens… 1.035
#&gt; # ℹ 5 more variables: unit &lt;chr&gt;, date &lt;chr&gt;, value_type &lt;fct&gt;,
#&gt; #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
</div>
<p>Filtering <strong>multiple studies by same lead author</strong> (e.g.&nbsp;Falster) and assigning it to an object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First, we need to identify all studies with an id that includes "Falster"</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>( dataset_ids <span class="ot">&lt;-</span> austraits<span class="sc">$</span>methods<span class="sc">$</span>dataset_id <span class="sc">%&gt;%</span> <span class="fu">unique</span>() <span class="sc">%&gt;%</span> <span class="fu">subset</span>(., <span class="fu">grepl</span>(<span class="st">"Falster"</span>,.))) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; [1] "Falster_2003"   "Falster_2005_1" "Falster_2005_2"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Then we extract</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>data_falster_studies <span class="ot">&lt;-</span> <span class="fu">extract_dataset</span>(austraits, dataset_ids)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>data_falster_studies<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 6 × 12
#&gt;   dataset_id   taxon_name site_name context_name observation_id trait_name value
#&gt;   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt;
#&gt; 1 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle 66.1 
#&gt; 2 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  319  
#&gt; 3 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp… simp…
#&gt; 4 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle 71.7 
#&gt; 5 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  562  
#&gt; 6 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp… simp…
#&gt; # ℹ 5 more variables: unit &lt;chr&gt;, date &lt;chr&gt;, value_type &lt;fct&gt;,
#&gt; #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
</div>
</section>
<section id="extracting-by-taxonomic-level" class="level3">
<h3 class="anchored" data-anchor-id="extracting-by-taxonomic-level">Extracting by taxonomic level</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># By family </span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>proteaceae <span class="ot">&lt;-</span> <span class="fu">extract_taxa</span>(austraits, <span class="at">family =</span> <span class="st">"Proteaceae"</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Checking that only taxa in Proteaceae have been extracted</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>proteaceae<span class="sc">$</span>taxa<span class="sc">$</span>family <span class="sc">%&gt;%</span> <span class="fu">unique</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; [1] "Proteaceae"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># By genus </span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>acacia <span class="ot">&lt;-</span> <span class="fu">extract_taxa</span>(austraits, <span class="at">genus =</span> <span class="st">"Acacia"</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Checking that only taxa in Acacia have been extracted</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>acacia<span class="sc">$</span>traits<span class="sc">$</span>taxon_name <span class="sc">%&gt;%</span> <span class="fu">unique</span>() <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; [1] "Acacia myrtifolia" "Acacia suaveolens" "Acacia floribunda"
#&gt; [4] "Acacia celsa"      "Acacia longifolia" "Acacia terminalis"</code></pre>
</div>
</div>
</section>
<section id="extracting-by-trait" class="level3">
<h3 class="anchored" data-anchor-id="extracting-by-trait">Extracting by trait</h3>
<p>Filtering <strong>one particular trait</strong> and assigning it to an object</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>data_wood_dens <span class="ot">&lt;-</span> <span class="fu">extract_trait</span>(austraits, <span class="st">"wood_density"</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data_wood_dens<span class="sc">$</span>traits)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 6 × 12
#&gt;   dataset_id   taxon_name site_name context_name observation_id trait_name value
#&gt;   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;      &lt;dbl&gt;
#&gt; 1 Falster_200… Acacia ce… Atherton  &lt;NA&gt;         Falster_2005_… wood_dens… 0.498
#&gt; 2 Falster_200… Acronychi… Atherton  &lt;NA&gt;         Falster_2005_… wood_dens… 0.525
#&gt; 3 Falster_200… Alphitoni… Atherton  &lt;NA&gt;         Falster_2005_… wood_dens… 0.413
#&gt; 4 Falster_200… Glochidio… Atherton  &lt;NA&gt;         Falster_2005_… wood_dens… 0.566
#&gt; 5 Falster_200… Homalanth… Atherton  &lt;NA&gt;         Falster_2005_… wood_dens… 0.319
#&gt; 6 Falster_200… Melicope … Atherton  &lt;NA&gt;         Falster_2005_… wood_dens… 0.346
#&gt; # ℹ 5 more variables: unit &lt;chr&gt;, date &lt;chr&gt;, value_type &lt;fct&gt;,
#&gt; #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
</div>
<p>Combining <code>lookup_trait</code> with <code>extract_trait</code> to obtain <strong>all traits with ‘leaf’ in the trait name</strong> and assigning it to an object. Note we use the <code>.</code> notation to pass on the <code>lookup_trait</code> results to <code>extract_trait</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>data_leaf <span class="ot">&lt;-</span> <span class="fu">lookup_trait</span>(austraits, <span class="st">"leaf"</span>) <span class="sc">%&gt;%</span> <span class="fu">extract_trait</span>(austraits, .) </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data_leaf<span class="sc">$</span>traits)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 6 × 12
#&gt;   dataset_id   taxon_name site_name context_name observation_id trait_name value
#&gt;   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;      &lt;dbl&gt;
#&gt; 1 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle  66.1
#&gt; 2 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  319  
#&gt; 3 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp…  NA  
#&gt; 4 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle  71.7
#&gt; 5 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  562  
#&gt; 6 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp…  NA  
#&gt; # ℹ 5 more variables: unit &lt;chr&gt;, date &lt;chr&gt;, value_type &lt;fct&gt;,
#&gt; #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
</div>
</section>
</section>
<section id="join-data-from-other-tables-and-elements" class="level2" data-number="32.4">
<h2 data-number="32.4" class="anchored" data-anchor-id="join-data-from-other-tables-and-elements"><span class="header-section-number">32.4</span> Join data from other tables and elements</h2>
<p>Once users have extracted the data they want, they may want to merge other study details into the main <code>traits</code> dataframe for their analyses. For example, users may require taxonomic information for a phylogenetic analysis. This is where the <code>join_</code> functions come in. There are five <code>join_</code> functions in total, each designed to append specific information from other tables and elements in the <code>austraits</code> object. Their suffixes refer to the type of information that is joined, e.g.&nbsp;<code>join_taxonomy</code> appends taxonomic information to the <code>traits</code> dataframe. See <code>?join_all</code> for more details.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Join taxonomic information </span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>(data_leaf <span class="sc">%&gt;%</span> join_taxonomy)<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 6 × 16
#&gt;   dataset_id   taxon_name site_name context_name observation_id trait_name value
#&gt;   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;      &lt;dbl&gt;
#&gt; 1 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle  66.1
#&gt; 2 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  319  
#&gt; 3 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp…  NA  
#&gt; 4 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle  71.7
#&gt; 5 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  562  
#&gt; 6 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp…  NA  
#&gt; # ℹ 9 more variables: unit &lt;chr&gt;, date &lt;chr&gt;, value_type &lt;fct&gt;,
#&gt; #   replicates &lt;chr&gt;, original_name &lt;chr&gt;, family &lt;chr&gt;, genus &lt;chr&gt;,
#&gt; #   taxonRank &lt;chr&gt;, acceptedNameUsageID &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Join methodological information </span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>(data_leaf <span class="sc">%&gt;%</span> join_methods)<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 6 × 16
#&gt;   dataset_id   taxon_name site_name context_name observation_id trait_name value
#&gt;   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;      &lt;dbl&gt;
#&gt; 1 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle  66.1
#&gt; 2 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  319  
#&gt; 3 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp…  NA  
#&gt; 4 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle  71.7
#&gt; 5 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  562  
#&gt; 6 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp…  NA  
#&gt; # ℹ 9 more variables: unit &lt;chr&gt;, date &lt;chr&gt;, value_type &lt;fct&gt;,
#&gt; #   replicates &lt;chr&gt;, original_name &lt;chr&gt;, methods &lt;chr&gt;,
#&gt; #   year_collected_start &lt;chr&gt;, year_collected_end &lt;chr&gt;, collection_type &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Join site based information </span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>(data_leaf <span class="sc">%&gt;%</span> join_sites)<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 6 × 14
#&gt;   dataset_id   taxon_name site_name context_name observation_id trait_name value
#&gt;   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;      &lt;dbl&gt;
#&gt; 1 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle  66.1
#&gt; 2 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  319  
#&gt; 3 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp…  NA  
#&gt; 4 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle  71.7
#&gt; 5 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  562  
#&gt; 6 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp…  NA  
#&gt; # ℹ 7 more variables: unit &lt;chr&gt;, date &lt;chr&gt;, value_type &lt;fct&gt;,
#&gt; #   replicates &lt;chr&gt;, original_name &lt;chr&gt;, `latitude (deg)` &lt;chr&gt;,
#&gt; #   `longitude (deg)` &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Join context information </span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>(data_leaf <span class="sc">%&gt;%</span> join_contexts)<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 6 × 16
#&gt;   dataset_id   taxon_name site_name context_name observation_id trait_name value
#&gt;   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;      &lt;dbl&gt;
#&gt; 1 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle  66.1
#&gt; 2 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  319  
#&gt; 3 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp…  NA  
#&gt; 4 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle  71.7
#&gt; 5 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  562  
#&gt; 6 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp…  NA  
#&gt; # ℹ 9 more variables: unit &lt;chr&gt;, date &lt;chr&gt;, value_type &lt;fct&gt;,
#&gt; #   replicates &lt;chr&gt;, original_name &lt;chr&gt;, type &lt;chr&gt;, description &lt;chr&gt;,
#&gt; #   leaf_parts &lt;chr&gt;, plant_age &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Alternatively users can join *all* information </span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>(data_leaf <span class="sc">%&gt;%</span> join_all)<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 6 × 22
#&gt;   dataset_id   taxon_name site_name context_name observation_id trait_name value
#&gt;   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;      &lt;dbl&gt;
#&gt; 1 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle  66.1
#&gt; 2 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  319  
#&gt; 3 Falster_2003 Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp…  NA  
#&gt; 4 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle  71.7
#&gt; 5 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  562  
#&gt; 6 Falster_2003 Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp…  NA  
#&gt; # ℹ 15 more variables: unit &lt;chr&gt;, date &lt;chr&gt;, value_type &lt;fct&gt;,
#&gt; #   replicates &lt;chr&gt;, original_name &lt;chr&gt;, `latitude (deg)` &lt;chr&gt;,
#&gt; #   `longitude (deg)` &lt;chr&gt;, family &lt;chr&gt;, genus &lt;chr&gt;, taxonRank &lt;chr&gt;,
#&gt; #   acceptedNameUsageID &lt;chr&gt;, methods &lt;chr&gt;, year_collected_start &lt;chr&gt;,
#&gt; #   year_collected_end &lt;chr&gt;, collection_type &lt;chr&gt;</code></pre>
</div>
</div>
</section>
<section id="visualising-data-by-site" class="level2" data-number="32.5">
<h2 data-number="32.5" class="anchored" data-anchor-id="visualising-data-by-site"><span class="header-section-number">32.5</span> Visualising data by site</h2>
<p><code>plot_site_locations</code> graphically summarises where trait data was collected from and how much data is available. The legend refers to the number of neighbouring points: the warmer the colour, the more data that is available. This function only works for studies that are geo-referenced. Users must first use <code>join_sites</code> to append latitude and longitude information into the trait dataframe before plotting</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>data_wood_dens <span class="ot">&lt;-</span> data_wood_dens <span class="sc">%&gt;%</span> <span class="fu">join_sites</span>()</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_site_locations</span>(data_wood_dens<span class="sc">$</span>traits)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="tmp/figures/site_plot-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="visualising-data-distribution-and-variance" class="level2" data-number="32.6">
<h2 data-number="32.6" class="anchored" data-anchor-id="visualising-data-distribution-and-variance"><span class="header-section-number">32.6</span> Visualising data distribution and variance</h2>
<p><code>plot_trait_distribution</code> creates histograms and <a href="https://github.com/eclarke/ggbeeswarm">beeswarm plots</a> for specific traits to help users visualise the variance of the data. Users can specify whether to create separate beeswarm plots at the level of taxonomic family or for each <code>dataset_id</code></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>austraits <span class="sc">%&gt;%</span> <span class="fu">plot_trait_distribution_beeswarm</span>(<span class="st">"wood_density"</span>, <span class="st">"family"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="tmp/figures/beeswarm-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>austraits <span class="sc">%&gt;%</span> <span class="fu">plot_trait_distribution_beeswarm</span>(<span class="st">"wood_density"</span>, <span class="st">"dataset_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="tmp/figures/beeswarm-2.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="pivotting-from-long-to-wide-format" class="level2" data-number="32.7">
<h2 data-number="32.7" class="anchored" data-anchor-id="pivotting-from-long-to-wide-format"><span class="header-section-number">32.7</span> Pivotting from long to wide format</h2>
<p>The table of traits in AusTraits comes in <strong>long</strong> format, where data for all trait information are denoted by two columns called <code>trait_name</code> and <code>value</code>. You can convert this to wide format, where each trait is in a separate column, using the function <code>trait_pivot_wider</code>.</p>
<p>The function operates differently depending on the version of AusTraits</p>
<section id="pivot-wider-3.0.2" class="level3 tabset">
<h3 class="tabset anchored" data-anchor-id="pivot-wider-3.0.2">Pivot wider &lt;=3.0.2</h3>
<p>In AusTraits &lt;=3.0.2, some studies will have <em>multiple observations</em> for some traits, e.g.&nbsp;huber values. This can prevent pivoting from long format to wide. There are two ways to collapse multiple observations. The first option <code>bind_trait_values</code> concatenates the observations into a single string, thereby retaining all information.</p>
<p>Alternatively, if you don’t want to bind the trait values, you can give <code>summarise_trait_means</code> a go, which computes means with multiple observations. Note that this method condenses the traits table and you won’t be able to revert it back unless you <code>load_austraits</code> again</p>
<section id="bind_trait_values" class="level4">
<h4 class="anchored" data-anchor-id="bind_trait_values"><code>bind_trait_values</code></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>data_wide_bound <span class="ot">&lt;-</span> data_falster_studies<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_trait_values</span>() <span class="sc">%&gt;%</span> <span class="co"># Joining multiple obs with `--`</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">trait_pivot_wider</span>() <span class="co">#Pivot wide</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>data_wide_bound<span class="sc">$</span>value <span class="co"># The trait values table</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 103 × 17
#&gt;    dataset_id   taxon_name      site_name context_name observation_id leaf_angle
#&gt;    &lt;chr&gt;        &lt;chr&gt;           &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;     
#&gt;  1 Falster_2003 Acacia myrtifo… Ku-ring-… &lt;NA&gt;         Falster_2003_… 66.1      
#&gt;  2 Falster_2003 Acacia suaveol… Ku-ring-… &lt;NA&gt;         Falster_2003_… 71.7      
#&gt;  3 Falster_2003 Angophora hisp… Ku-ring-… &lt;NA&gt;         Falster_2003_… 50.8      
#&gt;  4 Falster_2003 Banksia margin… Ku-ring-… &lt;NA&gt;         Falster_2003_… 53.1      
#&gt;  5 Falster_2003 Banksia oblong… Ku-ring-… &lt;NA&gt;         Falster_2003_… 45        
#&gt;  6 Falster_2003 Boronia pinnata Ku-ring-… &lt;NA&gt;         Falster_2003_… 43.9      
#&gt;  7 Falster_2003 Conospermum lo… Ku-ring-… &lt;NA&gt;         Falster_2003_… 72.3      
#&gt;  8 Falster_2003 Epacris pulche… Ku-ring-… &lt;NA&gt;         Falster_2003_… 42.9      
#&gt;  9 Falster_2003 Eriostemon aus… Ku-ring-… &lt;NA&gt;         Falster_2003_… 62.1      
#&gt; 10 Falster_2003 Corymbia gummi… Ku-ring-… &lt;NA&gt;         Falster_2003_… 59.3      
#&gt; # ℹ 93 more rows
#&gt; # ℹ 11 more variables: leaf_area &lt;chr&gt;, leaf_compoundness &lt;chr&gt;,
#&gt; #   branch_mass_fraction &lt;chr&gt;, huber_value &lt;chr&gt;, leaf_N_per_dry_mass &lt;chr&gt;,
#&gt; #   seed_mass &lt;chr&gt;, specific_leaf_area &lt;chr&gt;, wood_density &lt;chr&gt;,
#&gt; #   fire_response &lt;chr&gt;, plant_height &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check out the 'bounded' trait values</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>data_wide_bound<span class="sc">$</span>value <span class="sc">%&gt;%</span> </span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)) <span class="sc">%&gt;%</span> <span class="co">#Excluding values that are not traits so we can see which columns contains bounded values</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_all</span>(<span class="at">.vars_predicate =</span> <span class="fu">any_vars</span>(<span class="fu">str_detect</span>(., <span class="st">"--"</span>))) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 55 × 12
#&gt;    leaf_angle leaf_area leaf_compoundness branch_mass_fraction huber_value      
#&gt;    &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;             &lt;chr&gt;                &lt;chr&gt;            
#&gt;  1 &lt;NA&gt;       2786      &lt;NA&gt;              0.67                 0.00033590863285…
#&gt;  2 &lt;NA&gt;       14302     &lt;NA&gt;              0.48                 0.00013227513227…
#&gt;  3 &lt;NA&gt;       6820      &lt;NA&gt;              0.42                 0.00023041474654…
#&gt;  4 &lt;NA&gt;       3209      &lt;NA&gt;              0.41                 0.00050175614651…
#&gt;  5 &lt;NA&gt;       10682     &lt;NA&gt;              0.47                 0.00046816479400…
#&gt;  6 &lt;NA&gt;       6955      &lt;NA&gt;              0.58                 0.00023803856224…
#&gt;  7 &lt;NA&gt;       5228      &lt;NA&gt;              0.58                 0.00059311981020…
#&gt;  8 &lt;NA&gt;       6806      &lt;NA&gt;              0.42                 0.00016644474034…
#&gt;  9 &lt;NA&gt;       11157     &lt;NA&gt;              0                    0.00020092425155…
#&gt; 10 &lt;NA&gt;       3401      &lt;NA&gt;              0.34                 0.00023142791020…
#&gt; # ℹ 45 more rows
#&gt; # ℹ 7 more variables: leaf_N_per_dry_mass &lt;chr&gt;, seed_mass &lt;chr&gt;,
#&gt; #   specific_leaf_area &lt;chr&gt;, wood_density &lt;chr&gt;, fire_response &lt;chr&gt;,
#&gt; #   plant_height &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
</div>
<p>If you would like to revert the bounded trait values, you have to use <code>trait_pivot_longer</code> first, then call <code>separate_trait_values</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>data_wide_bound <span class="sc">%&gt;%</span> </span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">trait_pivot_longer</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate_trait_values</span>(., austraits<span class="sc">$</span>definitions)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 691 × 12
#&gt;    dataset_id  taxon_name site_name context_name observation_id trait_name value
#&gt;    &lt;chr&gt;       &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt;
#&gt;  1 Falster_20… Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle 66.1 
#&gt;  2 Falster_20… Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  319  
#&gt;  3 Falster_20… Acacia my… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp… simp…
#&gt;  4 Falster_20… Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle 71.7 
#&gt;  5 Falster_20… Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  562  
#&gt;  6 Falster_20… Acacia su… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp… simp…
#&gt;  7 Falster_20… Angophora… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle 50.8 
#&gt;  8 Falster_20… Angophora… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_area  1590 
#&gt;  9 Falster_20… Angophora… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_comp… simp…
#&gt; 10 Falster_20… Banksia m… Ku-ring-… &lt;NA&gt;         Falster_2003_… leaf_angle 53.1 
#&gt; # ℹ 681 more rows
#&gt; # ℹ 5 more variables: unit &lt;chr&gt;, date &lt;chr&gt;, value_type &lt;fct&gt;,
#&gt; #   replicates &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
</div>
</section>
<section id="summarise_trait_means" class="level4">
<h4 class="anchored" data-anchor-id="summarise_trait_means"><code>summarise_trait_means</code></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>data_wide_summarised <span class="ot">&lt;-</span> data_falster_studies<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise_trait_means</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">trait_pivot_wider</span>()</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>data_wide_summarised<span class="sc">$</span>value <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 6 × 17
#&gt;   dataset_id   taxon_name       site_name context_name observation_id leaf_angle
#&gt;   &lt;chr&gt;        &lt;chr&gt;            &lt;chr&gt;     &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt;
#&gt; 1 Falster_2003 Acacia myrtifol… Ku-ring-… &lt;NA&gt;         Falster_2003_…       66.1
#&gt; 2 Falster_2003 Acacia suaveole… Ku-ring-… &lt;NA&gt;         Falster_2003_…       71.7
#&gt; 3 Falster_2003 Angophora hispi… Ku-ring-… &lt;NA&gt;         Falster_2003_…       50.8
#&gt; 4 Falster_2003 Banksia margina… Ku-ring-… &lt;NA&gt;         Falster_2003_…       53.1
#&gt; 5 Falster_2003 Banksia oblongi… Ku-ring-… &lt;NA&gt;         Falster_2003_…       45  
#&gt; 6 Falster_2003 Boronia pinnata  Ku-ring-… &lt;NA&gt;         Falster_2003_…       43.9
#&gt; # ℹ 11 more variables: leaf_area &lt;dbl&gt;, leaf_compoundness &lt;dbl&gt;,
#&gt; #   branch_mass_fraction &lt;dbl&gt;, leaf_N_per_dry_mass &lt;dbl&gt;, seed_mass &lt;dbl&gt;,
#&gt; #   specific_leaf_area &lt;dbl&gt;, wood_density &lt;dbl&gt;, huber_value &lt;dbl&gt;,
#&gt; #   fire_response &lt;dbl&gt;, plant_height &lt;dbl&gt;, original_name &lt;chr&gt;</code></pre>
</div>
</div>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./spatial_data_example.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">Tutorial: Using AusTraits with spatial data</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./help.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">Getting help</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">How to get help: <a href="https://traitecoevo.github.io/traits.build-book/help.html" class="uri">https://traitecoevo.github.io/traits.build-book/help.html</a><br>Copyright 2023, Daniel Falster and Elizabeth Wenk</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>