<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>30&nbsp; AusTraits tutorial – The {traits.build} data model, R package, and workflow</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./traits_and_climate_example.html" rel="next">
<link href="./austraits_database.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-cd7de1037569933fbb609f06423bd096.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./austraits_database.html">Using outputs of <code>traits.build</code></a></li><li class="breadcrumb-item"><a href="./AusTraits_tutorial.html"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">AusTraits tutorial</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">The {traits.build} data model, R package, and workflow</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Introduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">About</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./motivation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Motivation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./usage_examples.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Usage examples</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Data structure and standard</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./long_wide.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Long vs Wide data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_standard.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data standard</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_structure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Data structure</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_metadata.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Overview of traits.build ontology</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dictionary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Trait dictionary</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Creating with <code>traits.build</code></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./traits_build.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">The package</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_compilation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Tutorial: Example compilation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./file_organisation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">File organisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./create_dictionary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Creating a dictionary</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./adding_data_brief.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Adding datasets, an introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./publishing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Publishing</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Guide to adding data</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Tutorial: Adding datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Tutorial 1: Adding a simple dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Tutorial 2: Adding a more complex dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Tutorial 3: Adding contexts and complex units</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Tutorial 4: Additional complexities</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Tutorial 5: Multiple columns for a trait</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_6.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Tutorial 6: Data with repeat measurements</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_7.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Tutorial 7: Adding long format dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./adding_data_long.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Adding datasets, a lengthy guide</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./check_dataset_functions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">Check data quality functions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_common_issues.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">Common issues</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./github.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">Using GitHub</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./versioned_releases.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">Version updating &amp; making a new release</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Using outputs of <code>traits.build</code></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./austraits_database.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">AusTraits, Australia’s Plant Trait Database</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./AusTraits_tutorial.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">AusTraits tutorial</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./traits_and_climate_example.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">Analysis example: Using AusTraits for trait-climate analyses</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spatial_data_example.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">Analysis example: Using AusTraits with spatial data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./austraits_package.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">The <code>austraits</code> package</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Getting help</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./help.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">34</span>&nbsp; <span class="chapter-title">Getting help</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./csv.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">CSV files</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./yaml.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Yaml files</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">30.1</span> Introduction</a></li>
  <li><a href="#download-austraits-data" id="toc-download-austraits-data" class="nav-link" data-scroll-target="#download-austraits-data"><span class="header-section-number">30.2</span> Download AusTraits data</a></li>
  <li><a href="#exploring" id="toc-exploring" class="nav-link" data-scroll-target="#exploring"><span class="header-section-number">30.3</span> A first look at data</a></li>
  <li><a href="#database_structure" id="toc-database_structure" class="nav-link" data-scroll-target="#database_structure"><span class="header-section-number">30.4</span> The database structure</a>
  <ul class="collapse">
  <li><a href="#traits-table" id="toc-traits-table" class="nav-link" data-scroll-target="#traits-table">Traits table</a></li>
  <li><a href="#ancillary-data-tables" id="toc-ancillary-data-tables" class="nav-link" data-scroll-target="#ancillary-data-tables">Ancillary data tables</a></li>
  </ul></li>
  <li><a href="#exploring-austraits" id="toc-exploring-austraits" class="nav-link" data-scroll-target="#exploring-austraits"><span class="header-section-number">30.5</span> Exploring AusTraits</a>
  <ul class="collapse">
  <li><a href="#austraits.r-function-reference" id="toc-austraits.r-function-reference" class="nav-link" data-scroll-target="#austraits.r-function-reference">austraits.R function reference</a></li>
  <li><a href="#summarise_functions" id="toc-summarise_functions" class="nav-link" data-scroll-target="#summarise_functions">Summarising data: data coverage</a></li>
  </ul></li>
  <li><a href="#wrangling-austraits" id="toc-wrangling-austraits" class="nav-link" data-scroll-target="#wrangling-austraits"><span class="header-section-number">30.6</span> Wrangling AusTraits</a>
  <ul class="collapse">
  <li><a href="#filtering_functions" id="toc-filtering_functions" class="nav-link" data-scroll-target="#filtering_functions">Filtering data</a></li>
  <li><a href="#join_functions" id="toc-join_functions" class="nav-link" data-scroll-target="#join_functions">Joining relational tables</a></li>
  <li><a href="#combining-extract_-and-join_-functions" id="toc-combining-extract_-and-join_-functions" class="nav-link" data-scroll-target="#combining-extract_-and-join_-functions">Combining <code>extract_</code> and <code>join_</code> functions</a></li>
  <li><a href="#binding-datasets" id="toc-binding-datasets" class="nav-link" data-scroll-target="#binding-datasets">Binding datasets</a></li>
  </ul></li>
  <li><a href="#summarising-data-trait-means-modes-etc." id="toc-summarising-data-trait-means-modes-etc." class="nav-link" data-scroll-target="#summarising-data-trait-means-modes-etc."><span class="header-section-number">30.7</span> Summarising data: trait means, modes, etc.</a>
  <ul class="collapse">
  <li><a href="#categorical-traits" id="toc-categorical-traits" class="nav-link" data-scroll-target="#categorical-traits">Categorical traits</a></li>
  <li><a href="#numeric-traits" id="toc-numeric-traits" class="nav-link" data-scroll-target="#numeric-traits">Numeric traits</a></li>
  </ul></li>
  <li><a href="#plotting_functions" id="toc-plotting_functions" class="nav-link" data-scroll-target="#plotting_functions"><span class="header-section-number">30.8</span> Plotting data</a>
  <ul class="collapse">
  <li><a href="#plotting-trait-distributions" id="toc-plotting-trait-distributions" class="nav-link" data-scroll-target="#plotting-trait-distributions">Plotting trait distributions</a></li>
  <li><a href="#plotting-data-distribution-by-location" id="toc-plotting-data-distribution-by-location" class="nav-link" data-scroll-target="#plotting-data-distribution-by-location">Plotting data distribution by location</a></li>
  <li><a href="#more-complex-workflows-some-examples" id="toc-more-complex-workflows-some-examples" class="nav-link" data-scroll-target="#more-complex-workflows-some-examples">More complex workflows – some examples</a></li>
  </ul></li>
  <li><a href="#a-complexity-pivoting-datasets-pivotting_datasets" id="toc-a-complexity-pivoting-datasets-pivotting_datasets" class="nav-link" data-scroll-target="#a-complexity-pivoting-datasets-pivotting_datasets"><span class="header-section-number">30.9</span> A complexity: pivoting datasets {pivotting_datasets}</a></li>
  <li><a href="#intepreting-trait-names-taxon-names" id="toc-intepreting-trait-names-taxon-names" class="nav-link" data-scroll-target="#intepreting-trait-names-taxon-names"><span class="header-section-number">30.10</span> Intepreting trait names, taxon names</a>
  <ul class="collapse">
  <li><a href="#trait-dictionary" id="toc-trait-dictionary" class="nav-link" data-scroll-target="#trait-dictionary">Trait dictionary</a></li>
  <li><a href="#understanding-taxon-names" id="toc-understanding-taxon-names" class="nav-link" data-scroll-target="#understanding-taxon-names">Understanding taxon names</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./austraits_database.html">Using outputs of <code>traits.build</code></a></li><li class="breadcrumb-item"><a href="./AusTraits_tutorial.html"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">AusTraits tutorial</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">AusTraits tutorial</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="introduction" class="level2" data-number="30.1">
<h2 data-number="30.1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">30.1</span> Introduction</h2>
<p>With more than 1.8 million data records, AusTraits is Australia’s <a href="austraits_database.html#plant_database">largest plant trait database</a>, created using the <a href="https://github.com/traitecoevo/traits.build"><code>{traits.build}</code> R package</a></p>
<p>This tutorial introduces:</p>
<ul>
<li><p><a href="#database_structure">the database structure</a></p></li>
<li><p><a href="#austraits_functions"><code>{austraits}</code> R package functions</a></p></li>
<li><p>additional <a href="#sample_analyses">examples of analyses</a> using the database</p></li>
</ul>
<p>To access more information about <code>traits.build</code>, see <a href="https://traitecoevo.github.io/traits.build-book/"><code>traits.build-book</code></a></p>
<p>Or you can visit the Github repositories for individual packages/data repos:</p>
<ul>
<li><p>the database structure: <a href="https://github.com/traitecoevo/traits.build"><code>traits.build</code></a></p></li>
<li><p>the database contents: <a href="https://github.com/traitecoevo/austraits.build"><code>austraits.build</code></a></p></li>
<li><p>an R package for exploring and wrangling the data: <a href="https://github.com/traitecoevo/austraits"><code>austraits</code></a></p></li>
</ul>
</section>
<section id="download-austraits-data" class="level2" data-number="30.2">
<h2 data-number="30.2" class="anchored" data-anchor-id="download-austraits-data"><span class="header-section-number">30.2</span> Download AusTraits data</h2>
<p>Before you begin, download and source essential packages and functions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Packages need to be installed the first time you use them.</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># They are commented out here, so they aren't reinstalled each time you run the code, </span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># but install any packages you require the first time you run this tutorial.</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("readr", "tidyr", "dplyr", "stringr", "remotes")</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co">#remotes::install_github("traitecoevo/austraits", dependencies = TRUE, upgrade = "ask")</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co">#remotes::install_github("traitecoevo/traits.build", dependencies = TRUE, upgrade = "ask")</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(austraits)    <span class="co"># functions for exploring a traits.build database, available at Github repo</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co"># XX library(traits.build) # additional functions for exploring a traits.build database, available at Github repo</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="co">#source("https://raw.githubusercontent.com/traitecoevo/traits.build-book/master/data/extra_functions.R")</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"data/extra_functions.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then download (or build) the latest AusTraits database, using one of the methods described <a href="austraits_database.html#access_data">here</a>.</p>
<p>This tutorial uses the most recent AusTraits release, version 6.0.0</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>most_recent <span class="ot">&lt;-</span> austraits<span class="sc">::</span><span class="fu">get_versions</span>() <span class="sc">%&gt;%</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">pull</span>(<span class="st">"doi"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">first</span>()</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>most_recent</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "10.5281/zenodo.11188867"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>austraits <span class="ot">&lt;-</span> austraits<span class="sc">::</span><span class="fu">load_austraits</span>(<span class="at">doi =</span> most_recent)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Downloading AusTraits to 'data/austraits'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading data from 'data/austraits/austraits-6.0.0.rds'</code></pre>
</div>
</div>
</section>
<section id="exploring" class="level2" data-number="30.3">
<h2 data-number="30.3" class="anchored" data-anchor-id="exploring"><span class="header-section-number">30.3</span> A first look at data</h2>
<p>If you’re not familiar with AusTraits, you may want to begin by exploring the breadth and depth of data within the database. The database can be explored by trait name, species, or genus using either <a href="#austraits_functions">austraits functions</a> or dplyr functions.</p>
<p><em>How many taxa have <code>leaf_N_per_dry_mass</code> data in AusTraits?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>(austraits <span class="sc">%&gt;%</span> </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_name =</span> <span class="st">"leaf_N_per_dry_mass"</span>))<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name) <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2230</code></pre>
</div>
</div>
<p><em>How are these data distributed across datasets?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_trait_distribution_beeswarm</span>(<span class="at">database =</span> austraits, <span class="at">trait_name =</span> <span class="st">"leaf_N_per_dry_mass"</span>, <span class="at">y_axis_category =</span> <span class="st">"dataset_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><em>How much data exist for other nitrogen traits?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"_N_"</span>) <span class="ot">-&gt;</span> N_traits</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>austraits <span class="sc">%&gt;%</span> </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_name =</span> N_traits) <span class="sc">%&gt;%</span> </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">summarise_database</span>(<span class="at">var =</span> <span class="st">"trait_name"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(<span class="sc">-</span>n_taxa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 16 × 5
   trait_name                           n_records n_dataset n_taxa percent_total
   &lt;chr&gt;                                    &lt;int&gt;     &lt;int&gt;  &lt;int&gt;         &lt;dbl&gt;
 1 leaf_N_per_dry_mass                      11314        87   2230      0.657   
 2 leaf_N_per_area                           3726        30    708      0.216   
 3 leaf_senesced_N_per_dry_mass               542         9     81      0.0315  
 4 wood_N_per_dry_mass                        539         8     69      0.0313  
 5 bark_N_per_dry_mass                        413         5     49      0.024   
 6 seed_N_per_seed_dry_mass                    47         6     42      0.00273 
 7 leaf_rubisco_N_per_total_leaf_N             68         1     29      0.00395 
 8 leaf_thylakoid_N_per_total_leaf_N           70         1     29      0.00407 
 9 stem_N_per_dry_mass                        157         3     28      0.00912 
10 fruit_N_per_dry_mass                        31         5     22      0.0018  
11 leaf_cell_wall_N_per_cell_wall_dry_…        29         1     22      0.00168 
12 leaf_cell_wall_N_per_leaf_N                 29         1     22      0.00168 
13 root_N_per_dry_mass                        141         4     22      0.00819 
14 leaf_N_resorption                           86         1     14      0.005   
15 wood_dead_N_per_dry_mass                     9         3      7      0.000523
16 flower_N_per_dry_mass                       10         2      2      0.000581</code></pre>
</div>
</div>
<p><em>How many “hydraulic” traits are in AusTraits? How much data exist for these traits?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"hydraulic"</span>) <span class="ot">-&gt;</span> hydraulic_traits</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>austraits <span class="sc">%&gt;%</span> </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_name =</span> hydraulic_traits) <span class="sc">%&gt;%</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">summarise_database</span>(<span class="at">var =</span> <span class="st">"trait_name"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(<span class="sc">-</span>n_taxa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 5
   trait_name                           n_records n_dataset n_taxa percent_total
   &lt;chr&gt;                                    &lt;int&gt;     &lt;int&gt;  &lt;int&gt;         &lt;dbl&gt;
 1 stem_vessel_diameter_hydraulic             741         9    264       0.285  
 2 sapwood_specific_hydraulic_conducti…       419         9    179       0.161  
 3 leaf_specific_hydraulic_conductivity       472         9    168       0.181  
 4 sapwood_specific_hydraulic_conducti…       343         6    164       0.132  
 5 leaf_specific_hydraulic_conductance         81         2     79       0.0311 
 6 stem_hydraulic_conductivity                261         5     51       0.1    
 7 hydraulic_safety_margin_50                  46         2     31       0.0177 
 8 hydraulic_safety_margin_88                  40         2     26       0.0154 
 9 leaf_hydraulic_vulnerability                20         1     20       0.00768
10 stem_specific_hydraulic_conductivity       181         3     14       0.0695 </code></pre>
</div>
</div>
<p><em>Where have trait data for Acacia aneura been collected?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">taxon_name =</span> <span class="st">"Acacia aneura"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_location_coordinates</span>()</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"taxon_name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><em>Where have data for Hibbertia species been collected?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">genus =</span> <span class="st">"Hibbertia"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_location_coordinates</span>() <span class="sc">%&gt;%</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_taxa</span>(<span class="at">var =</span> <span class="st">"genus"</span>)</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"genus"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="database_structure" class="level2" data-number="30.4">
<h2 data-number="30.4" class="anchored" data-anchor-id="database_structure"><span class="header-section-number">30.4</span> The database structure</h2>
<p>The <code>{traits.build}</code> R package is the workflow that builds AusTraits from its component datasets.</p>
<p>The database is output as a list, a collection of relational tables, described in detail <a href="database_structure.html">here</a>.</p>
<p>A <code>traits.build</code> data object includes both the relational data tables and additional tables documenting database metadata and a traits dictionary.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>austraits</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── This is 6.0.0 of AusTraits: a curated plant trait database for the Australian</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ This database is built using traits.build version 1.1.0.9000</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ This database contains a total of 1726024 records, for 33494 taxa and 497
  traits.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── This object is a 'list' with the following components: ──</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>• traits: A table containing measurements of traits.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>• locations: A table containing observations of location/site characteristics
associated with information in `traits`. Cross referencing between the two
dataframes is possible using combinations of the variables `dataset_id`,
`location_name`.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>• contexts: A table containing observations of contextual characteristics
associated with information in `traits`. Cross referencing between the two
dataframes is possible using combinations of the variables `dataset_id`,
`link_id`, and `link_vals`.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>• methods: A table containing details on methods with which data were
collected, including time frame and source. Cross referencing with the `traits`
table is possible using combinations of the variables `dataset_id`,
`trait_name`.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>• excluded_data: A table of data that did not pass quality test and so were
excluded from the master dataset.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>• taxonomic_updates: A table of all taxonomic changes implemented in the
construction of AusTraits. Changes are determined by comapring against the APC
(Australian Plant Census) and APNI (Australian Plant Names Index).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>• taxa: A table containing details on taxa associated with information in
`traits`. This information has been sourced from the APC (Australian Plant
Census) and APNI (Australian Plant Names Index) and is released under a CC-BY3
license.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>• contributors: A table of people contributing to each study.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>• sources: Bibtex entries for all primary and secondary sources in the
compilation.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>• definitions: A copy of the definitions for all tables and terms. Information
included here was used to process data and generate any documentation for the
study.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>• schema: A copy of the schema for all tables and terms. Information included
here was used to process data and generate any documentation for the study.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>• metadata: Metadata associated with the dataset, including title, creators,
license, subject, funding sources.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>• build_info: A description of the computing environment used to create this
version of the dataset, including version number, git commit and R
session_info.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ To access a component, try using the $ e.g. austraits$traits</code></pre>
</div>
</div>
<section id="traits-table" class="level3">
<h3 class="anchored" data-anchor-id="traits-table">Traits table</h3>
<p>The core AusTraits table is the traits table. It is in “long” format, with each row documenting a single trait measurement.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 26
   dataset_id taxon_name       observation_id trait_name value unit  entity_type
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 ABRS_1981  Acanthocarpus c… 0001           leaf_comp… simp… &lt;NA&gt;  species    
 2 ABRS_1981  Acanthocarpus c… 0001           seed_leng… 3     mm    species    
 3 ABRS_1981  Acanthocarpus h… 0002           leaf_comp… simp… &lt;NA&gt;  species    
 4 ABRS_1981  Acanthocarpus p… 0003           leaf_comp… simp… &lt;NA&gt;  species    
 5 ABRS_1981  Acanthocarpus p… 0004           leaf_comp… simp… &lt;NA&gt;  species    
 6 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 4     mm    species    
 7 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 5     mm    species    
 8 ABRS_1981  Acanthocarpus r… 0005           leaf_comp… simp… &lt;NA&gt;  species    
 9 ABRS_1981  Acanthocarpus r… 0006           leaf_comp… simp… &lt;NA&gt;  species    
10 ABRS_1981  Acanthocarpus r… 0006           seed_leng… 2.5   mm    species    
11 ABRS_1981  Acanthocarpus v… 0007           leaf_comp… simp… &lt;NA&gt;  species    
12 ABRS_1981  Acer pseudoplat… 0008           leaf_phen… deci… &lt;NA&gt;  species    
13 ABRS_1981  Acidonia microc… 0009           leaf_comp… comp… &lt;NA&gt;  species    
14 ABRS_1981  Callitris acumi… 0010           leaf_comp… simp… &lt;NA&gt;  species    
15 ABRS_1981  Callitris arena… 0011           leaf_comp… simp… &lt;NA&gt;  species    
16 ABRS_1981  Callitris pyram… 0012           leaf_comp… simp… &lt;NA&gt;  species    
17 ABRS_1981  Adenanthos acan… 0013           leaf_comp… simp… &lt;NA&gt;  species    
18 ABRS_1981  Adenanthos apic… 0014           leaf_comp… simp… &lt;NA&gt;  species    
19 ABRS_1981  Adenanthos apic… 0014           leaf_leng… 5     mm    species    
20 ABRS_1981  Adenanthos apic… 0014           leaf_leng… 20    mm    species    
# ℹ 19 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;,
#   replicates &lt;chr&gt;, basis_of_record &lt;chr&gt;, life_stage &lt;chr&gt;,
#   population_id &lt;chr&gt;, individual_id &lt;chr&gt;, repeat_measurements_id &lt;chr&gt;,
#   temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;, location_id &lt;chr&gt;,
#   entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;, treatment_context_id &lt;chr&gt;,
#   collection_date &lt;chr&gt;, measurement_remarks &lt;chr&gt;, method_id &lt;chr&gt;,
#   method_context_id &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
</div>
<p>The columns include:</p>
<ul>
<li>core columns<br>

<ul>
<li>dataset_id<br>
</li>
<li>taxon_name<br>
</li>
<li>trait_name<br>
</li>
<li>value (trait value)<br>
</li>
</ul></li>
<li>entity metadata
<ul>
<li>entity_type<br>
</li>
<li>life_stage</li>
</ul></li>
<li>value metadata
<ul>
<li>value_type<br>
</li>
<li>unit<br>
</li>
<li>basis_of_value<br>
</li>
<li>replicate<br>
</li>
<li>basis_of_record<br>
</li>
</ul></li>
<li>additional metadata
<ul>
<li>collection_date</li>
<li>measurement_remarks</li>
</ul></li>
<li>identifiers for specific observations, individuals, etc.
<ul>
<li>observation_id<br>
</li>
<li>individual_id<br>
</li>
<li>population_id<br>
</li>
<li>repeat_measurements_id<br>
</li>
</ul></li>
<li>identifiers that provide links to ancillary tables with additional metadata
<ul>
<li>location_id<br>
</li>
<li>treatment_context_id<br>
</li>
<li>plot_context_id<br>
</li>
<li>entity_context_id<br>
</li>
<li>temporal_context_id<br>
</li>
<li>method_context_id<br>
</li>
<li>method_id<br>
</li>
<li>source_id</li>
</ul></li>
</ul>
</section>
<section id="ancillary-data-tables" class="level3">
<h3 class="anchored" data-anchor-id="ancillary-data-tables">Ancillary data tables</h3>
<p>The remaining metadata accompanying each trait record is recorded across multiple relational tables.</p>
<p>These include:&nbsp;</p>
<ul>
<li>austraits$locations&nbsp;</li>
<li>austraits$contexts&nbsp;</li>
<li>austraits$methods&nbsp;</li>
<li>austraits$taxa&nbsp;</li>
<li>austraits$taxonomic_updates&nbsp;</li>
<li>austraits$contributors&nbsp;</li>
</ul>
<p>Like the core <code>traits</code> table, each is in ‘long’ format.</p>
<p>The tables <code>locations</code>, <code>contexts</code>, <code>methods</code>, <code>taxa</code> and <code>taxonomic_updates</code> include metadata that links seamlessly to individual rows within <code>traits</code>.</p>
<p>A collection of <code>join_</code> functions within <code>austraits</code> join the ancillary tables to the traits table, based on columns shared across tables.</p>
<table class="caption-top table">
<colgroup>
<col style="width: 25%">
<col style="width: 36%">
<col style="width: 38%">
</colgroup>
<thead>
<tr class="header">
<th>table</th>
<th>metadata in table</th>
<th>columns that link to austraits$traits</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>locations</td>
<td>location name, location properties, latitude, longitude</td>
<td>dataset_id, location_id</td>
</tr>
<tr class="even">
<td>contexts</td>
<td>context name, context category (method context, temporal, entity context, plot, treatment), context property</td>
<td>dataset_id, link_id (identifier to link to: method_context_id, temporal_context_id, entity_context_id, plot_context_id, treatment_context_id), link_vals (identifier value to link to)</td>
</tr>
<tr class="odd">
<td>methods</td>
<td>dataset description, dataset sampling strategy, trait collection method, data collectors, data curators, dataset citation, source_id &amp; citation</td>
<td>dataset_id, trait_name, method_id</td>
</tr>
<tr class="even">
<td>taxa</td>
<td>genus, family, scientific name, APC/APNI taxon concept/taxon name identifiers</td>
<td>taxon_name</td>
</tr>
<tr class="odd">
<td>taxonomic_updates</td>
<td>original name (name submitted), aligned name (typos removed; standardised syntax), identifiers for aligned name</td>
<td>dataset_id, taxon_name, original_name</td>
</tr>
<tr class="even">
<td>contributors</td>
<td>people who contributed data, including their ORCIDs, affiliations, roles</td>
<td>dataset_id</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="exploring-austraits" class="level2" data-number="30.5">
<h2 data-number="30.5" class="anchored" data-anchor-id="exploring-austraits"><span class="header-section-number">30.5</span> Exploring AusTraits</h2>
<p>With 1726024 rows of trait values in the main traits table, knowing how to explore the contents is essential.</p>
<p>The R package <a href="https://github.com/traitecoevo/austraits"><code>{austraits}</code></a> offers a collection of functions to explore and wrangle AusTraits data – or indeed any data using the traits.build format.</p>
<p>An austraits package vignette is available <a href="https://traitecoevo.github.io/traits.build-book/austraits_package.html">here</a>.</p>
<p>Function categories include:</p>
<ul>
<li><p><a href="#summarise_functions"><strong>summarise and lookup functions</strong></a>: These functions offer summaries by taxon name or trait, summarising taxa per trait (or other variable), datasets per trait, and observations per trait.</p></li>
<li><p><a href="#filtering_functions"><strong>filtering functions</strong></a>: These functions begin with the word <code>extract</code> and filter all of the relational tables simultaneously.</p></li>
<li><p><a href="#join_functions"><strong>join functions</strong></a>: These functions allow columns from the relational tables to be joined to the core traits table.</p></li>
<li><p><a href="#pivot_functions"><strong>pivot functions</strong></a>: These functions allow the traits table to be pivoted to wide format.</p></li>
<li><p><a href="#plotting_functions"><strong>plotting functions</strong></a>: These functions offer a means of rapidly visualising AusTraits data, either plotting collection locations on a map of Australia or plotting trait values by dataset.</p></li>
</ul>
<section id="austraits.r-function-reference" class="level3">
<h3 class="anchored" data-anchor-id="austraits.r-function-reference">austraits.R function reference</h3>
<p>Reference guide to: <a href="https://traitecoevo.github.io/austraits/reference/index.html">austraits functions</a></p>
</section>
<section id="summarise_functions" class="level3">
<h3 class="anchored" data-anchor-id="summarise_functions">Summarising data: data coverage</h3>
<p>There are two function families for summarising AusTraits data:</p>
<ul>
<li><strong>lookup_()</strong><br>
</li>
<li><strong>summarise_database()</strong></li>
</ul>
<p>Use the function <code>summarise_database</code> to output summaries of total records, datasets with records, and taxa with records across <code>families</code>, <code>genera</code> or <code>traits</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">summarise_database</span>(<span class="at">database =</span> austraits, <span class="at">var =</span> <span class="st">"trait_name"</span>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">100</span><span class="sc">:</span><span class="dv">130</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 31 × 5
   trait_name                      n_records n_dataset n_taxa percent_total
   &lt;chr&gt;                               &lt;int&gt;     &lt;int&gt;  &lt;int&gt;         &lt;dbl&gt;
 1 flower_pollen_length                  701         1    333     0.000406 
 2 flower_structural_carpels_count      1411         2   1003     0.000817 
 3 flower_structural_sex_type           1982         2   1675     0.00115  
 4 flower_style_differentiation         1656         2   1547     0.000959 
 5 flower_style_fusion                   319         2    267     0.000185 
 6 flowering_cues                        278         1    244     0.000161 
 7 flowering_time                      33685        36  18710     0.0195   
 8 fruit_Ca_per_dry_mass                  27         4     21     0.0000156
 9 fruit_K_per_dry_mass                   27         4     21     0.0000156
10 fruit_Mg_per_dry_mass                  27         4     21     0.0000156
# ℹ 21 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">summarise_database</span>(<span class="at">database =</span> austraits, <span class="at">var =</span> <span class="st">"family"</span>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 5
   family           n_records n_dataset n_taxa percent_total
   &lt;chr&gt;                &lt;int&gt;     &lt;int&gt;  &lt;int&gt;         &lt;dbl&gt;
 1 Acanthaceae           3719        57    149     0.00216  
 2 Achariaceae            162        14      3     0.0000939
 3 Actinidiaceae          186        16      3     0.000108 
 4 Agapanthaceae          107        13      3     0.000062 
 5 Aizoaceae             5004        63    102     0.0029   
 6 Akaniaceae             123        16      1     0.0000713
 7 Alismataceae           892        30     20     0.000517 
 8 Alliaceae              561        19     18     0.000325 
 9 Alseuosmiaceae         318        13      3     0.000184 
10 Alstroemeriaceae       175        15      2     0.000101 
11 Amaranthaceae         9873        85    267     0.00572  
12 Amaryllidaceae        1400        38     56     0.000812 
13 Anacardiaceae         2214        64     26     0.00128  
14 Anarthriaceae          779        32     20     0.000452 
15 Annonaceae            4172        50     62     0.00242  
16 Aphanopetalaceae       193        23      3     0.000112 
17 Apiaceae             12759        94    195     0.0074   
18 Apocynaceae          16341       111    238     0.00947  
19 Apodanthaceae           75        11      4     0.0000435
20 Aponogetonaceae        493        19     14     0.000286 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">summarise_database</span>(<span class="at">database =</span> austraits, <span class="at">var =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 5
   genus          n_records n_dataset n_taxa percent_total
   &lt;chr&gt;              &lt;int&gt;     &lt;int&gt;  &lt;int&gt;         &lt;dbl&gt;
 1 (Dockrillia            3         2      1    0.00000174
 2 Abelia                16         4      1    0.00000928
 3 Abelmoschus          271        19      8    0.000157  
 4 Abildgaardia          74         7      2    0.0000429 
 5 Abrodictyum          123        14      3    0.0000713 
 6 Abroma                39         7      2    0.0000226 
 7 Abrophyllum          181        19      3    0.000105  
 8 Abrotanella          183        18      4    0.000106  
 9 Abrus                202        26      3    0.000117  
10 Abutilon            1975        52     54    0.00115   
11 Acacia            109792       191   1558    0.0637    
12 Acaciella             66         5      3    0.0000383 
13 Acaena              1150        44     14    0.000667  
14 Acalypha             561        20     14    0.000325  
15 Acanthocarpus        986        22      9    0.000572  
16 Acanthocereus         22         4      1    0.0000128 
17 Acanthocladium        48        12      1    0.0000278 
18 Acanthospermum       117        16      2    0.0000679 
19 Acanthus             209        19      4    0.000121  
20 Acer                 221        14      4    0.000128  </code></pre>
</div>
</div>
<p>Since this function summarises the variable selected for <strong>ALL</strong> of AusTraits, you may want to first <a href="#filtering_functions">filter</a> the data before summarising by “taxon_name” – or even “trait_name”.</p>
<p>Alternatively, you can look up traits that contain a specific search term:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(<span class="at">database =</span> austraits, <span class="at">term =</span> <span class="st">"leaf"</span>) <span class="sc">%&gt;%</span> <span class="fu">length</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 213</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(<span class="at">database =</span> austraits, <span class="at">term =</span> <span class="st">"leaf"</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "leaf_compoundness"                   
 [2] "leaf_phenology"                      
 [3] "leaf_length"                         
 [4] "leaf_width"                          
 [5] "leaf_margin"                         
 [6] "leaf_shape"                          
 [7] "leaf_lamina_posture"                 
 [8] "leaf_base_shape"                     
 [9] "leaf_lobation"                       
[10] "leaf_phyllotaxis"                    
[11] "leaf_glaucousness"                   
[12] "leaf_arrangement"                    
[13] "leaf_margin_posture"                 
[14] "leaf_pendulousness"                  
[15] "leaf_N_per_dry_mass"                 
[16] "leaf_area"                           
[17] "leaf_delta13C"                       
[18] "leaf_mass_per_area"                  
[19] "leaf_photochemical_reflectance_index"
[20] "leaf_water_band_index"               
[21] "leaf_dry_matter_content"             
[22] "leaf_thickness"                      
[23] "leaf_water_use_efficiency_intrinsic" 
[24] "leaf_hairs_adult_leaves"             
[25] "leaf_hydraulic_vulnerability"        
[26] "leaf_specific_hydraulic_conductance" 
[27] "leaf_turgor_loss_point"              
[28] "leaf_NP_ratio"                       
[29] "leaf_N_per_area"                     
[30] "leaf_P_per_area"                     </code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(<span class="at">database =</span> austraits, <span class="at">term =</span> <span class="st">"_N_"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "leaf_N_per_dry_mass"                    
 [2] "leaf_N_per_area"                        
 [3] "wood_N_per_dry_mass"                    
 [4] "leaf_senesced_N_per_dry_mass"           
 [5] "bark_N_per_dry_mass"                    
 [6] "fruit_N_per_dry_mass"                   
 [7] "seed_N_per_seed_dry_mass"               
 [8] "stem_N_per_dry_mass"                    
 [9] "wood_dead_N_per_dry_mass"               
[10] "flower_N_per_dry_mass"                  
[11] "leaf_N_resorption"                      
[12] "leaf_cell_wall_N_per_cell_wall_dry_mass"
[13] "leaf_cell_wall_N_per_leaf_N"            
[14] "root_N_per_dry_mass"                    
[15] "leaf_rubisco_N_per_total_leaf_N"        
[16] "leaf_thylakoid_N_per_total_leaf_N"      </code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a> <span class="co"># elemental contents use their symbol and are *almost* always in the middle of a trait name</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(<span class="at">database =</span> austraits, <span class="at">term =</span> <span class="st">"photo"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "plant_photosynthetic_organ"                             
 [2] "leaf_photochemical_reflectance_index"                   
 [3] "leaf_photosynthetic_nitrogen_use_efficiency_saturated"  
 [4] "leaf_photosynthetic_phosphorus_use_efficiency_saturated"
 [5] "leaf_photosynthetic_rate_per_area_maximum"              
 [6] "leaf_photosynthetic_rate_per_area_saturated"            
 [7] "leaf_photosynthetic_rate_per_dry_mass_maximum"          
 [8] "leaf_photosynthetic_rate_per_dry_mass_saturated"        
 [9] "leaf_photosynthesis_Jmax_per_area_25C"                  
[10] "leaf_photosynthesis_Vcmax_per_area"                     
[11] "leaf_photosynthesis_Jmax_per_area"                      
[12] "leaf_photosynthetic_rate_per_area_ambient"              
[13] "photosynthetic_pathway"                                 
[14] "leaf_photosynthesis_Jmax_per_mass"                      
[15] "leaf_photosynthesis_Vcmax_per_mass"                     
[16] "leaf_photosynthetic_nitrogen_use_efficiency_maximum"    
[17] "leaf_photosynthetic_phosphorus_use_efficiency_maximum"  
[18] "leaf_photosynthesis_Jmax_over_Vcmax_25C"                
[19] "leaf_photosynthesis_Vcmax_per_area_25C"                 
[20] "bark_photosynthetic_status"                             </code></pre>
</div>
</div>
<p>Also visit the AusTraits Plant Dictionary to learn more about the traits included in AusTraits, https://w3id.org/APD.</p>
<p>You can also search the <code>locations</code> and <code>contexts</code> tables for <code>location_properties</code> and/or <code>context_properties</code> included as metadata for many trait measurements:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_location_property</span>(<span class="at">database =</span> austraits, <span class="at">term =</span> <span class="st">"soil"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "soil type"                               
 [2] "soil P, total (mg/kg)"                   
 [3] "soil pH (H2O)"                           
 [4] "soil P, SE total (mg/kg)"                
 [5] "soil clay (%)"                           
 [6] "soil sand (%)"                           
 [7] "soil silt (%)"                           
 [8] "soil P, extractable (mg/kg)"             
 [9] "soil exchangeable Al (mg/kg)"            
[10] "soil exchangeable Fe (mg/kg)"            
[11] "soil organic carbon (%)"                 
[12] "soil pH (CaCl2)"                         
[13] "soil carbonates content (%)"             
[14] "soil cation exchange capacity (cmol/kg)" 
[15] "soil K (mg/kg)"                          
[16] "soil N, total (%)"                       
[17] "soil P, total"                           
[18] "soil pH"                                 
[19] "soil P, description"                     
[20] "soil N, total (ppm)"                     
[21] "soil N, total (g/100g)"                  
[22] "soil bulk density (g/cm3)"               
[23] "soil cation exchange capacity (meq/kg)"  
[24] "soil conductivity (dS/m)"                
[25] "soil depth to gravel or laterite (m)"    
[26] "soil pH, 1-10 cm"                        
[27] "soil texture"                            
[28] "soil water holding capacity (v/v)"       
[29] "soil C, total (%)"                       
[30] "soil nutrient summary"                   
[31] "soil"                                    
[32] "soil Al (mg/kg)"                         
[33] "soil Ca (mg/kg)"                         
[34] "soil P, available (ppm)"                 
[35] "incubated soil NH4 content (mg/kg)"      
[36] "soil N:P"                                
[37] "soil NH4 content (mg/kg)"                
[38] "soil NO3 content"                        
[39] "soil TEB (cmol/kg)"                      
[40] "soil age"                                
[41] "soil disolved organic nitrogen"          
[42] "soil exchangeable Al (cmol/kg)"          
[43] "soil exchangeable Ca (cmol/kg, meq/100g)"
[44] "soil exchangeable Ca (mg/kg)"            
[45] "soil exchangeable Fe (cmol/kg, meq/100g)"
[46] "soil exchangeable K (cmol/kg, meq/100g)" 
[47] "soil exchangeable K (mg/kg)"             
[48] "soil exchangeable Mg (cmol/kg, meq/100g)"
[49] "soil exchangeable Mg (mg/kg)"            
[50] "soil exchangeable Mn (cmol/kg, meq/100g)"
[51] "soil exchangeable Mn (mg/kg)"            
[52] "soil exchangeable Na (cmol/kg, meq/100g)"
[53] "soil exchangeable Na (mg/kg)"            
[54] "soil Ca (cmol/kg)"                       
[55] "soil Fe (mg/kg)"                         
[56] "soil Mg (mg/kg)"                         
[57] "soil Mn (mg/kg)"                         
[58] "soil P, Olsen (mg/kg)"                   
[59] "soil P, Resin (mg/kg)"                   
[60] "soil organic content (%)"                
[61] "soil organic matter (%)"                 
[62] "soil N, total (mg/g)"                    
[63] "soil temperature (C)"                    
[64] "soil nutrient status"                    
[65] "soil P, total ICP"                       
[66] "soil N, Kjeldahl (mg/kg)"                
[67] "soil N, oxidised (mg/kg)"                
[68] "soil water content (%)"                  
[69] "soil series"                             
[70] "soil K"                                  
[71] "soil depth class"                        
[72] "soil morphology"                         
[73] "soil profile description"                
[74] "soil texture (subsoil)"                  
[75] "soil texture (surface)"                  
[76] "soil C:N"                                
[77] "soil C:P"                                
[78] "soil notes"                              </code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_location_property</span>(<span class="at">database =</span> austraits, <span class="at">term =</span> <span class="st">"temperature"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "temperature, max (C)"             "temperature, MAT (C)"            
 [3] "temperature, mean summer max (C)" "temperature, mean winter max (C)"
 [5] "temperature, max MAT (C)"         "temperature, min MAT (C)"        
 [7] "temperature, mean winter min (C)" "temperature variation (C)"       
 [9] "temperature, monthly max (C)"     "temperature, monthly min (C)"    
[11] "soil temperature (C)"             "temperature, mean daily max (C)" 
[13] "temperature, mean daily min (C)"  "temperature, mean yearly max (C)"
[15] "temperature, summer mean (C)"     "temperature, winter mean (C)"    </code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_context_property</span>(<span class="at">database =</span> austraits, <span class="at">term =</span> <span class="st">"season"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "sampling season" "fire season"    </code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_context_property</span>(<span class="at">database =</span> austraits, <span class="at">term =</span> <span class="st">"fire"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "fire intensity"     "fire history"       "fire response type"
[4] "fire severity"      "fire season"       </code></pre>
</div>
</div>
<p>For instance, to just look at number of records, datasets, and taxa with data for nitrogen-related traits:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>N_traits <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span> </span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_name =</span> <span class="st">"_N_"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">summarise_database</span>(<span class="at">var =</span> <span class="st">"trait_name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="wrangling-austraits" class="level2" data-number="30.6">
<h2 data-number="30.6" class="anchored" data-anchor-id="wrangling-austraits"><span class="header-section-number">30.6</span> Wrangling AusTraits</h2>
<section id="filtering_functions" class="level3">
<h3 class="anchored" data-anchor-id="filtering_functions">Filtering data</h3>
<p>There are four <code>austraits</code> functions that filter data: <code>extract_trait</code>, <code>extract_taxon</code> or <code>extract_dataset_id</code>, and <code>extract_data</code>.</p>
<p>Each of these functions simultaneously filters all database tables to only include trait measurements (and associated metadata) meeting the specified criteria, retaining the original database structure.</p>
<p><em>Note, although the <code>extract_</code> functions were explicitly developed to return the original <code>traits.build</code> database structure, they will also work when the “database” is just a single table, such as if prior manipulations have separated the traits table from the rest of the database.</em></p>
<section id="extract_trait-extract_dataset-and-extract_taxa" class="level4">
<h4 class="anchored" data-anchor-id="extract_trait-extract_dataset-and-extract_taxa"><code>extract_trait</code>, <code>extract_dataset</code> and <code>extract_taxa</code></h4>
<p>Three of the functions extract data based on pre-set columns: 1. <code>extract_trait</code> filters by <code>trait_name</code> 2. <code>extract_dataset</code> filters by <code>dataset_id</code> 3. <code>extract_taxa</code> filters by <code>taxon_name</code>, <code>genus</code> or <code>family</code></p>
<p>Search terms can either be exact or partial matches.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>leaf_mass_per_area_data <span class="ot">&lt;-</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> </span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_names =</span> <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>))</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>Westoby_2014_datasets <span class="ot">&lt;-</span></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="st">"Westoby_2014"</span>)</span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>all_Westoby_datasets <span class="ot">&lt;-</span></span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="st">"Westoby"</span>)</span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a>Eucalyptus_data <span class="ot">&lt;-</span></span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb63-15"><a href="#cb63-15" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">genus =</span> <span class="st">"Eucalyptus"</span>)</span>
<span id="cb63-16"><a href="#cb63-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-17"><a href="#cb63-17" aria-hidden="true" tabindex="-1"></a>Banksia_serrata_data <span class="ot">&lt;-</span></span>
<span id="cb63-18"><a href="#cb63-18" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb63-19"><a href="#cb63-19" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">taxon_name =</span> <span class="st">"Banksia serrata"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="extract_data" class="level4">
<h4 class="anchored" data-anchor-id="extract_data"><code>extract_data</code></h4>
<p><code>extract_data</code> offers the ability of filtering the database based on a value(s) in any column of any of the seven data tables (traits, locations, contexts, methods, taxa, taxonomic_updates, contributors).</p>
<p>See the <a href="database_structure.html">database structure chapter</a> for names and definitions of each column and, for those with controlled vocabulary, their allowed values.</p>
<p>Alternatively to see the list of column names to use:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(austraits<span class="sc">$</span>traits)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"             "taxon_name"             "observation_id"        
 [4] "trait_name"             "value"                  "unit"                  
 [7] "entity_type"            "value_type"             "basis_of_value"        
[10] "replicates"             "basis_of_record"        "life_stage"            
[13] "population_id"          "individual_id"          "repeat_measurements_id"
[16] "temporal_context_id"    "source_id"              "location_id"           
[19] "entity_context_id"      "plot_context_id"        "treatment_context_id"  
[22] "collection_date"        "measurement_remarks"    "method_id"             
[25] "method_context_id"      "original_name"         </code></pre>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(austraits<span class="sc">$</span>methods)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"                       "trait_name"                      
 [3] "methods"                          "method_id"                       
 [5] "description"                      "sampling_strategy"               
 [7] "source_primary_key"               "source_primary_citation"         
 [9] "source_secondary_key"             "source_secondary_citation"       
[11] "source_original_dataset_key"      "source_original_dataset_citation"
[13] "data_collectors"                  "assistants"                      
[15] "dataset_curators"                </code></pre>
</div>
</div>
<p>And to see the list of possible values for a column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(austraits<span class="sc">$</span>traits<span class="sc">$</span>life_stage)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "adult"                     "sapling"                  
 [3] "seedling"                  "adults"                   
 [5] "adult, seedling"           "unknown"                  
 [7] "seedling, adult"           "saplings"                 
 [9] "seedling sapling adult"    "juvenile"                 
[11] "adult juvenile"            NA                         
[13] "adult, sapling [by trait]"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(austraits<span class="sc">$</span>traits<span class="sc">$</span>basis_of_record)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "preserved_specimen"                "field_experiment"                 
 [3] "field"                             "lab"                              
 [5] "field field_experiment"            "literature field"                 
 [7] "literature"                        "unknown"                          
 [9] "captive_cultivated"                "literature field field_experiment"
[11] "field lab"                         "field preserved_specimen"         
[13] NA                                  "field captive_cultivated"         </code></pre>
</div>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(austraits<span class="sc">$</span>locations<span class="sc">$</span>location_property)[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "latitude (deg)"                   "longitude (deg)"                 
 [3] "aridity index (MAP/PET)"          "precipitation, MAP (mm)"         
 [5] "temperature, max (C)"             "locality"                        
 [7] "recorded by"                      "description"                     
 [9] "notes"                            "soil type"                       
[11] "temperature, MAT (C)"             "population"                      
[13] "elevation (m)"                    "biome"                           
[15] "crown height, max (m)"            "leaf area index"                 
[17] "soil P, total (mg/kg)"            "climate description"             
[19] "temperature, mean summer max (C)" "temperature, mean winter max (C)"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(austraits<span class="sc">$</span>contexts<span class="sc">$</span>context_property)[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "trait scoring method"       "entity_measured"           
 [3] "replicate observations"     "entity measured"           
 [5] "sampling season"            "measurement CO2"           
 [7] "seed provenance"            "measurement irradiance"    
 [9] "repeat observations"        "habitat flammability"      
[11] "time to flowering type"     "fire intensity"            
[13] "groundwater requirement"    "slope position"            
[15] "plot elevation (m)"         "inundation duration (days)"
[17] "fire history"               "canopy layer"              
[19] "drought treatment"          "sampling time of day"      </code></pre>
</div>
</div>
<p>The function then allows you to filter down to the components of each table that are relevant to the search criteria specified:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>field_data <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">extract_data</span>(<span class="at">table =</span> <span class="st">"traits"</span>, <span class="at">col =</span> <span class="st">"basis_of_record"</span>, <span class="at">col_value =</span> <span class="st">"field"</span>)</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>field_data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 26
  dataset_id  taxon_name       observation_id trait_name value unit  entity_type
  &lt;chr&gt;       &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
1 Ahrens_2019 Corymbia caloph… 001            leaf_N_pe… 9     mg/g  individual 
2 Ahrens_2019 Corymbia caloph… 001            leaf_area  1288… mm2   individual 
3 Ahrens_2019 Corymbia caloph… 001            leaf_delt… -27.… {del… individual 
4 Ahrens_2019 Corymbia caloph… 001            leaf_mass… 185.… g/m2  individual 
5 Ahrens_2019 Corymbia caloph… 001            leaf_phot… 0.02… {dim… individual 
6 Ahrens_2019 Corymbia caloph… 001            leaf_wate… 1.06… {dim… individual 
# ℹ 19 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;,
#   replicates &lt;chr&gt;, basis_of_record &lt;chr&gt;, life_stage &lt;chr&gt;,
#   population_id &lt;chr&gt;, individual_id &lt;chr&gt;, repeat_measurements_id &lt;chr&gt;,
#   temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;, location_id &lt;chr&gt;,
#   entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;, treatment_context_id &lt;chr&gt;,
#   collection_date &lt;chr&gt;, measurement_remarks &lt;chr&gt;, method_id &lt;chr&gt;,
#   method_context_id &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>data_with_soils_data <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">extract_data</span>(<span class="at">table =</span> <span class="st">"locations"</span>, <span class="at">col =</span> <span class="st">"location_property"</span>, <span class="at">col_value =</span> <span class="st">"soil"</span>)</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>data_with_soils_data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 26
  dataset_id  taxon_name       observation_id trait_name value unit  entity_type
  &lt;chr&gt;       &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
1 Apgaua_2017 Aglaia meridion… 001            leaf_delt… -34.… {del… individual 
2 Apgaua_2017 Aglaia meridion… 001            leaf_wate… 0.02… umol… individual 
3 Apgaua_2017 Aglaia meridion… 001            sapwood_s… 14.6… kg/m… individual 
4 Apgaua_2017 Aglaia meridion… 001            stem_vess… 71.0… {cou… individual 
5 Apgaua_2017 Aglaia meridion… 001            stem_vess… 53.9… um    individual 
6 Apgaua_2017 Aglaia meridion… 001            stem_vess… 82.5… um    individual 
# ℹ 19 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;,
#   replicates &lt;chr&gt;, basis_of_record &lt;chr&gt;, life_stage &lt;chr&gt;,
#   population_id &lt;chr&gt;, individual_id &lt;chr&gt;, repeat_measurements_id &lt;chr&gt;,
#   temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;, location_id &lt;chr&gt;,
#   entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;, treatment_context_id &lt;chr&gt;,
#   collection_date &lt;chr&gt;, measurement_remarks &lt;chr&gt;, method_id &lt;chr&gt;,
#   method_context_id &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>data_with_soils_data<span class="sc">$</span>locations <span class="sc">%&gt;%</span> <span class="fu">head</span>() <span class="co"># all location properties are retained for the measurements for measurements for which at least location property pertains to soil</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  dataset_id  location_id location_name                  location_property value
  &lt;chr&gt;       &lt;chr&gt;       &lt;chr&gt;                          &lt;chr&gt;             &lt;chr&gt;
1 Apgaua_2017 01          Daintree Rainforest Observato… description       lowl…
2 Apgaua_2017 01          Daintree Rainforest Observato… latitude (deg)    -16.…
3 Apgaua_2017 01          Daintree Rainforest Observato… longitude (deg)   145.…
4 Apgaua_2017 01          Daintree Rainforest Observato… precipitation, M… 4900 
5 Apgaua_2017 01          Daintree Rainforest Observato… soil type         high…
6 Apgaua_2017 01          Daintree Rainforest Observato… temperature, MAT… 24.4 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>data_contributed_by_Wright <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">extract_data</span>(<span class="at">table =</span> <span class="st">"contributors"</span>, <span class="at">col =</span> <span class="st">"last_name"</span>, <span class="at">col_value =</span> <span class="st">"Wright"</span>)</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>data_contributed_by_Wright<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 26
  dataset_id     taxon_name    observation_id trait_name value unit  entity_type
  &lt;chr&gt;          &lt;chr&gt;         &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
1 Gallagher_2018 Acacia myrti… 001            leaf_area  199.5 mm2   individual 
2 Gallagher_2018 Acacia myrti… 001            leaf_dry_… 27.51 mg    individual 
3 Gallagher_2018 Acacia myrti… 001            leaf_mass… 137.… g/m2  individual 
4 Gallagher_2018 Acacia myrti… 002            leaf_area  314.… mm2   individual 
5 Gallagher_2018 Acacia myrti… 002            leaf_dry_… 39.4  mg    individual 
6 Gallagher_2018 Acacia myrti… 002            leaf_mass… 125.… g/m2  individual 
# ℹ 19 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;,
#   replicates &lt;chr&gt;, basis_of_record &lt;chr&gt;, life_stage &lt;chr&gt;,
#   population_id &lt;chr&gt;, individual_id &lt;chr&gt;, repeat_measurements_id &lt;chr&gt;,
#   temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;, location_id &lt;chr&gt;,
#   entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;, treatment_context_id &lt;chr&gt;,
#   collection_date &lt;chr&gt;, measurement_remarks &lt;chr&gt;, method_id &lt;chr&gt;,
#   method_context_id &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>data_contributed_by_Wright<span class="sc">$</span>contributors <span class="sc">%&gt;%</span> <span class="fu">head</span>() <span class="co"># all contributors are retained for datasets where at least one of the contributors on the dataset has the last name "Wright"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 6
  dataset_id     last_name         given_name ORCID  affiliation additional_role
  &lt;chr&gt;          &lt;chr&gt;             &lt;chr&gt;      &lt;chr&gt;  &lt;chr&gt;       &lt;chr&gt;          
1 Gallagher_2018 Gallagher         Rachael    0000-… Hawkesbury… contact        
2 Gallagher_2018 Laugier-Kitchener Bree-Anne  0000-… Department… &lt;NA&gt;           
3 Gallagher_2018 McPherson         Hannah     0000-… National H… &lt;NA&gt;           
4 Gallagher_2018 Rossetto          Maurizio   0000-… National H… &lt;NA&gt;           
5 Gallagher_2018 van der Merwe     Marlien    0000-… Research C… &lt;NA&gt;           
6 Gallagher_2018 Wright            Ian        0000-… Hawkesbury… &lt;NA&gt;           </code></pre>
</div>
</div>
<p>Multiple <code>extract_</code>’s can be linked together to rapidly restrict data to the subset desired:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>subset <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_data</span>(<span class="at">table =</span> <span class="st">"traits"</span>, <span class="at">col =</span> <span class="st">"basis_of_record"</span>, <span class="at">col_value =</span> <span class="st">"field"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_name =</span> <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>, <span class="st">"leaf_thickness"</span>, <span class="st">"leaf_length"</span>, <span class="st">"leaf_area"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">genus =</span> <span class="st">"Eucalyptus"</span>)</span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>subset<span class="sc">$</span>traits[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10,067 × 20
   dataset_id      taxon_name  observation_id trait_name value unit  entity_type
   &lt;chr&gt;           &lt;chr&gt;       &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 Atkinson_2020_2 Eucalyptus… 35             leaf_area  4065  mm2   individual 
 2 Atkinson_2020_2 Eucalyptus… 35             leaf_mass… 189.… g/m2  individual 
 3 Atkinson_2020_2 Eucalyptus… 36             leaf_area  4012… mm2   individual 
 4 Atkinson_2020_2 Eucalyptus… 36             leaf_mass… 172.… g/m2  individual 
 5 Atkinson_2020_2 Eucalyptus… 37             leaf_area  2561… mm2   individual 
 6 Atkinson_2020_2 Eucalyptus… 37             leaf_mass… 138.… g/m2  individual 
 7 Atkinson_2020_2 Eucalyptus… 38             leaf_area  2617… mm2   individual 
 8 Atkinson_2020_2 Eucalyptus… 38             leaf_mass… 141.… g/m2  individual 
 9 Atkinson_2020_2 Eucalyptus… 39             leaf_area  4364  mm2   individual 
10 Atkinson_2020_2 Eucalyptus… 39             leaf_mass… 172.… g/m2  individual 
# ℹ 10,057 more rows
# ℹ 13 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;,
#   replicates &lt;chr&gt;, basis_of_record &lt;chr&gt;, life_stage &lt;chr&gt;,
#   population_id &lt;chr&gt;, individual_id &lt;chr&gt;, repeat_measurements_id &lt;chr&gt;,
#   temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;, location_id &lt;chr&gt;,
#   entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;</code></pre>
</div>
</div>
</section>
</section>
<section id="join_functions" class="level3">
<h3 class="anchored" data-anchor-id="join_functions">Joining relational tables</h3>
<p>For many research purposes you will want to join metadata from one of the relational tables to the core traits table. There are eight <code>{austraits}</code> functions that facilitate this by adding the columns you select from the ancillary data tables to the database’s <code>traits</code> table, seven functions that merge information from a single table (<code>join_...</code>) and a function that joins columns from all seven ancillary data tables (<code>flatten_database</code>). All functions output the database with the original database structure allowing you to follow up <code>joining</code> with <code>extracting</code> and to continue <code>joining</code> additional columns.</p>
<section id="joining-location-metadata" class="level4">
<h4 class="anchored" data-anchor-id="joining-location-metadata">Joining location metadata</h4>
<p>The locations table includes information on all location properties measured, including the actual location (latitude/longitude), climatic data, soil properties, fire history, vegetation history, geologic history, etc.</p>
<p>The austraits function <code>join_location_coordinates</code> just adds location name, latitude, and longitude to the core traits table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>traits_with_lat_long <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span> </span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="at">dataset_id =</span> <span class="st">"Westoby"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_location_coordinates</span>()</span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>traits_with_lat_long<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"             "taxon_name"             "observation_id"        
 [4] "trait_name"             "value"                  "unit"                  
 [7] "entity_type"            "value_type"             "basis_of_value"        
[10] "replicates"             "basis_of_record"        "life_stage"            
[13] "population_id"          "individual_id"          "repeat_measurements_id"
[16] "temporal_context_id"    "source_id"              "location_id"           
[19] "entity_context_id"      "plot_context_id"        "treatment_context_id"  
[22] "collection_date"        "measurement_remarks"    "method_id"             
[25] "method_context_id"      "original_name"          "location_name"         
[28] "latitude (deg)"         "longitude (deg)"       </code></pre>
</div>
</div>
<p>The function <code>join_location_properties</code> joins other location properties to the <code>traits</code> table. It has two arguments:</p>
<ol type="1">
<li><p><code>vars</code>, specifies the location properties, via complete or partial string matches, that should be added to the traits table; defaults to “all”</p></li>
<li><p><code>format</code> offers three output formats:</p></li>
</ol>
<pre><code>*   `many_columns` (each location property is added as a separate column)  
*   `single_column_pretty` (all location properties compacted into a single column delimited in a way that is easy for humans to read; this is the default)
*   `single_column_json` (all location properties compacted into a single column, using json formatting)  </code></pre>
<p>Examples of joining location properties to the traits table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="co"># method to add location properties that you know exist from previous database exploration; this example showcases `format = "many_columns"</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>locations1 <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_location_properties</span>(<span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"description"</span>, <span class="st">"aridity index (MAP/PET)"</span>, </span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>                                         <span class="st">"soil type"</span>, <span class="st">"fire history"</span>), <span class="at">format =</span> <span class="st">"many_columns"</span>)</span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>locations1<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"                                
 [2] "taxon_name"                                
 [3] "observation_id"                            
 [4] "trait_name"                                
 [5] "value"                                     
 [6] "unit"                                      
 [7] "entity_type"                               
 [8] "value_type"                                
 [9] "basis_of_value"                            
[10] "replicates"                                
[11] "basis_of_record"                           
[12] "life_stage"                                
[13] "population_id"                             
[14] "individual_id"                             
[15] "repeat_measurements_id"                    
[16] "temporal_context_id"                       
[17] "source_id"                                 
[18] "location_id"                               
[19] "entity_context_id"                         
[20] "plot_context_id"                           
[21] "treatment_context_id"                      
[22] "collection_date"                           
[23] "measurement_remarks"                       
[24] "method_id"                                 
[25] "method_context_id"                         
[26] "original_name"                             
[27] "location_name"                             
[28] "location_property: aridity index (MAP/PET)"
[29] "location_property: description"            
[30] "location_property: soil type"              
[31] "location_property: fire history"           </code></pre>
</div>
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>locations1<span class="sc">$</span>traits[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,726,024 × 10
   dataset_id taxon_name       observation_id trait_name value unit  entity_type
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 ABRS_1981  Acanthocarpus c… 0001           leaf_comp… simp… &lt;NA&gt;  species    
 2 ABRS_1981  Acanthocarpus c… 0001           seed_leng… 3     mm    species    
 3 ABRS_1981  Acanthocarpus h… 0002           leaf_comp… simp… &lt;NA&gt;  species    
 4 ABRS_1981  Acanthocarpus p… 0003           leaf_comp… simp… &lt;NA&gt;  species    
 5 ABRS_1981  Acanthocarpus p… 0004           leaf_comp… simp… &lt;NA&gt;  species    
 6 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 4     mm    species    
 7 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 5     mm    species    
 8 ABRS_1981  Acanthocarpus r… 0005           leaf_comp… simp… &lt;NA&gt;  species    
 9 ABRS_1981  Acanthocarpus r… 0006           leaf_comp… simp… &lt;NA&gt;  species    
10 ABRS_1981  Acanthocarpus r… 0006           seed_leng… 2.5   mm    species    
# ℹ 1,726,014 more rows
# ℹ 3 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;, replicates &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="co"># method where you first lookup location properties using the function `lookup_location_property`; this example showcases `format = "single_column_pretty"</span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>precipitation_properties <span class="ot">&lt;-</span> <span class="fu">lookup_location_property</span>(<span class="at">database =</span> austraits, <span class="at">term =</span> <span class="st">"precipitation"</span>)</span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>locations2 <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span></span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_location_properties</span>(<span class="at">vars =</span> precipitation_properties, <span class="at">format =</span> <span class="st">"single_column_pretty"</span>)</span>
<span id="cb95-6"><a href="#cb95-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-7"><a href="#cb95-7" aria-hidden="true" tabindex="-1"></a>locations2<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"             "taxon_name"             "observation_id"        
 [4] "trait_name"             "value"                  "unit"                  
 [7] "entity_type"            "value_type"             "basis_of_value"        
[10] "replicates"             "basis_of_record"        "life_stage"            
[13] "population_id"          "individual_id"          "repeat_measurements_id"
[16] "temporal_context_id"    "source_id"              "location_id"           
[19] "entity_context_id"      "plot_context_id"        "treatment_context_id"  
[22] "collection_date"        "measurement_remarks"    "method_id"             
[25] "method_context_id"      "original_name"          "location_name"         
[28] "location_properties"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>locations2<span class="sc">$</span>traits[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,726,024 × 10
   dataset_id taxon_name       observation_id trait_name value unit  entity_type
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 ABRS_1981  Acanthocarpus c… 0001           leaf_comp… simp… &lt;NA&gt;  species    
 2 ABRS_1981  Acanthocarpus c… 0001           seed_leng… 3     mm    species    
 3 ABRS_1981  Acanthocarpus h… 0002           leaf_comp… simp… &lt;NA&gt;  species    
 4 ABRS_1981  Acanthocarpus p… 0003           leaf_comp… simp… &lt;NA&gt;  species    
 5 ABRS_1981  Acanthocarpus p… 0004           leaf_comp… simp… &lt;NA&gt;  species    
 6 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 4     mm    species    
 7 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 5     mm    species    
 8 ABRS_1981  Acanthocarpus r… 0005           leaf_comp… simp… &lt;NA&gt;  species    
 9 ABRS_1981  Acanthocarpus r… 0006           leaf_comp… simp… &lt;NA&gt;  species    
10 ABRS_1981  Acanthocarpus r… 0006           seed_leng… 2.5   mm    species    
# ℹ 1,726,014 more rows
# ℹ 3 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;, replicates &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co"># method where you add all location properties; this example showcases `format = "single_column_json"</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>locations3 <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span></span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_location_properties</span>(<span class="at">vars =</span> <span class="st">"all"</span>, <span class="at">format =</span> <span class="st">"single_column_json"</span>)</span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-5"><a href="#cb99-5" aria-hidden="true" tabindex="-1"></a>locations3<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"             "taxon_name"             "observation_id"        
 [4] "trait_name"             "value"                  "unit"                  
 [7] "entity_type"            "value_type"             "basis_of_value"        
[10] "replicates"             "basis_of_record"        "life_stage"            
[13] "population_id"          "individual_id"          "repeat_measurements_id"
[16] "temporal_context_id"    "source_id"              "location_id"           
[19] "entity_context_id"      "plot_context_id"        "treatment_context_id"  
[22] "collection_date"        "measurement_remarks"    "method_id"             
[25] "method_context_id"      "original_name"          "location_name"         
[28] "location_properties"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>locations3<span class="sc">$</span>traits[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,726,024 × 10
   dataset_id taxon_name       observation_id trait_name value unit  entity_type
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 ABRS_1981  Acanthocarpus c… 0001           leaf_comp… simp… &lt;NA&gt;  species    
 2 ABRS_1981  Acanthocarpus c… 0001           seed_leng… 3     mm    species    
 3 ABRS_1981  Acanthocarpus h… 0002           leaf_comp… simp… &lt;NA&gt;  species    
 4 ABRS_1981  Acanthocarpus p… 0003           leaf_comp… simp… &lt;NA&gt;  species    
 5 ABRS_1981  Acanthocarpus p… 0004           leaf_comp… simp… &lt;NA&gt;  species    
 6 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 4     mm    species    
 7 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 5     mm    species    
 8 ABRS_1981  Acanthocarpus r… 0005           leaf_comp… simp… &lt;NA&gt;  species    
 9 ABRS_1981  Acanthocarpus r… 0006           leaf_comp… simp… &lt;NA&gt;  species    
10 ABRS_1981  Acanthocarpus r… 0006           seed_leng… 2.5   mm    species    
# ℹ 1,726,014 more rows
# ℹ 3 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;, replicates &lt;chr&gt;</code></pre>
</div>
</div>
</section>
<section id="joining-contexts-metadata" class="level4">
<h4 class="anchored" data-anchor-id="joining-contexts-metadata">Joining contexts metadata</h4>
<p>The context table documents additional context properties/ancillary data which may be useful for interpreting trait values. Context properties are divided into 5 categories: <code>treatment context</code>, <code>plot context</code>, <code>entity context</code>, <code>temporal context</code>, and <code>method context</code>.</p>
<table class="caption-top table">
<colgroup>
<col style="width: 23%">
<col style="width: 76%">
</colgroup>
<thead>
<tr class="header">
<th>context category</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>treatment context</td>
<td>Context property that is an experimental manipulation, that might affect the trait values measured on an individual, population or species-level entity.</td>
</tr>
<tr class="even">
<td>plot context</td>
<td>Context property that is a feature of a plot (subset of a location) that might affect the trait values measured on an individual, population or species-level entity.</td>
</tr>
<tr class="odd">
<td>entity context</td>
<td>Context property that is information about an organismal entity (individual, population or taxon) that does not comprise a trait-centered observation but might affect the trait values measured on the entity.</td>
</tr>
<tr class="even">
<td>temporal context</td>
<td>Context property that is a feature of a “point in time” that might affect the trait values measured on an individual, population or species-level entity.</td>
</tr>
<tr class="odd">
<td>method context</td>
<td>Context property that records specific information about a measurement method that is modified between measurements.</td>
</tr>
</tbody>
</table>
<p>The austraits function <code>join_context_properties</code> joins context properties to the <code>traits</code> table. It has three arguments:</p>
<ol type="1">
<li><p><code>vars</code>, specifies the context properties, via complete or partial string matches, that should be added to the traits table; defaults to “all”</p></li>
<li><p><code>format</code> offers three output formats:</p></li>
</ol>
<pre><code>*   `many_columns` (each context property is added as a separate column)  
*   `single_column_pretty` (all context properties compacted into a single column delimited in a way that is easy for humans to read; this is the default)
*   `single_column_json` (all context properties compacted into a single column, using json formatting)</code></pre>
<ol start="3" type="1">
<li><code>include_description</code> is a logical argument (TRUE/FALSE) that indicates where the context property value descriptions should be included or excluded when context data are joined; defaults to “TRUE”</li>
</ol>
<p>Each category of context property is added to a separate column for the compacted columns, retaining this important information about the different groupings of context properties. When <code>format = "many_columns"</code> is selected the context category is indicated in the column name.</p>
<p>Examples of joining context properties to the traits table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="co"># method to add context properties that you know exist from previous database exploration; this example showcases `format = "many_columns"</span></span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>contexts1 <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span></span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_context_properties</span>(</span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"sampling season"</span>, <span class="st">"plant sex"</span>, <span class="st">"leaf surface"</span>, <span class="st">"leaf age"</span>, <span class="st">"fire intensity"</span>,</span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>             <span class="st">"slope position"</span>, <span class="st">"fire season"</span>, <span class="st">"drought treatment"</span>, <span class="st">"temperature treatment"</span>), </span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">format =</span> <span class="st">"many_columns"</span>,</span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">include_description =</span> <span class="cn">TRUE</span></span>
<span id="cb104-8"><a href="#cb104-8" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb104-9"><a href="#cb104-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-10"><a href="#cb104-10" aria-hidden="true" tabindex="-1"></a>contexts1<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"                              
 [2] "taxon_name"                              
 [3] "observation_id"                          
 [4] "trait_name"                              
 [5] "value"                                   
 [6] "unit"                                    
 [7] "entity_type"                             
 [8] "value_type"                              
 [9] "basis_of_value"                          
[10] "replicates"                              
[11] "basis_of_record"                         
[12] "life_stage"                              
[13] "population_id"                           
[14] "individual_id"                           
[15] "repeat_measurements_id"                  
[16] "temporal_context_id"                     
[17] "source_id"                               
[18] "location_id"                             
[19] "entity_context_id"                       
[20] "plot_context_id"                         
[21] "treatment_context_id"                    
[22] "collection_date"                         
[23] "measurement_remarks"                     
[24] "method_id"                               
[25] "method_context_id"                       
[26] "original_name"                           
[27] "treatment_context: drought treatment"    
[28] "treatment_context: temperature treatment"
[29] "plot_context: fire intensity"            
[30] "plot_context: slope position"            
[31] "entity_context: plant sex"               
[32] "temporal_context: sampling season"       
[33] "temporal_context: fire season"           
[34] "method_context: leaf age"                
[35] "method_context: leaf surface"            </code></pre>
</div>
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>contexts1<span class="sc">$</span>traits[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,726,024 × 10
   dataset_id taxon_name       observation_id trait_name value unit  entity_type
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 ABRS_1981  Acanthocarpus c… 0001           leaf_comp… simp… &lt;NA&gt;  species    
 2 ABRS_1981  Acanthocarpus c… 0001           seed_leng… 3     mm    species    
 3 ABRS_1981  Acanthocarpus h… 0002           leaf_comp… simp… &lt;NA&gt;  species    
 4 ABRS_1981  Acanthocarpus p… 0003           leaf_comp… simp… &lt;NA&gt;  species    
 5 ABRS_1981  Acanthocarpus p… 0004           leaf_comp… simp… &lt;NA&gt;  species    
 6 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 4     mm    species    
 7 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 5     mm    species    
 8 ABRS_1981  Acanthocarpus r… 0005           leaf_comp… simp… &lt;NA&gt;  species    
 9 ABRS_1981  Acanthocarpus r… 0006           leaf_comp… simp… &lt;NA&gt;  species    
10 ABRS_1981  Acanthocarpus r… 0006           seed_leng… 2.5   mm    species    
# ℹ 1,726,014 more rows
# ℹ 3 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;, replicates &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co"># method where you first lookup context properties using the function `lookup_context_property`; this example showcases `format = "single_column_pretty"</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a>leaf_properties <span class="ot">&lt;-</span> <span class="fu">lookup_context_property</span>(<span class="at">database =</span> austraits, <span class="at">term =</span> <span class="st">"leaf"</span>)</span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a>contexts2 <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span></span>
<span id="cb108-5"><a href="#cb108-5" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_context_properties</span>(</span>
<span id="cb108-6"><a href="#cb108-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">vars =</span> leaf_properties,</span>
<span id="cb108-7"><a href="#cb108-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">format =</span> <span class="st">"single_column_pretty"</span>,</span>
<span id="cb108-8"><a href="#cb108-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">include_description =</span> <span class="cn">TRUE</span></span>
<span id="cb108-9"><a href="#cb108-9" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb108-10"><a href="#cb108-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb108-11"><a href="#cb108-11" aria-hidden="true" tabindex="-1"></a>contexts2<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"                   "taxon_name"                  
 [3] "observation_id"               "trait_name"                  
 [5] "value"                        "unit"                        
 [7] "entity_type"                  "value_type"                  
 [9] "basis_of_value"               "replicates"                  
[11] "basis_of_record"              "life_stage"                  
[13] "population_id"                "individual_id"               
[15] "repeat_measurements_id"       "temporal_context_id"         
[17] "source_id"                    "location_id"                 
[19] "entity_context_id"            "plot_context_id"             
[21] "treatment_context_id"         "collection_date"             
[23] "measurement_remarks"          "method_id"                   
[25] "method_context_id"            "original_name"               
[27] "treatment_context_properties" "plot_context_properties"     
[29] "entity_context_properties"    "temporal_context_properties" 
[31] "method_context_properties"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>contexts2<span class="sc">$</span>traits[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,726,024 × 10
   dataset_id taxon_name       observation_id trait_name value unit  entity_type
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 ABRS_1981  Acanthocarpus c… 0001           leaf_comp… simp… &lt;NA&gt;  species    
 2 ABRS_1981  Acanthocarpus c… 0001           seed_leng… 3     mm    species    
 3 ABRS_1981  Acanthocarpus h… 0002           leaf_comp… simp… &lt;NA&gt;  species    
 4 ABRS_1981  Acanthocarpus p… 0003           leaf_comp… simp… &lt;NA&gt;  species    
 5 ABRS_1981  Acanthocarpus p… 0004           leaf_comp… simp… &lt;NA&gt;  species    
 6 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 4     mm    species    
 7 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 5     mm    species    
 8 ABRS_1981  Acanthocarpus r… 0005           leaf_comp… simp… &lt;NA&gt;  species    
 9 ABRS_1981  Acanthocarpus r… 0006           leaf_comp… simp… &lt;NA&gt;  species    
10 ABRS_1981  Acanthocarpus r… 0006           seed_leng… 2.5   mm    species    
# ℹ 1,726,014 more rows
# ℹ 3 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;, replicates &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="co"># method where you add all context properties; this example showcases `format = "single_column_json"</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>contexts3 <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span></span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_context_properties</span>(</span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">vars =</span> <span class="st">"all"</span>,</span>
<span id="cb112-5"><a href="#cb112-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">format =</span> <span class="st">"single_column_json"</span>,</span>
<span id="cb112-6"><a href="#cb112-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">include_description =</span> <span class="cn">FALSE</span></span>
<span id="cb112-7"><a href="#cb112-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb112-8"><a href="#cb112-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-9"><a href="#cb112-9" aria-hidden="true" tabindex="-1"></a>contexts3<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"                   "taxon_name"                  
 [3] "observation_id"               "trait_name"                  
 [5] "value"                        "unit"                        
 [7] "entity_type"                  "value_type"                  
 [9] "basis_of_value"               "replicates"                  
[11] "basis_of_record"              "life_stage"                  
[13] "population_id"                "individual_id"               
[15] "repeat_measurements_id"       "temporal_context_id"         
[17] "source_id"                    "location_id"                 
[19] "entity_context_id"            "plot_context_id"             
[21] "treatment_context_id"         "collection_date"             
[23] "measurement_remarks"          "method_id"                   
[25] "method_context_id"            "original_name"               
[27] "treatment_context_properties" "plot_context_properties"     
[29] "entity_context_properties"    "temporal_context_properties" 
[31] "method_context_properties"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>contexts3<span class="sc">$</span>traits[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,726,024 × 10
   dataset_id taxon_name       observation_id trait_name value unit  entity_type
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 ABRS_1981  Acanthocarpus c… 0001           leaf_comp… simp… &lt;NA&gt;  species    
 2 ABRS_1981  Acanthocarpus c… 0001           seed_leng… 3     mm    species    
 3 ABRS_1981  Acanthocarpus h… 0002           leaf_comp… simp… &lt;NA&gt;  species    
 4 ABRS_1981  Acanthocarpus p… 0003           leaf_comp… simp… &lt;NA&gt;  species    
 5 ABRS_1981  Acanthocarpus p… 0004           leaf_comp… simp… &lt;NA&gt;  species    
 6 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 4     mm    species    
 7 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 5     mm    species    
 8 ABRS_1981  Acanthocarpus r… 0005           leaf_comp… simp… &lt;NA&gt;  species    
 9 ABRS_1981  Acanthocarpus r… 0006           leaf_comp… simp… &lt;NA&gt;  species    
10 ABRS_1981  Acanthocarpus r… 0006           seed_leng… 2.5   mm    species    
# ℹ 1,726,014 more rows
# ℹ 3 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;, replicates &lt;chr&gt;</code></pre>
</div>
</div>
</section>
<section id="joining-methods-columns" class="level4">
<h4 class="anchored" data-anchor-id="joining-methods-columns">Joining methods columns</h4>
<p>The methods table documents a selection of metadata recorded about the entire dataset and methods used for individual trait measurements. There is a single row of data per <code>dataset_id</code> x <code>trait_name</code> x <code>method_id</code> combination. <code>Method_id</code> is used to distinguish between instances where a single trait is measured twice using two separate protocols and is separate to <code>method_context_id</code>, which documents specific components of a method that are modified between measurements.</p>
<p>The austraits function <code>join_methods</code> joins columns from the methods table to the <code>traits</code> table. It has one argument:</p>
<ol type="1">
<li><code>vars</code> which specifies which columns from the <code>methods</code> table are joined to the <code>traits</code> table; defaults to <code>vars = c("all")</code></li>
</ol>
<p>First, check the schema file embedded within AusTraits to see what information is documented in each column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>schema<span class="sc">$</span>austraits<span class="sc">$</span>elements<span class="sc">$</span>methods<span class="sc">$</span>elements <span class="sc">%&gt;%</span> </span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">convert_list_to_df1</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 15 × 2
   key                              value                                       
   &lt;chr&gt;                            &lt;chr&gt;                                       
 1 dataset_id                       Primary identifier for each study contribut…
 2 trait_name                       Name of the trait sampled. Allowable values…
 3 methods                          A textual description of the methods used t…
 4 method_id                        A unique integer identifier to distinguish …
 5 description                      A 1-2 sentence description of the purpose o…
 6 sampling_strategy                A written description of how study location…
 7 source_primary_key               Citation key for the primary source in `sou…
 8 source_primary_citation          Citation for the primary source. This detai…
 9 source_secondary_key             Citation key for the secondary source in `s…
10 source_secondary_citation        Citations for the secondary source. This de…
11 source_original_dataset_key      Citation key for the original dataset_id in…
12 source_original_dataset_citation Citations for the original dataset_id in so…
13 data_collectors                  The person (people) leading data collection…
14 assistants                       Names of additional people who played a mor…
15 dataset_curators                 Names of AusTraits team member(s) who conta…</code></pre>
</div>
</div>
<p>Examples using <code>join_methods</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co"># join methods column only, the default</span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>traits_with_methods <span class="ot">&lt;-</span> </span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_methods</span>()</span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb118-5"><a href="#cb118-5" aria-hidden="true" tabindex="-1"></a>traits_with_methods<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"             "taxon_name"             "observation_id"        
 [4] "trait_name"             "value"                  "unit"                  
 [7] "entity_type"            "value_type"             "basis_of_value"        
[10] "replicates"             "basis_of_record"        "life_stage"            
[13] "population_id"          "individual_id"          "repeat_measurements_id"
[16] "temporal_context_id"    "source_id"              "location_id"           
[19] "entity_context_id"      "plot_context_id"        "treatment_context_id"  
[22] "collection_date"        "measurement_remarks"    "method_id"             
[25] "method_context_id"      "original_name"          "methods"               </code></pre>
</div>
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="co"># join all methods table columns</span></span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>traits_with_methods <span class="ot">&lt;-</span> </span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_methods</span>(<span class="at">vars =</span> <span class="st">"all"</span>)</span>
<span id="cb120-4"><a href="#cb120-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-5"><a href="#cb120-5" aria-hidden="true" tabindex="-1"></a>traits_with_methods<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"                       "taxon_name"                      
 [3] "observation_id"                   "trait_name"                      
 [5] "value"                            "unit"                            
 [7] "entity_type"                      "value_type"                      
 [9] "basis_of_value"                   "replicates"                      
[11] "basis_of_record"                  "life_stage"                      
[13] "population_id"                    "individual_id"                   
[15] "repeat_measurements_id"           "temporal_context_id"             
[17] "source_id"                        "location_id"                     
[19] "entity_context_id"                "plot_context_id"                 
[21] "treatment_context_id"             "collection_date"                 
[23] "measurement_remarks"              "method_id"                       
[25] "method_context_id"                "original_name"                   
[27] "methods"                          "description"                     
[29] "sampling_strategy"                "source_primary_key"              
[31] "source_primary_citation"          "source_secondary_key"            
[33] "source_secondary_citation"        "source_original_dataset_key"     
[35] "source_original_dataset_citation" "data_collectors"                 
[37] "assistants"                       "dataset_curators"                </code></pre>
</div>
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="co"># join all specifically selected methods table columns</span></span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a>traits_with_methods <span class="ot">&lt;-</span> </span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_methods</span>(<span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"methods"</span>, <span class="st">"description"</span>, <span class="st">"source_secondary_key"</span>))</span>
<span id="cb122-4"><a href="#cb122-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb122-5"><a href="#cb122-5" aria-hidden="true" tabindex="-1"></a>traits_with_methods<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"             "taxon_name"             "observation_id"        
 [4] "trait_name"             "value"                  "unit"                  
 [7] "entity_type"            "value_type"             "basis_of_value"        
[10] "replicates"             "basis_of_record"        "life_stage"            
[13] "population_id"          "individual_id"          "repeat_measurements_id"
[16] "temporal_context_id"    "source_id"              "location_id"           
[19] "entity_context_id"      "plot_context_id"        "treatment_context_id"  
[22] "collection_date"        "measurement_remarks"    "method_id"             
[25] "method_context_id"      "original_name"          "methods"               
[28] "description"            "source_secondary_key"  </code></pre>
</div>
</div>
</section>
<section id="joining-taxa" class="level4">
<h4 class="anchored" data-anchor-id="joining-taxa">Joining taxa</h4>
<p>The <code>taxa</code> table documents a collection of names and identifiers for each taxon. Within AusTraits, <code>names</code> submitted as identifiers within a dataset might be resolved to a species, an infraspecific taxon, or sometimes just to a genus- or family-level name; the name’s resolution is recorded as the <code>taxon_rank</code>. The <code>taxon_rank</code> determines which information is filled in in the taxa table.</p>
<p>The <code>{austraits}</code> function <code>join_taxa</code> joins columns from the <code>taxa</code> table to the <code>traits</code> table. It has one argument:</p>
<ol type="1">
<li><code>vars</code> which specifies which columns from the <code>taxa</code> table are joined to the <code>traits</code> table; defaults to <code>vars = c("family", "genus", "taxon_rank", "establishment_means").</code></li>
</ol>
<p>First, check the schema file embedded within AusTraits to see what information is documented in each column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>schema<span class="sc">$</span>austraits<span class="sc">$</span>elements<span class="sc">$</span>taxa<span class="sc">$</span>elements <span class="sc">%&gt;%</span> </span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">convert_list_to_df1</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 15 × 2
   key                 value                                                    
   &lt;chr&gt;               &lt;chr&gt;                                                    
 1 taxon_name          Scientific name of the taxon on which traits were sample…
 2 taxonomic_dataset   Name of the taxonomy (tree) that contains this concept. …
 3 taxon_rank          The taxonomic rank of the most specific name in the scie…
 4 trinomial           The infraspecific taxon name match for an original name.…
 5 binomial            The species-level taxon name match for an original name.…
 6 genus               Genus of the taxon without authorship.                   
 7 family              Family of the taxon.                                     
 8 taxon_distribution  Known distribution of the taxon, by Australian state.    
 9 establishment_means Statement about whether an organism or organisms have be…
10 taxonomic_status    The status of the use of the scientificName as a label f…
11 taxon_id            An identifier for the set of taxon information (data ass…
12 taxon_id_genus      An identifier for the set of taxon information (data ass…
13 taxon_id_family     An identifier for the set of taxon information (data ass…
14 scientific_name     The full scientific name, with authorship and date infor…
15 scientific_name_id  An identifier for the set of taxon information (data ass…</code></pre>
</div>
</div>
<p>Examples using <code>join_taxa</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="co"># join the default columns</span></span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a>traits_with_taxa <span class="ot">&lt;-</span> </span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_taxa</span>()</span>
<span id="cb126-4"><a href="#cb126-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb126-5"><a href="#cb126-5" aria-hidden="true" tabindex="-1"></a>traits_with_taxa<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"             "taxon_name"             "observation_id"        
 [4] "trait_name"             "value"                  "unit"                  
 [7] "entity_type"            "value_type"             "basis_of_value"        
[10] "replicates"             "basis_of_record"        "life_stage"            
[13] "population_id"          "individual_id"          "repeat_measurements_id"
[16] "temporal_context_id"    "source_id"              "location_id"           
[19] "entity_context_id"      "plot_context_id"        "treatment_context_id"  
[22] "collection_date"        "measurement_remarks"    "method_id"             
[25] "method_context_id"      "original_name"          "family"                
[28] "genus"                  "taxon_rank"             "establishment_means"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="co"># join all taxa table columns</span></span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>traits_with_taxa <span class="ot">&lt;-</span> </span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_taxa</span>(<span class="at">vars =</span> <span class="st">"all"</span>)</span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb128-5"><a href="#cb128-5" aria-hidden="true" tabindex="-1"></a>traits_with_taxa<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"              "taxon_name"             
 [3] "observation_id"          "trait_name"             
 [5] "value"                   "unit"                   
 [7] "entity_type"             "value_type"             
 [9] "basis_of_value"          "replicates"             
[11] "basis_of_record"         "life_stage"             
[13] "population_id"           "individual_id"          
[15] "repeat_measurements_id"  "temporal_context_id"    
[17] "source_id"               "location_id"            
[19] "entity_context_id"       "plot_context_id"        
[21] "treatment_context_id"    "collection_date"        
[23] "measurement_remarks"     "method_id"              
[25] "method_context_id"       "original_name"          
[27] "taxon_rank"              "taxonomic_status"       
[29] "taxonomic_dataset"       "taxon_name_alternatives"
[31] "genus"                   "family"                 
[33] "binomial"                "trinomial"              
[35] "taxon_distribution"      "establishment_means"    
[37] "scientific_name"         "taxon_id"               
[39] "taxon_id_genus"          "taxon_id_family"        
[41] "scientific_name_id"     </code></pre>
</div>
</div>
</section>
<section id="joining-taxonomic-updates" class="level4">
<h4 class="anchored" data-anchor-id="joining-taxonomic-updates">Joining taxonomic updates</h4>
<p>The taxonomic updates table documents all taxonomic changes implemented in the construction of AusTraits, including both the correction of typos and the updating of outdated synonyms to the currently accepted name.</p>
<p>The <code>{austraits}</code> function <code>join_taxonomic_updates</code> joins columns from the <code>taxonomic_updates</code> table to the <code>traits</code> table. It has one argument:</p>
<ol type="1">
<li><code>vars</code> which specifies which columns from the <code>taxonomic_updates</code> table are joined to the <code>traits</code> table; defaults to <code>vars = c("aligned_name").</code></li>
</ol>
<p>First, check the schema file embedded within AusTraits to see what information is documented in each column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>schema<span class="sc">$</span>austraits<span class="sc">$</span>elements<span class="sc">$</span>taxonomic_updates<span class="sc">$</span>elements <span class="sc">%&gt;%</span> </span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">convert_list_to_df1</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 7 × 2
  key                           value                                           
  &lt;chr&gt;                         &lt;chr&gt;                                           
1 dataset_id                    Primary identifier for each study contributed t…
2 original_name                 Name given to taxon in the original data suppli…
3 aligned_name                  The taxon name without authorship after impleme…
4 taxonomic_resolution          The rank of the most specific taxon name (or sc…
5 taxon_name                    Scientific name of the taxon on which traits we…
6 aligned_name_taxon_id         An identifier for the aligned name before it is…
7 aligned_name_taxonomic_status The status of the use of the `aligned_name` as …</code></pre>
</div>
</div>
<p>Examples using <code>join_taxonomic_updates</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="co"># join the default columns</span></span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>traits_with_taxonomic_updates <span class="ot">&lt;-</span> </span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_taxonomic_updates</span>()</span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a>traits_with_taxonomic_updates<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"             "taxon_name"             "observation_id"        
 [4] "trait_name"             "value"                  "unit"                  
 [7] "entity_type"            "value_type"             "basis_of_value"        
[10] "replicates"             "basis_of_record"        "life_stage"            
[13] "population_id"          "individual_id"          "repeat_measurements_id"
[16] "temporal_context_id"    "source_id"              "location_id"           
[19] "entity_context_id"      "plot_context_id"        "treatment_context_id"  
[22] "collection_date"        "measurement_remarks"    "method_id"             
[25] "method_context_id"      "original_name"          "aligned_name"          </code></pre>
</div>
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="co"># join all methods columns</span></span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a>traits_with_taxonomic_updates <span class="ot">&lt;-</span> </span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_taxonomic_updates</span>(<span class="at">vars =</span> <span class="st">"all"</span>)</span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-5"><a href="#cb134-5" aria-hidden="true" tabindex="-1"></a>traits_with_taxonomic_updates<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"                    "taxon_name"                   
 [3] "observation_id"                "trait_name"                   
 [5] "value"                         "unit"                         
 [7] "entity_type"                   "value_type"                   
 [9] "basis_of_value"                "replicates"                   
[11] "basis_of_record"               "life_stage"                   
[13] "population_id"                 "individual_id"                
[15] "repeat_measurements_id"        "temporal_context_id"          
[17] "source_id"                     "location_id"                  
[19] "entity_context_id"             "plot_context_id"              
[21] "treatment_context_id"          "collection_date"              
[23] "measurement_remarks"           "method_id"                    
[25] "method_context_id"             "original_name"                
[27] "aligned_name"                  "taxonomic_resolution"         
[29] "aligned_name_taxon_id"         "aligned_name_taxonomic_status"</code></pre>
</div>
</div>
</section>
<section id="joining-contributors" class="level4">
<h4 class="anchored" data-anchor-id="joining-contributors">Joining contributors</h4>
<p>The contributors table documents all basic metadata about all dataset contributors, including their name, ORCID, and role for various datasets.</p>
<p>The <code>{austraits}</code> function <code>join_contributors</code> joins columns from the <code>contributors</code> table to the <code>traits</code> table. It has two arguments:</p>
<ol type="1">
<li><p><code>vars</code> which specifies which columns from the <code>contributors</code> table are joined to the <code>traits</code> table; defaults to `vars = c(“aligned_name”).</p></li>
<li><p><code>format</code> offers two output formats:</p></li>
</ol>
<pre><code>*   `single_column_pretty` (data in selected columns from `contributor` table compacted into a single column delimited in a way that is easy for humans to read; this is the default)
*   `single_column_json` (data in selected columns from `contributor` table compacted into a single column, using json formatting)  </code></pre>
<p>First, check the schema file embedded within AusTraits to see what information is documented in each column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>schema<span class="sc">$</span>austraits<span class="sc">$</span>elements<span class="sc">$</span>contributors<span class="sc">$</span>elements <span class="sc">%&gt;%</span> </span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">convert_list_to_df1</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  key             value                                                         
  &lt;chr&gt;           &lt;chr&gt;                                                         
1 dataset_id      Primary identifier for each study contributed to AusTraits; m…
2 last_name       Last name of the data collector.                              
3 given_name      Given names of the data collector.                            
4 ORCID           ORCID of the data collector.                                  
5 affiliation     Last known institution or affiliation.                        
6 additional_role Additional roles of data collector, mostly contact person.    </code></pre>
</div>
</div>
<p>Examples using <code>join_contributors</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="co"># join all columns (the default)</span></span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>traits_with_contributors <span class="ot">&lt;-</span> </span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_contributors</span>(<span class="at">format =</span> <span class="st">"single_column_json"</span>)</span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a>traits_with_contributors<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"             "taxon_name"             "observation_id"        
 [4] "trait_name"             "value"                  "unit"                  
 [7] "entity_type"            "value_type"             "basis_of_value"        
[10] "replicates"             "basis_of_record"        "life_stage"            
[13] "population_id"          "individual_id"          "repeat_measurements_id"
[16] "temporal_context_id"    "source_id"              "location_id"           
[19] "entity_context_id"      "plot_context_id"        "treatment_context_id"  
[22] "collection_date"        "measurement_remarks"    "method_id"             
[25] "method_context_id"      "original_name"          "data_contributors"     </code></pre>
</div>
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="co"># join select contributors columns</span></span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a>traits_with_contributors <span class="ot">&lt;-</span> </span>
<span id="cb141-3"><a href="#cb141-3" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_contributors</span>(</span>
<span id="cb141-4"><a href="#cb141-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"last_name"</span>, <span class="st">"first_name"</span>, <span class="st">"ORCID"</span>),</span>
<span id="cb141-5"><a href="#cb141-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">format =</span> <span class="st">"single_column_pretty"</span>)</span>
<span id="cb141-6"><a href="#cb141-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-7"><a href="#cb141-7" aria-hidden="true" tabindex="-1"></a>traits_with_contributors<span class="sc">$</span>traits <span class="sc">%&gt;%</span> <span class="fu">names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"             "taxon_name"             "observation_id"        
 [4] "trait_name"             "value"                  "unit"                  
 [7] "entity_type"            "value_type"             "basis_of_value"        
[10] "replicates"             "basis_of_record"        "life_stage"            
[13] "population_id"          "individual_id"          "repeat_measurements_id"
[16] "temporal_context_id"    "source_id"              "location_id"           
[19] "entity_context_id"      "plot_context_id"        "treatment_context_id"  
[22] "collection_date"        "measurement_remarks"    "method_id"             
[25] "method_context_id"      "original_name"          "data_contributors"     </code></pre>
</div>
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a>traits_with_contributors<span class="sc">$</span>traits</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,726,024 × 27
   dataset_id taxon_name       observation_id trait_name value unit  entity_type
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 ABRS_1981  Acanthocarpus c… 0001           leaf_comp… simp… &lt;NA&gt;  species    
 2 ABRS_1981  Acanthocarpus c… 0001           seed_leng… 3     mm    species    
 3 ABRS_1981  Acanthocarpus h… 0002           leaf_comp… simp… &lt;NA&gt;  species    
 4 ABRS_1981  Acanthocarpus p… 0003           leaf_comp… simp… &lt;NA&gt;  species    
 5 ABRS_1981  Acanthocarpus p… 0004           leaf_comp… simp… &lt;NA&gt;  species    
 6 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 4     mm    species    
 7 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 5     mm    species    
 8 ABRS_1981  Acanthocarpus r… 0005           leaf_comp… simp… &lt;NA&gt;  species    
 9 ABRS_1981  Acanthocarpus r… 0006           leaf_comp… simp… &lt;NA&gt;  species    
10 ABRS_1981  Acanthocarpus r… 0006           seed_leng… 2.5   mm    species    
# ℹ 1,726,014 more rows
# ℹ 20 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;,
#   replicates &lt;chr&gt;, basis_of_record &lt;chr&gt;, life_stage &lt;chr&gt;,
#   population_id &lt;chr&gt;, individual_id &lt;chr&gt;, repeat_measurements_id &lt;chr&gt;,
#   temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;, location_id &lt;chr&gt;,
#   entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;, treatment_context_id &lt;chr&gt;,
#   collection_date &lt;chr&gt;, measurement_remarks &lt;chr&gt;, method_id &lt;chr&gt;, …</code></pre>
</div>
</div>
</section>
<section id="joining-all-data" class="level4">
<h4 class="anchored" data-anchor-id="joining-all-data">Joining all data</h4>
<p>If you want to join data from all ancillary tables onto the traits table, effectively “flattening” the relational table into a flat table, it is simplest to use the <code>{austraits}</code> function <code>flatten_database</code>.</p>
<p><code>flatten_database</code> calls each of the join functions, selecting <code>vars = "all"</code> as the default for each function.</p>
<p>It has three arguments:</p>
<ol type="1">
<li><p><code>vars</code>, specifies the context properties, via complete or partial string matches, that should be added to the traits table; defaults to “all”</p></li>
<li><p><code>format</code> offers three output formats that apply to the functions <code>join_locations</code>, <code>join_contexts</code> and <code>join_contributors</code></p></li>
</ol>
<pre><code>*   `many_columns` (each location or context property is added as a separate column)
*   `single_column_pretty` (all location or context properties or all contributor columns compacted into a single column delimited in a way that is easy for humans to read; this is the default)
*   `single_column_json` (all location or context properties or all contributor columns compacted into a single column, using json formatting)</code></pre>
<ol start="3" type="1">
<li><code>include_description</code> is a logical argument (TRUE/FALSE) that indicates where the context property value descriptions should be included or excluded when context data are joined; defaults to “TRUE”; this argument is only used to parameterise <code>join_contexts</code></li>
</ol>
<p>Examples using <code>flatten_database</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="co"># using the defaults</span></span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a>flat_database <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span> <span class="fu">flatten_database</span>()</span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb146-4"><a href="#cb146-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(flat_database)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"                       "taxon_name"                      
 [3] "observation_id"                   "trait_name"                      
 [5] "value"                            "unit"                            
 [7] "entity_type"                      "value_type"                      
 [9] "basis_of_value"                   "replicates"                      
[11] "basis_of_record"                  "life_stage"                      
[13] "population_id"                    "individual_id"                   
[15] "repeat_measurements_id"           "temporal_context_id"             
[17] "source_id"                        "location_id"                     
[19] "entity_context_id"                "plot_context_id"                 
[21] "treatment_context_id"             "collection_date"                 
[23] "measurement_remarks"              "method_id"                       
[25] "method_context_id"                "original_name"                   
[27] "location_name"                    "latitude (deg)"                  
[29] "longitude (deg)"                  "location_properties"             
[31] "treatment_context_properties"     "plot_context_properties"         
[33] "entity_context_properties"        "temporal_context_properties"     
[35] "method_context_properties"        "methods"                         
[37] "description"                      "sampling_strategy"               
[39] "source_primary_key"               "source_primary_citation"         
[41] "source_secondary_key"             "source_secondary_citation"       
[43] "source_original_dataset_key"      "source_original_dataset_citation"
[45] "assistants"                       "dataset_curators"                
[47] "data_contributors"                "taxon_rank"                      
[49] "taxonomic_status"                 "taxonomic_dataset"               
[51] "taxon_name_alternatives"          "genus"                           
[53] "family"                           "binomial"                        
[55] "trinomial"                        "taxon_distribution"              
[57] "establishment_means"              "scientific_name"                 
[59] "taxon_id"                         "taxon_id_genus"                  
[61] "taxon_id_family"                  "scientific_name_id"              
[63] "aligned_name"                     "taxonomic_resolution"            
[65] "aligned_name_taxon_id"            "aligned_name_taxonomic_status"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="co"># specifying vars for each column</span></span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a>flat_database <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span> <span class="fu">flatten_database</span>(</span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">vars =</span> <span class="fu">list</span>(</span>
<span id="cb148-4"><a href="#cb148-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">location =</span> <span class="st">"all"</span>,</span>
<span id="cb148-5"><a href="#cb148-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">context =</span> <span class="st">"sampling_season"</span>,</span>
<span id="cb148-6"><a href="#cb148-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">contributors =</span> <span class="fu">c</span>(<span class="st">"last_name"</span>, <span class="st">"first_name"</span>, <span class="st">"ORCID"</span>),</span>
<span id="cb148-7"><a href="#cb148-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">taxonomy =</span> <span class="fu">c</span>(<span class="st">"family"</span>, <span class="st">"establishment_means"</span>),</span>
<span id="cb148-8"><a href="#cb148-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">taxonomic_updates =</span> <span class="st">"aligned_name"</span>,</span>
<span id="cb148-9"><a href="#cb148-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">methods =</span> <span class="st">"methods"</span></span>
<span id="cb148-10"><a href="#cb148-10" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb148-11"><a href="#cb148-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb148-12"><a href="#cb148-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb148-13"><a href="#cb148-13" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(flat_database)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "dataset_id"                   "taxon_name"                  
 [3] "observation_id"               "trait_name"                  
 [5] "value"                        "unit"                        
 [7] "entity_type"                  "value_type"                  
 [9] "basis_of_value"               "replicates"                  
[11] "basis_of_record"              "life_stage"                  
[13] "population_id"                "individual_id"               
[15] "repeat_measurements_id"       "temporal_context_id"         
[17] "source_id"                    "location_id"                 
[19] "entity_context_id"            "plot_context_id"             
[21] "treatment_context_id"         "collection_date"             
[23] "measurement_remarks"          "method_id"                   
[25] "method_context_id"            "original_name"               
[27] "location_name"                "latitude (deg)"              
[29] "longitude (deg)"              "location_properties"         
[31] "treatment_context_properties" "plot_context_properties"     
[33] "entity_context_properties"    "temporal_context_properties" 
[35] "method_context_properties"    "methods"                     
[37] "data_contributors"            "family"                      
[39] "establishment_means"          "aligned_name"                </code></pre>
</div>
</div>
</section>
</section>
<section id="combining-extract_-and-join_-functions" class="level3">
<h3 class="anchored" data-anchor-id="combining-extract_-and-join_-functions">Combining <code>extract_</code> and <code>join_</code> functions</h3>
<p>As both the <code>extract</code> and <code>join</code> functions output a database with the original database structure they can be used sequentially to extract, then join exactly the data desired.</p>
<p>For instance:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a>subset2 <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span></span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_data</span>(<span class="at">table =</span> <span class="st">"traits"</span>, <span class="at">col =</span> <span class="st">"basis_of_record"</span>, <span class="at">col_value =</span> <span class="st">"field"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_name =</span> <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>, <span class="st">"leaf_thickness"</span>, <span class="st">"leaf_length"</span>, <span class="st">"leaf_area"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb150-4"><a href="#cb150-4" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">genus =</span> <span class="st">"Eucalyptus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb150-5"><a href="#cb150-5" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_location_coordinates</span>() <span class="sc">%&gt;%</span></span>
<span id="cb150-6"><a href="#cb150-6" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_taxa</span>(<span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"family"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb150-7"><a href="#cb150-7" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_context_properties</span>(<span class="at">vars =</span> <span class="st">"all"</span>, <span class="at">format =</span> <span class="st">"many_columns"</span>)</span>
<span id="cb150-8"><a href="#cb150-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb150-9"><a href="#cb150-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(subset2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "traits"            "locations"         "contexts"         
 [4] "methods"           "excluded_data"     "taxonomic_updates"
 [7] "taxa"              "contributors"      "sources"          
[10] "definitions"       "schema"            "metadata"         
[13] "build_info"       </code></pre>
</div>
</div>
<p>Having joined all context properties are separate columns, you may now look at the expanded <code>traits</code> table and decide that you only want data that was sampled during <code>wet seasons</code>, documented in the column <code>temporal_context: sampling season</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(subset2<span class="sc">$</span>traits<span class="sc">$</span><span class="st">`</span><span class="at">temporal_context: sampling season</span><span class="st">`</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] NA                                                                                                            
 [2] "summer favourable &lt;&lt;Measurements made during summer in a location where this is the favourable season.&gt;&gt;"    
 [3] "winter unfavourable &lt;&lt;Measurements made during winter in a location where this is the unfavourable season.&gt;&gt;"
 [4] "dry &lt;&lt;Measurements made during the dry season.&gt;&gt;"                                                            
 [5] "summer unfavourable &lt;&lt;Measurements made during summer in a location where this is the unfavourable season.&gt;&gt;"
 [6] "winter favourable &lt;&lt;Measurements made during winter in a location where this is the favourable season.&gt;&gt;"    
 [7] "October"                                                                                                     
 [8] "wet &lt;&lt;Measurements made during the wet season.&gt;&gt;"                                                            
 [9] "dry season &lt;&lt;Measurements made during the dry season.&gt;&gt;"                                                     
[10] "wet season &lt;&lt;Measurements made during the wet season.&gt;&gt;"                                                     
[11] "dry season (Sep 2010) &lt;&lt;Measurements made during the dry season (Sep 2010).&gt;&gt;"                               
[12] "wet season (May 2011) &lt;&lt;Measurements made during the wet season (May 2011).&gt;&gt;"                               </code></pre>
</div>
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a>subset2 <span class="ot">&lt;-</span> subset2 <span class="sc">%&gt;%</span></span>
<span id="cb154-2"><a href="#cb154-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_data</span>(<span class="at">table =</span> <span class="st">"traits"</span>, <span class="at">col =</span> <span class="st">"temporal_context: sampling season"</span>, <span class="at">col_value =</span> <span class="st">"wet"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="binding-datasets" class="level3">
<h3 class="anchored" data-anchor-id="binding-datasets">Binding datasets</h3>
<p>For some applications, you may wish to extract two different subsets of data, based on the values of different columns, then merge those extracted database subsets together, but still retain the original database structure.</p>
<p>This is possible with the function <code>bind_databases</code>.</p>
<p>This function binds each of the relational tables, removing any duplicate entries.</p>
<p>For instance, you might want all measurements where <em>either</em> the <code>location_property</code> or the <code>context_property</code> references the word “fire”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a>subset_a <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span></span>
<span id="cb155-2"><a href="#cb155-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_data</span>(<span class="at">table =</span> <span class="st">"locations"</span>, <span class="at">col =</span> <span class="st">"location_property"</span>, <span class="at">col_value =</span> <span class="st">"fire"</span>)</span>
<span id="cb155-3"><a href="#cb155-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb155-4"><a href="#cb155-4" aria-hidden="true" tabindex="-1"></a>subset_b <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span></span>
<span id="cb155-5"><a href="#cb155-5" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_data</span>(<span class="at">table =</span> <span class="st">"contexts"</span>, <span class="at">col =</span> <span class="st">"context_property"</span>, <span class="at">col_value =</span> <span class="st">"fire"</span>)</span>
<span id="cb155-6"><a href="#cb155-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb155-7"><a href="#cb155-7" aria-hidden="true" tabindex="-1"></a>subset_ab <span class="ot">&lt;-</span> <span class="fu">bind_databases</span>(subset_a, subset_b)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="summarising-data-trait-means-modes-etc." class="level2" data-number="30.7">
<h2 data-number="30.7" class="anchored" data-anchor-id="summarising-data-trait-means-modes-etc."><span class="header-section-number">30.7</span> Summarising data: trait means, modes, etc.</h2>
<p>The function <code>summarise_trait_means</code> that was in older <code>{austraits}</code> versions was has been deprecated, as it is not appropriate for AusTraits versions &gt; 5.0 – that is all databases built using <code>{traits.build}</code>. A new version is in development and will be released in 2025. In the meantime, if you’ve sourced the file <code>extra_functions.R</code>, there are a few functions that allow you to summarise trait values.</p>
<section id="categorical-traits" class="level3">
<h3 class="anchored" data-anchor-id="categorical-traits">Categorical traits</h3>
<p>For instance, <code>categorical_summary</code> indicates how many times a specific trait value is reported for a given taxa (across all datasets):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a>cat_summary <span class="ot">&lt;-</span> <span class="fu">categorical_summary</span>(austraits, <span class="st">"resprouting_capacity"</span>)</span>
<span id="cb156-2"><a href="#cb156-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb156-3"><a href="#cb156-3" aria-hidden="true" tabindex="-1"></a>cat_summary</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12,817 × 3
   taxon_name                             trait_name           value_summary  
   &lt;chr&gt;                                  &lt;chr&gt;                &lt;chr&gt;          
 1 Abelmoschus ficulneus                  resprouting_capacity fire_killed (4)
 2 Abelmoschus manihot                    resprouting_capacity fire_killed (3)
 3 Abelmoschus moschatus                  resprouting_capacity resprouts (1)  
 4 Abelmoschus moschatus subsp. tuberosus resprouting_capacity resprouts (3)  
 5 Abrodictyum caudatum                   resprouting_capacity fire_killed (2)
 6 Abrodictyum obscurum                   resprouting_capacity fire_killed (2)
 7 Abrotanella forsteroides               resprouting_capacity resprouts (1)  
 8 Abrotanella nivigena                   resprouting_capacity resprouts (1)  
 9 Abrotanella sp. [White_2020]           resprouting_capacity resprouts (1)  
10 Abrus precatorius                      resprouting_capacity resprouts (1)  
# ℹ 12,807 more rows</code></pre>
</div>
</div>
<p>Alternatively, create a wider matrix with possible trait values as columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb158"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a>categorical_summary_wider <span class="ot">&lt;-</span> </span>
<span id="cb158-2"><a href="#cb158-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">categorical_summary_by_value</span>(austraits, <span class="st">"resprouting_capacity"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb158-3"><a href="#cb158-3" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> value_tmp, <span class="at">values_from =</span> replicates)</span>
<span id="cb158-4"><a href="#cb158-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb158-5"><a href="#cb158-5" aria-hidden="true" tabindex="-1"></a>categorical_summary_wider</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12,817 × 5
   taxon_name               trait_name fire_killed resprouts partial_resprouting
   &lt;chr&gt;                    &lt;chr&gt;            &lt;dbl&gt;     &lt;dbl&gt;               &lt;dbl&gt;
 1 Abelmoschus ficulneus    resprouti…           4        NA                  NA
 2 Abelmoschus manihot      resprouti…           3        NA                  NA
 3 Abelmoschus moschatus    resprouti…          NA         1                  NA
 4 Abelmoschus moschatus s… resprouti…          NA         3                  NA
 5 Abrodictyum caudatum     resprouti…           2        NA                  NA
 6 Abrodictyum obscurum     resprouti…           2        NA                  NA
 7 Abrotanella forsteroides resprouti…          NA         1                  NA
 8 Abrotanella nivigena     resprouti…          NA         1                  NA
 9 Abrotanella sp. [White_… resprouti…          NA         1                  NA
10 Abrus precatorius        resprouti…          NA         1                  NA
# ℹ 12,807 more rows</code></pre>
</div>
</div>
</section>
<section id="numeric-traits" class="level3">
<h3 class="anchored" data-anchor-id="numeric-traits">Numeric traits</h3>
<p>One of the problems with writing functions that summarise numeric traits is that they make statistical assumptions that are hidden within the function code and might not be appropriate for your data use case.</p>
<p>The datasets that comprise AusTraits were collected by different people, with a different number of replicates and different entity types reported. One dataset might include 20 measurements on individuals for a trait and another might have submitted a single population-level mean derived from 5 measurements.</p>
<p>How do you take the mean of these trait values?</p>
<p>Do you want to include both data from experiments and plants growing under natural conditions? This information is recorded in the <code>basis_of_record</code> column.</p>
<p>One function we’re developing calculates weighted group means for field and experiment-sourced data, by first grouping values at the site level, then at the taxon level. For trait data sourced from floras where trait values are documented as a minimum and maximum value, the function takes the mean of these. The two subsets of data are then merged together.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a>weighted <span class="ot">&lt;-</span> <span class="fu">austraits_weighted_means</span>(austraits, <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>, </span>
<span id="cb160-2"><a href="#cb160-2" aria-hidden="true" tabindex="-1"></a>                                                  <span class="st">"leaf_length"</span>))</span>
<span id="cb160-3"><a href="#cb160-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-4"><a href="#cb160-4" aria-hidden="true" tabindex="-1"></a>weighted</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 24,995 × 10
   taxon_name       trait_name  mean   min   max median geom_mean all_replicates
   &lt;chr&gt;            &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;     &lt;dbl&gt;          &lt;dbl&gt;
 1 Abelia x grandi… leaf_leng…   35     20    50     NA      35                1
 2 Abelmoschus fic… leaf_leng…  105     50   160     NA     105                1
 3 Abelmoschus man… leaf_leng…  232.   150   380     NA     230.               2
 4 Abelmoschus man… leaf_leng…  200    200   200     NA     200                1
 5 Abelmoschus mos… leaf_leng…  103.    27   150     NA      97.6              3
 6 Abelmoschus mos… leaf_leng…  150    150   150     NA     150                1
 7 Abildgaardia ov… leaf_leng…  230     60   400     NA     230                1
 8 Abrodictyum bra… leaf_leng…  143.    40   250     NA     137.               2
 9 Abrodictyum cau… leaf_leng…  134.    35   200     NA     129.               4
10 Abrodictyum obs… leaf_leng…  190     20   400     NA     161.               5
# ℹ 24,985 more rows
# ℹ 2 more variables: location_replicates &lt;dbl&gt;, flora_replicates &lt;dbl&gt;</code></pre>
</div>
</div>
<p>This function may be sufficient for exploratory purposes. Alternatively, you can download the file with the function and edit the code to suit your purposes.</p>
</section>
</section>
<section id="plotting_functions" class="level2" data-number="30.8">
<h2 data-number="30.8" class="anchored" data-anchor-id="plotting_functions"><span class="header-section-number">30.8</span> Plotting data</h2>
<section id="plotting-trait-distributions" class="level3">
<h3 class="anchored" data-anchor-id="plotting-trait-distributions">Plotting trait distributions</h3>
<p>Another way to summarise AusTraits data by trait, and determine whether AusTraits offers sufficient data coverage for a trait of choice, is to plot the distribution of trait values in AusTraits.</p>
<p>As seen in <a href="#exploring"><code>A first look at data</code></a>, the function <code>austraits::plot_trait_distribution_beeswarm()</code> plots trait data by <code>dataset_id</code>, <code>genus</code>, <code>family</code> or indeed any column in the traits table, such as <code>life_stage</code> or <code>basis_of_record</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="co"># How does leaf N vary by dataset?</span></span>
<span id="cb162-2"><a href="#cb162-2" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_trait_distribution_beeswarm</span>(austraits, <span class="st">"leaf_N_per_dry_mass"</span>, </span>
<span id="cb162-3"><a href="#cb162-3" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">y_axis_category =</span> <span class="st">"dataset_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="co"># How does leaf N vary across Banksia species?</span></span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a>Banksia_data <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span> <span class="fu">extract_taxa</span>(<span class="at">genus =</span> <span class="st">"Banksia"</span>)</span>
<span id="cb163-3"><a href="#cb163-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb163-4"><a href="#cb163-4" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_trait_distribution_beeswarm</span>(Banksia_data, <span class="st">"leaf_N_per_dry_mass"</span>,</span>
<span id="cb163-5"><a href="#cb163-5" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">y_axis_category =</span> <span class="st">"taxon_name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-38-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Does leaf mass per area shift in Eucalyptus seedlings versus adults, which is captured in `life_stage`? What about amongst Eucalypts where information about the age of the leaves was recorded, captured as the context property "leaf age"?</span></span>
<span id="cb164-2"><a href="#cb164-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb164-3"><a href="#cb164-3" aria-hidden="true" tabindex="-1"></a>Euc_data <span class="ot">&lt;-</span> austraits <span class="sc">%&gt;%</span> <span class="fu">extract_taxa</span>(<span class="at">genus =</span> <span class="st">"Eucalyptus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb164-4"><a href="#cb164-4" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_context_properties</span>(<span class="at">vars =</span> <span class="st">"leaf age"</span>, <span class="at">format =</span> <span class="st">"many_columns"</span>, <span class="at">include_description =</span> <span class="cn">FALSE</span>)</span>
<span id="cb164-5"><a href="#cb164-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb164-6"><a href="#cb164-6" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_trait_distribution_beeswarm</span>(Euc_data, <span class="st">"leaf_mass_per_area"</span>, </span>
<span id="cb164-7"><a href="#cb164-7" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">y_axis_category =</span> <span class="st">"life_stage"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-38-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb165"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_trait_distribution_beeswarm</span>(Euc_data, <span class="st">"leaf_mass_per_area"</span>, </span>
<span id="cb165-2"><a href="#cb165-2" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">y_axis_category =</span> <span class="st">"method_context: leaf age"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-38-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="plotting-data-distribution-by-location" class="level3">
<h3 class="anchored" data-anchor-id="plotting-data-distribution-by-location">Plotting data distribution by location</h3>
<p>To plot locations, begin by merging on the latitude &amp; longitude data from austraits$locations using <code>austraits::join_location_coordinates</code>.</p>
<p>The <code>plot_locations</code> function plots the selected data, separating data into a series of plots based on the variable name selected. You can separate data based on the values of <strong>any</strong> column within the traits table – including <code>basis_of_record</code>, <code>life_stage</code> and <code>value_type</code> – or higher taxon categories (<code>genus</code>, <code>family</code>).</p>
<p>For instance, <code>austraits::plot_locations("trait_name")</code> will output a separate plot for each <code>trait_name</code> within the selected data.</p>
<p>A warning: <code>austraits::plot_locations()</code> WILL BE VERY SLOW if you request more than ~20 plots. For instance, do not attempt to generate plots for all traits simultaneously. Always first use extract/filter to just select a narrow range of traits, datasets, or taxa.</p>
<section id="plot-locations-by-trait-dataset-or-other-column" class="level4">
<h4 class="anchored" data-anchor-id="plot-locations-by-trait-dataset-or-other-column">Plot locations by trait, dataset, or other column</h4>
<p>See where Eucalyptus data have been collected, divided by <code>life_stage</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a>Euc_data <span class="ot">&lt;-</span> Euc_data <span class="sc">%&gt;%</span></span>
<span id="cb166-2"><a href="#cb166-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_location_coordinates</span>()</span>
<span id="cb166-3"><a href="#cb166-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb166-4"><a href="#cb166-4" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="at">database =</span> Euc_data, <span class="at">feature =</span> <span class="st">"life_stage"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>See where Banksia leaf area data have been collected, divided by <code>taxon_name</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a>Banksia_data <span class="ot">&lt;-</span> Banksia_data <span class="sc">%&gt;%</span></span>
<span id="cb167-2"><a href="#cb167-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_location_coordinates</span>() <span class="sc">%&gt;%</span></span>
<span id="cb167-3"><a href="#cb167-3" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_name =</span> <span class="st">"leaf_area"</span>)</span>
<span id="cb167-4"><a href="#cb167-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb167-5"><a href="#cb167-5" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="at">database =</span> Banksia_data, <span class="at">feature =</span> <span class="st">"taxon_name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Where were the various Westoby datasets collected?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb168"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb168-1"><a href="#cb168-1" aria-hidden="true" tabindex="-1"></a>Westoby <span class="ot">&lt;-</span></span>
<span id="cb168-2"><a href="#cb168-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb168-3"><a href="#cb168-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="at">dataset_id =</span> <span class="st">"Westoby"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb168-4"><a href="#cb168-4" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_location_coordinates</span>()</span>
<span id="cb168-5"><a href="#cb168-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb168-6"><a href="#cb168-6" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="at">database =</span> Westoby, <span class="at">feature =</span> <span class="st">"dataset_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that while the `dataset` is intended to be a relational database, this function also works with just the traits table, should you have separated it out of the relational structure.</span></span>
<span id="cb169-2"><a href="#cb169-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb169-3"><a href="#cb169-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Westoby_traits &lt;- Westoby$traits</span></span>
<span id="cb169-4"><a href="#cb169-4" aria-hidden="true" tabindex="-1"></a><span class="co"># austraits::plot_locations(dataset = Westoby_traits, feature = "dataset_id")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Where were data for Acacia aneura collected?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb170"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb170-1"><a href="#cb170-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb170-2"><a href="#cb170-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb170-3"><a href="#cb170-3" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">taxon_name =</span> <span class="st">"Acacia aneura"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb170-4"><a href="#cb170-4" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_location_coordinates</span>()</span>
<span id="cb170-5"><a href="#cb170-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-6"><a href="#cb170-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="ot">&lt;-</span> data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb170-7"><a href="#cb170-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(<span class="st">`</span><span class="at">latitude (deg)</span><span class="st">`</span>)) </span>
<span id="cb170-8"><a href="#cb170-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-9"><a href="#cb170-9" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_locations</span>(data, <span class="st">"taxon_name"</span>)     <span class="co"># actually 4 taxa, because of subspecies</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_locations</span>(data, <span class="st">"dataset_id"</span>)     <span class="co"># 1 plot for each dataset_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-42-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="more-complex-workflows-some-examples" class="level3">
<h3 class="anchored" data-anchor-id="more-complex-workflows-some-examples">More complex workflows – some examples</h3>
<section id="an-example-looking-at-trait-climate-gradients" class="level4">
<h4 class="anchored" data-anchor-id="an-example-looking-at-trait-climate-gradients">An example looking at trait-climate gradients</h4>
<p>A simple workflow allows one to look at <a href="traits_and_climate_example.html">trait values across a climate gradient</a></p>
</section>
<section id="an-example-incorporating-ala-distribution-data" class="level4">
<h4 class="anchored" data-anchor-id="an-example-incorporating-ala-distribution-data">An example incorporating ALA distribution data</h4>
<p>A recent tutorial posted by ALA shows how one can combine AusTraits trait data and ALA spatial occurrence data:</p>
<p>https://labs.ala.org.au/posts/2023-08-28_alternatives-to-box-plots/post.html</p>
<p>We’ve adopted it <a href="spatial_data_example.html">here</a>.</p>
</section>
</section>
</section>
<section id="a-complexity-pivoting-datasets-pivotting_datasets" class="level2" data-number="30.9">
<h2 data-number="30.9" class="anchored" data-anchor-id="a-complexity-pivoting-datasets-pivotting_datasets"><span class="header-section-number">30.9</span> A complexity: pivoting datasets {pivotting_datasets}</h2>
<p>The AusTraits tables are all in <code>long</code> format with an individual row for each trait measurement. This is the most compact way to store data and offers the flexibility of documenting diverse metadata for each trait measurement.</p>
<p>However, for many research uses, it may be more useful to view data in a <code>wide</code> format, with the multiple traits that comprise a single observation displayed as consecutive columns.</p>
<p>The <code>{austraits}</code> function <code>trait_pivot_wider</code> allows AusTraits datasets to be pivoted from <code>long</code> to <code>wide</code> format.</p>
<p>It is recommended to only use this function on individual datasets – or perhaps a small selection of datasets – as each dataset includes a different collection of traits and pivoting wider otherwise creates a very “holey” dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb172"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a>Farrell_2017_values <span class="ot">&lt;-</span></span>
<span id="cb172-2"><a href="#cb172-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb172-3"><a href="#cb172-3" aria-hidden="true" tabindex="-1"></a>    austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="at">dataset_id =</span> <span class="st">"Farrell_2017"</span>)</span>
<span id="cb172-4"><a href="#cb172-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-5"><a href="#cb172-5" aria-hidden="true" tabindex="-1"></a>Farrell_2017_pivoted <span class="ot">&lt;-</span> </span>
<span id="cb172-6"><a href="#cb172-6" aria-hidden="true" tabindex="-1"></a>  Farrell_2017_values<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb172-7"><a href="#cb172-7" aria-hidden="true" tabindex="-1"></a>    austraits<span class="sc">::</span><span class="fu">trait_pivot_wider</span>()</span>
<span id="cb172-8"><a href="#cb172-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb172-9"><a href="#cb172-9" aria-hidden="true" tabindex="-1"></a>Farrell_2017_pivoted</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 177 × 25
   dataset_id   taxon_name observation_id entity_type basis_of_record life_stage
   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;          &lt;chr&gt;       &lt;chr&gt;           &lt;chr&gt;     
 1 Farrell_2017 Arthropod… 001            individual  lab             adult     
 2 Farrell_2017 Arthropod… 002            individual  lab             adult     
 3 Farrell_2017 Arthropod… 003            individual  lab             adult     
 4 Farrell_2017 Arthropod… 004            individual  lab             adult     
 5 Farrell_2017 Arthropod… 005            individual  lab             adult     
 6 Farrell_2017 Arthropod… 006            individual  lab             adult     
 7 Farrell_2017 Arthropod… 007            individual  lab             adult     
 8 Farrell_2017 Arthropod… 008            individual  lab             adult     
 9 Farrell_2017 Arthropod… 009            individual  lab             adult     
10 Farrell_2017 Chrysocep… 010            individual  lab             adult     
# ℹ 167 more rows
# ℹ 19 more variables: population_id &lt;chr&gt;, individual_id &lt;chr&gt;,
#   repeat_measurements_id &lt;chr&gt;, temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;,
#   location_id &lt;chr&gt;, entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;,
#   treatment_context_id &lt;chr&gt;, collection_date &lt;chr&gt;, method_id &lt;chr&gt;,
#   method_context_id &lt;chr&gt;, original_name &lt;chr&gt;,
#   bulk_modulus_of_elasticity &lt;chr&gt;, …</code></pre>
</div>
</div>
<p>This example pivots “nicely” as all observations have <code>entity_type = individual</code>.</p>
<p>Compare this first example to the dataset <code>Edwards_2000</code> which includes individual-, population-, and species-level observations:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb174"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb174-1"><a href="#cb174-1" aria-hidden="true" tabindex="-1"></a>Edwards_2000_values <span class="ot">&lt;-</span></span>
<span id="cb174-2"><a href="#cb174-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb174-3"><a href="#cb174-3" aria-hidden="true" tabindex="-1"></a>    austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="at">dataset_id =</span> <span class="st">"Edwards_2000"</span>)</span>
<span id="cb174-4"><a href="#cb174-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb174-5"><a href="#cb174-5" aria-hidden="true" tabindex="-1"></a>Edwards_2000_pivoted <span class="ot">&lt;-</span> </span>
<span id="cb174-6"><a href="#cb174-6" aria-hidden="true" tabindex="-1"></a>  Edwards_2000_values<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb174-7"><a href="#cb174-7" aria-hidden="true" tabindex="-1"></a>    austraits<span class="sc">::</span><span class="fu">trait_pivot_wider</span>()</span>
<span id="cb174-8"><a href="#cb174-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb174-9"><a href="#cb174-9" aria-hidden="true" tabindex="-1"></a>Edwards_2000_pivoted</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 139 × 35
   dataset_id   taxon_name observation_id entity_type basis_of_record life_stage
   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;          &lt;chr&gt;       &lt;chr&gt;           &lt;chr&gt;     
 1 Edwards_2000 Acacia me… 001            individual  field           adult     
 2 Edwards_2000 Acacia me… 002            individual  field           adult     
 3 Edwards_2000 Acacia me… 003            individual  field           adult     
 4 Edwards_2000 Acacia me… 004            individual  field           adult     
 5 Edwards_2000 Acacia me… 005            individual  field           adult     
 6 Edwards_2000 Acacia me… 006            population  field           adult     
 7 Edwards_2000 Acacia me… 007            species     field           adult     
 8 Edwards_2000 Acacia su… 008            individual  field           adult     
 9 Edwards_2000 Acacia su… 009            individual  field           adult     
10 Edwards_2000 Acacia su… 010            individual  field           adult     
# ℹ 129 more rows
# ℹ 29 more variables: population_id &lt;chr&gt;, individual_id &lt;chr&gt;,
#   repeat_measurements_id &lt;chr&gt;, temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;,
#   location_id &lt;chr&gt;, entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;,
#   treatment_context_id &lt;chr&gt;, collection_date &lt;chr&gt;, method_id &lt;chr&gt;,
#   method_context_id &lt;chr&gt;, original_name &lt;chr&gt;, leaf_N_per_area &lt;chr&gt;,
#   leaf_N_per_dry_mass &lt;chr&gt;, leaf_area &lt;chr&gt;, leaf_mass_per_area &lt;chr&gt;, …</code></pre>
</div>
</div>
<p>The values at the individual, population and species level do not collapse together, because traits measured on different <code>entity_types</code> have separate <code>observation_id</code>’s.</p>
<p>One of the core identifiers assigned to data points is the <code>observation_id</code>. An observation is a collection of measurements made on a specific entity at a single point in time.</p>
<p>Observation_id’s are, therefore, unique combinations of:</p>
<ul>
<li>dataset_id</li>
<li>source_id</li>
<li>entity_type</li>
<li>taxon_name</li>
<li>population_id (location_id, plot_context_id, treatment_context_id)</li>
<li>individual_id</li>
<li>basis_of_record</li>
<li>entity_context_id</li>
<li>life_stage</li>
<li>temporal_context_id</li>
<li>collection_date</li>
<li>original_name</li>
</ul>
<p>If a single dataset includes traits that are attributed to different entity types, they are assigned separate <code>observation_id</code>’s. For instance, many datasets are comprised of individual-level physiological trait data and a column <code>growth_form</code>, documenting the growth form (i.e.&nbsp;tree, shrub, herb, etc.) of each <em>species</em>.</p>
<p>We’re developing a function, <code>merge_entity_types</code> that collapses the pivoted data into a more condensed table, but this loses some of the metadata. This function is currently in the R file <code>extra_functions.R</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb176"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a>Edwards_2000_pivoted_merged <span class="ot">&lt;-</span></span>
<span id="cb176-2"><a href="#cb176-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">merge_entity_types</span>(<span class="st">"Edwards_2000"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><p>This function will duplicate any “higher-entity” trait values (e.g.&nbsp;A single species-level value is filled in for all individuals or populations)</p></li>
<li><p>Metadata fields, like <code>entity_type</code> or <code>value_type</code>, are only retained if their values are identical for all measurements</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a>Westoby_2014_pivoted_merged <span class="ot">&lt;-</span></span>
<span id="cb177-2"><a href="#cb177-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">merge_entity_types</span>(<span class="st">"Westoby_2014"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="intepreting-trait-names-taxon-names" class="level2" data-number="30.10">
<h2 data-number="30.10" class="anchored" data-anchor-id="intepreting-trait-names-taxon-names"><span class="header-section-number">30.10</span> Intepreting trait names, taxon names</h2>
<section id="trait-dictionary" class="level3">
<h3 class="anchored" data-anchor-id="trait-dictionary">Trait dictionary</h3>
<p>The <code>{traits.build}</code> pipeline requires a trait dictionary that documents 4 pieces of information about each trait:</p>
<ul>
<li>trait name&nbsp;</li>
<li>trait type (categorical vs numeric)&nbsp;</li>
<li>allowable trait values (for categorical traits)&nbsp;</li>
<li>allowable trait range and units (for numeric traits)&nbsp;</li>
</ul>
<p>The trait dictionary embedded within AusTraits also has:</p>
<ul>
<li>trait labels&nbsp;</li>
<li>trait definitions&nbsp;</li>
<li>definitions for all categorical trait values&nbsp;</li>
</ul>
<p>Together these clarify each “trait concept”, which we define as: “a circumscribed set of trait measurements”. Much like a taxon concept delimits a collection of organisms, a trait concept delimits a collection of trait values pertaining to a distinct characteristic of a specific part of an organism (cell, tissue, organ, or whole organism).</p>
<p>The <a href="http:///w3id.org/APD">AusTraits Plant Dictionary (APD)</a> offers detailed descriptions for all trait concepts included in AusTraits. With the APD, each trait is given a unique, resolvable identifier, allowing trait definitions to be reused and shared.</p>
<p>The trait dictionary also includes:&nbsp;</p>
<ul>
<li>keywords&nbsp;</li>
<li>plant structure measured&nbsp;</li>
<li>characteristic measured&nbsp;</li>
<li>references&nbsp;</li>
<li>links to the same (or similar) trait concepts in other databases and dictionaries&nbsp;</li>
</ul>
</section>
<section id="understanding-taxon-names" class="level3">
<h3 class="anchored" data-anchor-id="understanding-taxon-names">Understanding taxon names</h3>
<p>AusTraits uses the taxon names in the Australian Plant Census (APC) and the scientific names in the Australian Plant Names Index (APNI).</p>
<p>The R package <a href="https://github.com/traitecoevo/APCalign"><code>{APCalign}</code></a> is used to align and update taxon names submitted to AusTraits with those in the APC/APNI.</p>
<p><code>{APCalign}</code> can be installed directly from CRAN</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb178"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("APCalign")</span></span>
<span id="cb178-2"><a href="#cb178-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-3"><a href="#cb178-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(APCalign) <span class="co"># Australian plant taxon alignment function, available on CRAN </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are two key components to the workflow:&nbsp;</p>
<ol type="1">
<li>aligning names&nbsp;</li>
</ol>
<ul>
<li><p>syntax is standardised, including for phrase names&nbsp;</p></li>
<li><p>most spelling mistakes are corrected&nbsp;</p></li>
<li><p>names that indicate the plant can only be identified to genus are reformatted to <code>genus sp. [available notes; dataset_id]</code>&nbsp;</p>
<ol type="1">
<li><p>they are linked to an APC-accepted genus but not to an APC-accepted binomial.&nbsp;</p></li>
<li><p>they include the dataset_id so people don’t mistakenly group all <code>Eucalyptus sp.</code> as a single “species”&nbsp;</p></li>
</ol></li>
</ul>
<ol start="2" type="1">
<li>updating names&nbsp;</li>
</ol>
<ul>
<li>all aligned names that are in the APC, but that have a <code>taxonomic status</code> other than <code>accepted</code>, are updated to their currently accepted name.&nbsp;</li>
</ul>
<p>Examples:&nbsp;</p>
<p>Identical <code>genus sp.</code> inputs from disparate datasets are given unique “names”:&nbsp;</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb179"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb179-1"><a href="#cb179-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb179-2"><a href="#cb179-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(stringr<span class="sc">::</span><span class="fu">str_detect</span>(original_name, <span class="st">"Eucalyptus sp</span><span class="sc">\\</span><span class="st">.$"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb179-3"><a href="#cb179-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(dataset_id, taxon_name, original_name) <span class="sc">%&gt;%</span></span>
<span id="cb179-4"><a href="#cb179-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(original_name <span class="sc">!=</span> taxon_name) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 3
  dataset_id     taxon_name                      original_name 
  &lt;chr&gt;          &lt;chr&gt;                           &lt;chr&gt;         
1 Clarke_2015    Eucalyptus sp. [Clarke_2015]    Eucalyptus sp.
2 Nolan_2022     Eucalyptus sp. [Nolan_2022]     Eucalyptus sp.
3 Soper_2014     Eucalyptus sp. [Soper_2014]     Eucalyptus sp.
4 Zieminska_2015 Eucalyptus sp. [Zieminska_2015] Eucalyptus sp.</code></pre>
</div>
</div>
<p>Outdated names are updated:&nbsp;</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb181"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb181-1"><a href="#cb181-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb181-2"><a href="#cb181-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(stringr<span class="sc">::</span><span class="fu">str_detect</span>(original_name, <span class="st">"Dryandra"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb181-3"><a href="#cb181-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name, original_name) <span class="sc">%&gt;%</span></span>
<span id="cb181-4"><a href="#cb181-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(original_name <span class="sc">!=</span> taxon_name) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 15 × 2
   taxon_name                                 original_name                     
   &lt;chr&gt;                                      &lt;chr&gt;                             
 1 Banksia acanthopoda                        Dryandra acanthopoda              
 2 Banksia anatona                            Dryandra anatona                  
 3 Banksia arborea                            Dryandra arborea                  
 4 Banksia arctotidis                         Dryandra arctotidis               
 5 Banksia armata var. armata                 Dryandra armata var. armata       
 6 Banksia armata var. ignicida               Dryandra armata var. ignicida     
 7 Banksia armata                             Dryandra armata                   
 8 Banksia aurantia                           Dryandra aurantia                 
 9 Banksia biterax                            Dryandra baxteri                  
10 Banksia bipinnatifida subsp. bipinnatifida Dryandra bipinnatifida subsp. bip…
11 Banksia bipinnatifida subsp. multifida     Dryandra bipinnatifida subsp. mul…
12 Banksia bipinnatifida                      Dryandra bipinnatifida            
13 Banksia pellaeifolia                       Dryandra blechnifolia             
14 Banksia borealis subsp. borealis           Dryandra borealis subsp. borealis 
15 Banksia borealis subsp. elatior            Dryandra borealis subsp. elatior  </code></pre>
</div>
</div>
<p>Phrase name syntax across datasets is aligned:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb183"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb183-1"><a href="#cb183-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb183-2"><a href="#cb183-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(stringr<span class="sc">::</span><span class="fu">str_detect</span>(taxon_name, <span class="st">"Argyrodendron sp. Whyanbeel"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb183-3"><a href="#cb183-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name, original_name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  taxon_name                                       original_name                
  &lt;chr&gt;                                            &lt;chr&gt;                        
1 Argyrodendron sp. Whyanbeel (B.P.Hyland RFK1106) Argyrodendron (Whyanbeel)    
2 Argyrodendron sp. Whyanbeel (B.P.Hyland RFK1106) Argyrodendron ssp. (Whyanbee…
3 Argyrodendron sp. Whyanbeel (B.P.Hyland RFK1106) Argyrodendron Whyanbeel      
4 Argyrodendron sp. Whyanbeel (B.P.Hyland RFK1106) Argyrodendron sp. (Whyanbeel…
5 Argyrodendron sp. Whyanbeel (B.P.Hyland RFK1106) Argyrodendron sp. Whyanbeel …
6 Argyrodendron sp. Whyanbeel (B.P.Hyland RFK1106) Argyrodendron sp. Whyanbeel …</code></pre>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./austraits_database.html" class="pagination-link" aria-label="AusTraits, Australia's Plant Trait Database">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">AusTraits, Australia’s Plant Trait Database</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./traits_and_climate_example.html" class="pagination-link" aria-label="Analysis example: Using AusTraits for trait-climate analyses">
        <span class="nav-page-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">Analysis example: Using AusTraits for trait-climate analyses</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>How to get help: <a href="https://traitecoevo.github.io/traits.build-book/help.html" class="uri">https://traitecoevo.github.io/traits.build-book/help.html</a><br>Copyright 2023, Daniel Falster and Elizabeth Wenk</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>