<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>The {traits.build} data standard, R package, and workflow - 29&nbsp; AusTraits tutorial</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./traits_and_climate_example.html" rel="next">
<link href="./austraits_database.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./austraits_database.html">Using outputs of <code>traits.build</code></a></li><li class="breadcrumb-item"><a href="./AusTraits_tutorial.html"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">AusTraits tutorial</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">The {traits.build} data standard, R package, and workflow</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Introduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">About</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./motivation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Motivation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./usage_examples.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Usage examples</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Data structure and standard</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./long_wide.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Long vs Wide data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_standard.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data standard</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_structure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Data structure</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_metadata.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Overview of traits.build ontology</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dictionary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Trait dictionary</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Creating with <code>traits.build</code></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./traits_build.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">The package</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_compilation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Tutorial: Example compilation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./file_organisation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">File organisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./create_dictionary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Creating a dictionary</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./adding_data_brief.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Adding datasets (brief)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./custom_R_code.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Adding custom_R_code</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./publishing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Publishing</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Guide to adding data</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Tutorial: Adding datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Tutorial 1: Adding a simple dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Tutorial 2: Adding a more complex dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Tutorial 3: Adding contexts and complex units</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Tutorial 4: Additional complexities</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Tutorial 5: Multiple columns for a trait</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_6.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Tutorial 6: Data with repeat measurements</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_7.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Tutorial 7: Adding long format dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./adding_data_long.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">Adding datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./check_dataset_functions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">Check data quality functions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_common_issues.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">Common issues</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Using outputs of <code>traits.build</code></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./austraits_database.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">AusTraits, Australia’s Plant Trait Database</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./AusTraits_tutorial.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">AusTraits tutorial</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./traits_and_climate_example.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">Combining trait values and climate data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spatial_data_example.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">Tutorial: Using AusTraits with spatial data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./austraits_package.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">The <code>austraits</code> package</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Getting help</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./help.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">Getting help</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./debugging.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">34</span>&nbsp; <span class="chapter-title">Debugging</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./csv.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">CSV files</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./yaml.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Yaml files</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">29.1</span> Introduction</a></li>
  <li><a href="#download-austraits-data" id="toc-download-austraits-data" class="nav-link" data-scroll-target="#download-austraits-data"><span class="header-section-number">29.2</span> Download AusTraits data</a></li>
  <li><a href="#exploring" id="toc-exploring" class="nav-link" data-scroll-target="#exploring"><span class="header-section-number">29.3</span> A first look at data</a></li>
  <li><a href="#database_structure" id="toc-database_structure" class="nav-link" data-scroll-target="#database_structure"><span class="header-section-number">29.4</span> The database structure</a>
  <ul class="collapse">
  <li><a href="#traits-table" id="toc-traits-table" class="nav-link" data-scroll-target="#traits-table">Traits table</a></li>
  <li><a href="#ancillary-data-tables" id="toc-ancillary-data-tables" class="nav-link" data-scroll-target="#ancillary-data-tables">Ancillary data tables</a></li>
  </ul></li>
  <li><a href="#exploring-austraits" id="toc-exploring-austraits" class="nav-link" data-scroll-target="#exploring-austraits"><span class="header-section-number">29.5</span> Exploring AusTraits</a>
  <ul class="collapse">
  <li><a href="#austraits.r-function-reference" id="toc-austraits.r-function-reference" class="nav-link" data-scroll-target="#austraits.r-function-reference">austraits.R function reference</a></li>
  <li><a href="#summarise_functions" id="toc-summarise_functions" class="nav-link" data-scroll-target="#summarise_functions">Summarising data: data coverage</a></li>
  <li><a href="#filtering_functions" id="toc-filtering_functions" class="nav-link" data-scroll-target="#filtering_functions">Filtering data</a></li>
  <li><a href="#join_functions" id="toc-join_functions" class="nav-link" data-scroll-target="#join_functions">Joining relational tables</a></li>
  </ul></li>
  <li><a href="#summarising-data-trait-means-modes-etc." id="toc-summarising-data-trait-means-modes-etc." class="nav-link" data-scroll-target="#summarising-data-trait-means-modes-etc."><span class="header-section-number">29.6</span> Summarising data: trait means, modes, etc.</a>
  <ul class="collapse">
  <li><a href="#categorical-traits" id="toc-categorical-traits" class="nav-link" data-scroll-target="#categorical-traits">Categorical traits</a></li>
  <li><a href="#numeric-traits" id="toc-numeric-traits" class="nav-link" data-scroll-target="#numeric-traits">Numeric traits</a></li>
  </ul></li>
  <li><a href="#plotting_functions" id="toc-plotting_functions" class="nav-link" data-scroll-target="#plotting_functions"><span class="header-section-number">29.7</span> Plotting data</a>
  <ul class="collapse">
  <li><a href="#plotting-trait-distributions" id="toc-plotting-trait-distributions" class="nav-link" data-scroll-target="#plotting-trait-distributions">Plotting trait distributions</a></li>
  <li><a href="#plotting-data-distribution-by-location" id="toc-plotting-data-distribution-by-location" class="nav-link" data-scroll-target="#plotting-data-distribution-by-location">Plotting data distribution by location</a></li>
  <li><a href="#more-complex-workflows---some-examples" id="toc-more-complex-workflows---some-examples" class="nav-link" data-scroll-target="#more-complex-workflows---some-examples">More complex workflows - some examples</a></li>
  </ul></li>
  <li><a href="#a-complexity-pivoting-datasets" id="toc-a-complexity-pivoting-datasets" class="nav-link" data-scroll-target="#a-complexity-pivoting-datasets"><span class="header-section-number">29.8</span> A complexity: pivoting datasets</a></li>
  <li><a href="#intepreting-trait-names-taxon-names" id="toc-intepreting-trait-names-taxon-names" class="nav-link" data-scroll-target="#intepreting-trait-names-taxon-names"><span class="header-section-number">29.9</span> Intepreting trait names, taxon names</a>
  <ul class="collapse">
  <li><a href="#trait-dictionary" id="toc-trait-dictionary" class="nav-link" data-scroll-target="#trait-dictionary">Trait dictionary</a></li>
  <li><a href="#understanding-taxon-names" id="toc-understanding-taxon-names" class="nav-link" data-scroll-target="#understanding-taxon-names">Understanding taxon names</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">AusTraits tutorial</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="introduction" class="level2" data-number="29.1">
<h2 data-number="29.1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">29.1</span> Introduction</h2>
<p>With more than 1.8 million data records, AusTraits is Australia’s <a href="austraits_database.html#plant_database">largest plant trait database</a>, created using the <a href="https://github.com/traitecoevo/traits.build"><code>{traits.build}</code> R package</a></p>
<p>This tutorial introduces:</p>
<ul>
<li><p><a href="#database_structure">the database structure</a></p></li>
<li><p><a href="#austraits_functions"><code>{austraits}</code> R package functions</a></p></li>
<li><p>additional <a href="#sample_analyses">examples of analyses</a> using the database</p></li>
</ul>
<p>To access more information about <code>traits.build</code> see <a href="https://traitecoevo.github.io/traits.build-book/"><code>traits.build-book</code></a></p>
<p>Or you can visit the Github repositories for individual packages/data repos:</p>
<ul>
<li><p>the database structure, <a href="https://github.com/traitecoevo/traits.build"><code>traits.build</code></a></p></li>
<li><p>the database contents, <a href="https://github.com/traitecoevo/austraits.build"><code>austraits.build</code></a></p></li>
<li><p>an R package for exploring and wrangling the data, <a href="https://github.com/traitecoevo/austraits"><code>austraits</code></a></p></li>
</ul>
</section>
<section id="download-austraits-data" class="level2" data-number="29.2">
<h2 data-number="29.2" class="anchored" data-anchor-id="download-austraits-data"><span class="header-section-number">29.2</span> Download AusTraits data</h2>
<p>Before you begin, download and source essential packages and functions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co">#remotes::install_github("traitecoevo/austraits", quick=TRUE)</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(austraits)    <span class="co"># functions for exploring a traits.build database, available at Github repo</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(traits.build) <span class="co"># additional functions for exploring a traits.build database, available at Github repo</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"data/extra_functions.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then download (or build) the latest AusTraits data, by one of the methods described <a href="austraits_database.html#access_data">here</a>.</p>
<p>This tutorial uses the most recent AusTraits release, version 5.0.0</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>most_recent <span class="ot">&lt;-</span> austraits<span class="sc">::</span><span class="fu">get_versions</span>()<span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">pull</span>(<span class="st">"doi"</span>) <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">first</span>()</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>most_recent</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "10.5281/zenodo.10156222"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>austraits <span class="ot">&lt;-</span> austraits<span class="sc">::</span><span class="fu">load_austraits</span>(<span class="at">doi =</span> most_recent)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Downloading AusTraits to 'data/austraits'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading data from 'data/austraits/austraits-5.0.0.rds'</code></pre>
</div>
</div>
</section>
<section id="exploring" class="level2" data-number="29.3">
<h2 data-number="29.3" class="anchored" data-anchor-id="exploring"><span class="header-section-number">29.3</span> A first look at data</h2>
<p>If you’re not familiar with AusTraits, you may want to begin by exploring the breadth and depth of data within the database. The database can be explored by trait name, species, or genus using either <a href="#austraits_functions">austraits functions</a> or dplyr functions.</p>
<p><em>How many taxa have <code>leaf_N_per_dry_mass</code> data in AusTraits?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(trait_name <span class="sc">==</span> <span class="st">"leaf_N_per_dry_mass"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name) <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2230</code></pre>
</div>
</div>
<p><em>How are these data distributed across datasets?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_trait_distribution_beeswarm</span>(austraits, <span class="st">"leaf_N_per_dry_mass"</span>, <span class="at">y_axis_category =</span> <span class="st">"dataset_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Orientation inferred to be along y-axis; override with
`position_quasirandom(orientation = 'x')`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><em>How much data exist for other nitrogen traits?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"_N_"</span>) <span class="ot">-&gt;</span> N_traits</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(trait_name <span class="sc">%in%</span> N_traits) <span class="sc">%&gt;%</span> </span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name, trait_name) <span class="sc">%&gt;%</span> </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(trait_name) <span class="sc">%&gt;%</span> </span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">n_taxa =</span> <span class="fu">length</span>(trait_name)) <span class="sc">%&gt;%</span> </span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(trait_name, n_taxa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 16 × 2
   trait_name                              n_taxa
   &lt;chr&gt;                                    &lt;int&gt;
 1 leaf_N_per_dry_mass                       2230
 2 leaf_N_per_area                            708
 3 wood_N_per_dry_mass                         69
 4 leaf_senesced_N_per_dry_mass                81
 5 bark_N_per_dry_mass                         49
 6 fruit_N_per_dry_mass                        22
 7 seed_N_per_seed_dry_mass                    42
 8 stem_N_per_dry_mass                         28
 9 wood_dead_N_per_dry_mass                     7
10 flower_N_per_dry_mass                        2
11 leaf_N_resorption                           14
12 leaf_cell_wall_N_per_cell_wall_dry_mass     22
13 leaf_cell_wall_N_per_leaf_N                 22
14 root_N_per_dry_mass                         22
15 leaf_rubisco_N_per_total_leaf_N             29
16 leaf_thylakoid_N_per_total_leaf_N           29</code></pre>
</div>
</div>
<p><em>How many “hydraulic” traits are in AusTraits? How much data exist for these traits?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"hydraulic"</span>) <span class="ot">-&gt;</span> hydraulic_traits</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(trait_name <span class="sc">%in%</span> hydraulic_traits) <span class="sc">%&gt;%</span> </span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name, trait_name) <span class="sc">%&gt;%</span> </span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(trait_name) <span class="sc">%&gt;%</span> </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">n_taxa =</span> <span class="fu">length</span>(trait_name)) <span class="sc">%&gt;%</span> </span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(trait_name, n_taxa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 2
   trait_name                                          n_taxa
   &lt;chr&gt;                                                &lt;int&gt;
 1 sapwood_specific_hydraulic_conductivity_theoretical    164
 2 stem_vessel_diameter_hydraulic                         264
 3 leaf_hydraulic_vulnerability                            20
 4 leaf_specific_hydraulic_conductance                     79
 5 leaf_specific_hydraulic_conductivity                   168
 6 sapwood_specific_hydraulic_conductivity                179
 7 stem_specific_hydraulic_conductivity                    14
 8 stem_hydraulic_conductivity                             51
 9 hydraulic_safety_margin_50                              31
10 hydraulic_safety_margin_88                              26</code></pre>
</div>
</div>
<p><em>Where have trait data for Acacia aneura been collected?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">taxon_name =</span> <span class="st">"Acacia aneura"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"taxon_name"</span>) <span class="co"># 1 taxon, so 1 plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><em>Where have data for Hibbertia species been collected?</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">genus =</span> <span class="st">"Hibbertia"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_locations</span>() <span class="sc">%&gt;%</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_taxonomy</span>()</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"genus"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="database_structure" class="level2" data-number="29.4">
<h2 data-number="29.4" class="anchored" data-anchor-id="database_structure"><span class="header-section-number">29.4</span> The database structure</h2>
<p>The <code>{traits.build}</code> R package is the workflow that builds AusTraits from its component datasets.</p>
<p>The database is output as a collection of relational tables, described in detail <a href="database_structure.html">here</a>.</p>
<p>The data object <code>austraits</code> includes both the relational data tables and additional tables documenting database metadata and the traits dictionary referenced for trait concepts.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(austraits)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "traits"            "locations"         "contexts"         
 [4] "methods"           "excluded_data"     "taxonomic_updates"
 [7] "taxa"              "contributors"      "sources"          
[10] "definitions"       "schema"            "metadata"         
[13] "build_info"       </code></pre>
</div>
</div>
<section id="traits-table" class="level3">
<h3 class="anchored" data-anchor-id="traits-table">Traits table</h3>
<p>The core AusTraits table is the traits table. It is in “long” format, with each row documenting a single trait measurement.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 26
   dataset_id taxon_name       observation_id trait_name value unit  entity_type
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 ABRS_1981  Acanthocarpus c… 0001           leaf_comp… simp… &lt;NA&gt;  species    
 2 ABRS_1981  Acanthocarpus c… 0001           leaf_leng… 3     mm    species    
 3 ABRS_1981  Acanthocarpus c… 0001           leaf_leng… 15    mm    species    
 4 ABRS_1981  Acanthocarpus c… 0001           seed_heig… 3     mm    species    
 5 ABRS_1981  Acanthocarpus c… 0001           seed_leng… 3     mm    species    
 6 ABRS_1981  Acanthocarpus c… 0001           seed_width 3     mm    species    
 7 ABRS_1981  Acanthocarpus h… 0002           leaf_comp… simp… &lt;NA&gt;  species    
 8 ABRS_1981  Acanthocarpus h… 0002           leaf_leng… 4     mm    species    
 9 ABRS_1981  Acanthocarpus h… 0002           leaf_leng… 12    mm    species    
10 ABRS_1981  Acanthocarpus h… 0002           seed_leng… 4     mm    species    
11 ABRS_1981  Acanthocarpus p… 0003           leaf_comp… simp… &lt;NA&gt;  species    
12 ABRS_1981  Acanthocarpus p… 0003           leaf_leng… 3     mm    species    
13 ABRS_1981  Acanthocarpus p… 0003           seed_leng… 3.5   mm    species    
14 ABRS_1981  Acanthocarpus p… 0003           seed_leng… 4     mm    species    
15 ABRS_1981  Acanthocarpus p… 0004           leaf_comp… simp… &lt;NA&gt;  species    
16 ABRS_1981  Acanthocarpus p… 0004           leaf_leng… 20    mm    species    
17 ABRS_1981  Acanthocarpus p… 0004           leaf_leng… 70    mm    species    
18 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 4     mm    species    
19 ABRS_1981  Acanthocarpus p… 0004           seed_leng… 5     mm    species    
20 ABRS_1981  Acanthocarpus r… 0005           leaf_comp… simp… &lt;NA&gt;  species    
# ℹ 19 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;,
#   replicates &lt;chr&gt;, basis_of_record &lt;chr&gt;, life_stage &lt;chr&gt;,
#   population_id &lt;chr&gt;, individual_id &lt;chr&gt;, repeat_measurements_id &lt;chr&gt;,
#   temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;, location_id &lt;chr&gt;,
#   entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;, treatment_context_id &lt;chr&gt;,
#   collection_date &lt;chr&gt;, measurement_remarks &lt;chr&gt;, method_id &lt;chr&gt;,
#   method_context_id &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
</div>
<p>The columns include:</p>
<ul>
<li>core columns&nbsp;
<ul>
<li>dataset_id&nbsp;</li>
<li>taxon_name&nbsp;</li>
<li>trait_name&nbsp;</li>
<li>value (trait value)&nbsp;</li>
</ul></li>
<li>entity metadata
<ul>
<li>entity_type<br>
</li>
<li>life_stage</li>
</ul></li>
<li>value metadata
<ul>
<li>value_type<br>
</li>
<li>unit<br>
</li>
<li>basis_of_value<br>
</li>
<li>replicate<br>
</li>
<li>basis_of_record<br>
</li>
</ul></li>
<li>additional metadata
<ul>
<li>collection_date</li>
<li>measurement_remarks</li>
</ul></li>
<li>identifiers for specific observations, individuals, etc.
<ul>
<li>observation_id<br>
</li>
<li>individual_id<br>
</li>
<li>population_id<br>
</li>
<li>repeat_measurements_id<br>
</li>
</ul></li>
<li>identifiers that provide links to ancillary tables with additional metadata
<ul>
<li>location_id<br>
</li>
<li>treatment_context_id<br>
</li>
<li>plot_context_id<br>
</li>
<li>entity_context_id<br>
</li>
<li>temporal_context_id<br>
</li>
<li>method_context_id<br>
</li>
<li>method_id<br>
</li>
<li>source_id</li>
</ul></li>
</ul>
</section>
<section id="ancillary-data-tables" class="level3">
<h3 class="anchored" data-anchor-id="ancillary-data-tables">Ancillary data tables</h3>
<p>The remaining metadata accompanying each trait record is recorded across multiple relational tables.</p>
<p>These include:&nbsp;</p>
<ul>
<li>austraits$locations&nbsp;</li>
<li>austraits$contexts&nbsp;</li>
<li>austraits$methods&nbsp;</li>
<li>austraits$taxa&nbsp;</li>
<li>austraits$taxonomic_updates&nbsp;</li>
<li>austraits$contributors&nbsp;</li>
</ul>
<p>Like the core <code>traits</code> table, each is in ‘long’ format.</p>
<p>The tables <code>locations</code>, <code>contexts</code>, <code>methods</code> and <code>taxa</code> include metadata that links seamlessly to individual rows within <code>traits</code>.</p>
<p>As such, you can use the various <code>{dplyr}</code> join functions to join together the relational tables into a single (much) wider table. Some <code>{austraits}</code> package functions also join key columns to the traits table.</p>
<table class="table">
<colgroup>
<col style="width: 25%">
<col style="width: 36%">
<col style="width: 38%">
</colgroup>
<thead>
<tr class="header">
<th>table</th>
<th>metadata in table</th>
<th>columns that link to austraits$traits</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>locations</td>
<td>location name, location properties, latitude, longitude</td>
<td>dataset_id, location_id</td>
</tr>
<tr class="even">
<td>contexts</td>
<td>context name, context category (method context, temporal, entity context, plot, treatment), context property</td>
<td>dataset_id, link_id (identifier to link to: method_context_id, temporal_id, entity_context_id, plot_id, treatment_id), link_vals (identifier value to link to)</td>
</tr>
<tr class="odd">
<td>methods</td>
<td>dataset description, dataset sampling strategy, trait collection method, data collectors, data curators, dataset citation, source_id &amp; citation</td>
<td>dataset_id, trait_name, method_id</td>
</tr>
<tr class="even">
<td>taxa</td>
<td>genus, family, scientific name, APC/APNI taxon concept/taxon name identifiers</td>
<td>taxon_name</td>
</tr>
<tr class="odd">
<td>taxonomic_updates</td>
<td>original name (name submitted), aligned name (typos removed; standardised syntax), identifiers for aligned name</td>
<td>dataset_id, taxon_name, original_name</td>
</tr>
<tr class="even">
<td>contributors</td>
<td>people who contributed data, including their ORCIDs, affiliations, roles</td>
<td>dataset_id</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="exploring-austraits" class="level2" data-number="29.5">
<h2 data-number="29.5" class="anchored" data-anchor-id="exploring-austraits"><span class="header-section-number">29.5</span> Exploring AusTraits</h2>
<p>With 1813897 rows of trait values in the main traits table, knowing how to explore the contents is essential.</p>
<p>The R package <a href="https://github.com/traitecoevo/austraits"><code>{austraits}</code></a> offers a collection of functions to explore and wrangle AusTraits data - or indeed any data using the traits.build format.</p>
<p>Function categories include:</p>
<ul>
<li><p><a href="#summarise_functions"><strong>summarise and lookup functions</strong></a>: These functions offer summaries by taxon name or trait, summarising taxa per trait (or other variable), datasets per trait, and observations per trait.</p></li>
<li><p><a href="#filtering_functions"><strong>filtering functions</strong></a>: These functions begin with the word <code>extract</code> and filter all of the relational tables simultaneously.</p></li>
<li><p><a href="#join_functions"><strong>join functions</strong></a>: These functions allow columns from the relational tables to be joined to the core traits table.</p></li>
<li><p><a href="#pivot_functions"><strong>pivot functions</strong></a>: These functions allow the traits table to be pivoted between long and wide formats.</p></li>
<li><p><a href="#plotting_functions"><strong>plotting functions</strong></a>: These functions offer a means to rapidly visualise AusTraits data, either plotting collection locations on a map of Australia or plotting trait values by dataset.</p></li>
</ul>
<section id="austraits.r-function-reference" class="level3">
<h3 class="anchored" data-anchor-id="austraits.r-function-reference">austraits.R function reference</h3>
<p>Reference guide to: <a href="https://traitecoevo.github.io/austraits/reference/index.html">austraits functions</a></p>
<p>(There is also a standalone <a href="austraits_package.html">austraits introduction</a>, which repeats some of the same material covered here.)</p>
</section>
<section id="summarise_functions" class="level3">
<h3 class="anchored" data-anchor-id="summarise_functions">Summarising data: data coverage</h3>
<p>There are three functions for summarising AusTraits data:</p>
<ul>
<li><strong>summarise_austraits()</strong><br>
</li>
<li><strong>summarise_trait_means()</strong><br>
</li>
<li><strong>lookup_trait()</strong></li>
</ul>
<p>Use the function <code>summarise_austraits</code> to output summaries of total records, datasets with records, and taxa with records across <code>families</code>, <code>genera</code> or <code>traits</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">summarise_austraits</span>(austraits, <span class="st">"trait_name"</span>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">100</span><span class="sc">:</span><span class="dv">130</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                     trait_name n_records n_dataset n_taxa percent_total
   flower_style_differentiation      1656         2   1547      9.13e-04
            flower_style_fusion       319         2    267      1.76e-04
                 flowering_cues       278         1    244      1.53e-04
                 flowering_time     32966        35  18215      1.82e-02
          fruit_Ca_per_dry_mass        27         4     21      1.49e-05
           fruit_K_per_dry_mass        27         4     21      1.49e-05
          fruit_Mg_per_dry_mass        27         4     21      1.49e-05
           fruit_N_per_dry_mass        31         5     22      1.71e-05
           fruit_P_per_dry_mass        29         5     23      1.60e-05
           fruit_S_per_dry_mass        27         4     21      1.49e-05
                   fruit_colour      6644         6   4745      3.66e-03
               fruit_dehiscence     41671        17  24297      2.30e-02
                 fruit_dry_mass       495         7    138      2.73e-04
               fruit_fleshiness     32575        15  21421      1.80e-02
                   fruit_height      8128         7   3316      4.48e-03
                   fruit_length     36965        16  12748      2.04e-02
                     fruit_type     56734        18  25352      3.13e-02
           fruit_wall_thickness       329         1     16      1.81e-04
                    fruit_width     23652        15   8273      1.30e-02
                  fruiting_time      3770         7   3470      2.08e-03
                    genome_size      1080         3    976      5.95e-04
                    huber_value      1494        19    358      8.24e-04
     hydraulic_safety_margin_50        46         2     31      2.54e-05
     hydraulic_safety_margin_88        40         2     26      2.21e-05
 integrated_plant_transpiration       202         3     32      1.11e-04
           leaf_Al_per_dry_mass       968         7    234      5.34e-04
            leaf_B_per_dry_mass      1140        10    352      6.28e-04
                  leaf_CN_ratio      1045         9    125      5.76e-04
            leaf_C_per_dry_mass      4797        29   1098      2.64e-03
           leaf_Ca_per_dry_mass      1886        20    509      1.04e-03
           leaf_Cl_per_dry_mass         6         2      6      3.31e-06</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">summarise_austraits</span>(austraits, <span class="at">var =</span>  <span class="st">"family"</span>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           family n_records n_dataset n_taxa percent_total
      Acanthaceae      3437        56    149      1.89e-03
      Achariaceae       162        14      3      8.93e-05
    Actinidiaceae       172        15      3      9.48e-05
    Agapanthaceae       105        12      3      5.79e-05
        Aizoaceae      4963        62    102      2.74e-03
       Akaniaceae       116        16      1      6.40e-05
     Alismataceae       853        30     20      4.70e-04
        Alliaceae       565        18     18      3.12e-04
   Alseuosmiaceae       318        13      3      1.75e-04
 Alstroemeriaceae       173        14      2      9.54e-05
    Amaranthaceae     10257        84    267      5.65e-03
   Amaryllidaceae      1391        37     56      7.67e-04
    Anacardiaceae      2192        64     26      1.21e-03
    Anarthriaceae       805        32     20      4.44e-04
       Annonaceae      4621        49     62      2.55e-03
 Aphanopetalaceae       188        22      3      1.04e-04
         Apiaceae     12372        93    195      6.82e-03
      Apocynaceae     15859       110    238      8.74e-03
    Apodanthaceae        79        11      4      4.36e-05
  Aponogetonaceae       475        19     14      2.62e-04</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">summarise_austraits</span>(austraits, <span class="at">var =</span>  <span class="st">"genus"</span>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          genus n_records n_dataset n_taxa percent_total
    (Dockrillia         3         2      1      1.65e-06
         Abelia        16         4      1      8.82e-06
    Abelmoschus       291        19      8      1.61e-04
   Abildgaardia        80         7      2      4.41e-05
    Abrodictyum       131        14      3      7.23e-05
         Abroma        39         7      2      2.15e-05
    Abrophyllum       191        19      3      1.05e-04
    Abrotanella       191        18      4      1.05e-04
          Abrus       219        26      3      1.21e-04
       Abutilon      2089        52     54      1.15e-03
         Acacia    111994       191   1558      6.18e-02
      Acaciella        66         5      3      3.64e-05
         Acaena      1205        44     14      6.65e-04
       Acalypha       594        20     14      3.28e-04
  Acanthocarpus      1030        22      9      5.68e-04
  Acanthocereus        22         4      1      1.21e-05
 Acanthocladium        60        13      1      3.31e-05
 Acanthospermum       132        16      2      7.28e-05
       Acanthus       198        18      4      1.09e-04
           Acer       247        15      4      1.36e-04</code></pre>
</div>
</div>
<p>Since this function summarises the variable selected for <strong>ALL</strong> of AusTraits, you will want to first <a href="#filtering_functions">filter</a> the data before summarising by “taxon_name” – or even “trait_name”.</p>
<p>Alternatively you can look up traits that contain a specific search term:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"leaf"</span>) <span class="sc">%&gt;%</span> <span class="fu">length</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 213</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"leaf"</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "leaf_compoundness"                   
 [2] "leaf_length"                         
 [3] "leaf_phenology"                      
 [4] "leaf_width"                          
 [5] "leaf_margin"                         
 [6] "leaf_shape"                          
 [7] "leaf_lamina_posture"                 
 [8] "leaf_base_shape"                     
 [9] "leaf_lobation"                       
[10] "leaf_phyllotaxis"                    
[11] "leaf_glaucousness"                   
[12] "leaf_arrangement"                    
[13] "leaf_margin_posture"                 
[14] "leaf_pendulousness"                  
[15] "leaf_N_per_dry_mass"                 
[16] "leaf_area"                           
[17] "leaf_delta13C"                       
[18] "leaf_mass_per_area"                  
[19] "leaf_photochemical_reflectance_index"
[20] "leaf_water_band_index"               
[21] "leaf_dry_matter_content"             
[22] "leaf_thickness"                      
[23] "leaf_water_use_efficiency_intrinsic" 
[24] "leaf_hairs_adult_leaves"             
[25] "leaf_hydraulic_vulnerability"        
[26] "leaf_specific_hydraulic_conductance" 
[27] "leaf_turgor_loss_point"              
[28] "leaf_NP_ratio"                       
[29] "leaf_N_per_area"                     
[30] "leaf_P_per_area"                     </code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"_N_"</span>) <span class="co"># elemental contents use their symbol and are always in the middle of a trait name</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "leaf_N_per_dry_mass"                    
 [2] "leaf_N_per_area"                        
 [3] "wood_N_per_dry_mass"                    
 [4] "leaf_senesced_N_per_dry_mass"           
 [5] "bark_N_per_dry_mass"                    
 [6] "fruit_N_per_dry_mass"                   
 [7] "seed_N_per_seed_dry_mass"               
 [8] "stem_N_per_dry_mass"                    
 [9] "wood_dead_N_per_dry_mass"               
[10] "flower_N_per_dry_mass"                  
[11] "leaf_N_resorption"                      
[12] "leaf_cell_wall_N_per_cell_wall_dry_mass"
[13] "leaf_cell_wall_N_per_leaf_N"            
[14] "root_N_per_dry_mass"                    
[15] "leaf_rubisco_N_per_total_leaf_N"        
[16] "leaf_thylakoid_N_per_total_leaf_N"      </code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"photo"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "plant_photosynthetic_organ"                             
 [2] "leaf_photochemical_reflectance_index"                   
 [3] "leaf_photosynthetic_nitrogen_use_efficiency_saturated"  
 [4] "leaf_photosynthetic_phosphorus_use_efficiency_saturated"
 [5] "leaf_photosynthetic_rate_per_area_maximum"              
 [6] "leaf_photosynthetic_rate_per_area_saturated"            
 [7] "leaf_photosynthetic_rate_per_dry_mass_maximum"          
 [8] "leaf_photosynthetic_rate_per_dry_mass_saturated"        
 [9] "leaf_photosynthesis_Jmax_per_area_25C"                  
[10] "leaf_photosynthesis_Vcmax_per_area"                     
[11] "leaf_photosynthesis_Jmax_per_area"                      
[12] "leaf_photosynthetic_rate_per_area_ambient"              
[13] "photosynthetic_pathway"                                 
[14] "leaf_photosynthesis_Jmax_per_mass"                      
[15] "leaf_photosynthesis_Vcmax_per_mass"                     
[16] "leaf_photosynthetic_nitrogen_use_efficiency_maximum"    
[17] "leaf_photosynthetic_phosphorus_use_efficiency_maximum"  
[18] "leaf_photosynthesis_Jmax_over_Vcmax_25C"                
[19] "leaf_photosynthesis_Vcmax_per_area_25C"                 
[20] "bark_photosynthetic_status"                             </code></pre>
</div>
</div>
<p>If you’d like to summarise similar metrics for a subset of data, it is best to write your own code to filter, then summarise based on specific variable values.</p>
<p>For instance, to just look at info for nitrogen-related traits:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"_N_"</span>) <span class="ot">-&gt;</span> N_traits</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(trait_name <span class="sc">%in%</span> N_traits) <span class="sc">%&gt;%</span> </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name, trait_name) <span class="sc">%&gt;%</span> </span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(trait_name) <span class="sc">%&gt;%</span> </span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">n_taxa =</span> <span class="fu">length</span>(trait_name)) <span class="sc">%&gt;%</span> </span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(trait_name, n_taxa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 16 × 2
   trait_name                              n_taxa
   &lt;chr&gt;                                    &lt;int&gt;
 1 leaf_N_per_dry_mass                       2230
 2 leaf_N_per_area                            708
 3 wood_N_per_dry_mass                         69
 4 leaf_senesced_N_per_dry_mass                81
 5 bark_N_per_dry_mass                         49
 6 fruit_N_per_dry_mass                        22
 7 seed_N_per_seed_dry_mass                    42
 8 stem_N_per_dry_mass                         28
 9 wood_dead_N_per_dry_mass                     7
10 flower_N_per_dry_mass                        2
11 leaf_N_resorption                           14
12 leaf_cell_wall_N_per_cell_wall_dry_mass     22
13 leaf_cell_wall_N_per_leaf_N                 22
14 root_N_per_dry_mass                         22
15 leaf_rubisco_N_per_total_leaf_N             29
16 leaf_thylakoid_N_per_total_leaf_N           29</code></pre>
</div>
</div>
<p>Or look at how many datasets have each trait:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">lookup_trait</span>(austraits, <span class="st">"_N_"</span>) <span class="ot">-&gt;</span> N_traits</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(trait_name <span class="sc">%in%</span> N_traits) <span class="sc">%&gt;%</span> </span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(dataset_id, trait_name) <span class="sc">%&gt;%</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(trait_name) <span class="sc">%&gt;%</span> </span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">n_dataset =</span> <span class="fu">length</span>(trait_name)) <span class="sc">%&gt;%</span> </span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(trait_name, n_dataset)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 16 × 2
   trait_name                              n_dataset
   &lt;chr&gt;                                       &lt;int&gt;
 1 leaf_N_per_dry_mass                            87
 2 leaf_N_per_area                                30
 3 wood_N_per_dry_mass                             8
 4 leaf_senesced_N_per_dry_mass                    9
 5 bark_N_per_dry_mass                             5
 6 fruit_N_per_dry_mass                            5
 7 seed_N_per_seed_dry_mass                        6
 8 stem_N_per_dry_mass                             3
 9 wood_dead_N_per_dry_mass                        3
10 flower_N_per_dry_mass                           2
11 leaf_N_resorption                               1
12 leaf_cell_wall_N_per_cell_wall_dry_mass         1
13 leaf_cell_wall_N_per_leaf_N                     1
14 root_N_per_dry_mass                             4
15 leaf_rubisco_N_per_total_leaf_N                 1
16 leaf_thylakoid_N_per_total_leaf_N               1</code></pre>
</div>
</div>
</section>
<section id="filtering_functions" class="level3">
<h3 class="anchored" data-anchor-id="filtering_functions">Filtering data</h3>
<p>There are a trio of austraits.R functions to filter data by <code>trait_name</code>, <code>taxon_name</code> or <code>dataset_id</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>leaf_mass_per_area_values <span class="ot">&lt;-</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> </span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_names =</span> <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>))</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>Westoby_2014_values <span class="ot">&lt;-</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="st">"Westoby_2014"</span>)</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>Eucalyptus_values <span class="ot">&lt;-</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">genus =</span> <span class="st">"Eucalyptus"</span>)</span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>Banksia_serrata_values <span class="ot">&lt;-</span></span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">genus =</span> <span class="st">"Banksia serrata"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The advantage of the <code>{austraits}</code> package functions over <code>dplyr::filter()</code> is that they simultaneously subset both the core traits table and all the ancillary data tables.</p>
<p>However, there are many columns within the traits tables that you cannot (currently) filter by using <code>{austraits}</code> functions.</p>
<p>If you wish to also subset data based on fields such as <code>basis_of_record</code> or <code>life_stage</code> you need to use <code>dplyr::filter()</code>. This allows you to pipe together multiple filtering calls, but only filters the traits table.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>my_data <span class="ot">&lt;-</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(stringr<span class="sc">::</span><span class="fu">str_detect</span>(taxon_name, <span class="st">"Eucalyptus"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(trait_name <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>, <span class="st">"leaf_length"</span>, <span class="st">"leaf_thickness"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(basis_of_record <span class="sc">==</span> <span class="st">"field"</span>)</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>my_data <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 5375</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>my_data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 26
   dataset_id      taxon_name  observation_id trait_name value unit  entity_type
   &lt;chr&gt;           &lt;chr&gt;       &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 Atkinson_2020_2 Eucalyptus… 35             leaf_mass… 189.… g/m2  individual 
 2 Atkinson_2020_2 Eucalyptus… 36             leaf_mass… 172.… g/m2  individual 
 3 Atkinson_2020_2 Eucalyptus… 37             leaf_mass… 138.… g/m2  individual 
 4 Atkinson_2020_2 Eucalyptus… 38             leaf_mass… 141.… g/m2  individual 
 5 Atkinson_2020_2 Eucalyptus… 39             leaf_mass… 172.… g/m2  individual 
 6 Atkinson_2020_2 Eucalyptus… 40             leaf_mass… 171.… g/m2  individual 
 7 Atkinson_2020_2 Eucalyptus… 41             leaf_mass… 168.… g/m2  individual 
 8 Atkinson_2020_2 Eucalyptus… 42             leaf_mass… 210.… g/m2  individual 
 9 Atkinson_2020_2 Eucalyptus… 43             leaf_mass… 178.… g/m2  individual 
10 Atkinson_2020_2 Eucalyptus… 44             leaf_mass… 226.… g/m2  individual 
# ℹ 19 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;,
#   replicates &lt;chr&gt;, basis_of_record &lt;chr&gt;, life_stage &lt;chr&gt;,
#   population_id &lt;chr&gt;, individual_id &lt;chr&gt;, repeat_measurements_id &lt;chr&gt;,
#   temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;, location_id &lt;chr&gt;,
#   entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;, treatment_context_id &lt;chr&gt;,
#   collection_date &lt;chr&gt;, measurement_remarks &lt;chr&gt;, method_id &lt;chr&gt;,
#   method_context_id &lt;chr&gt;, original_name &lt;chr&gt;</code></pre>
</div>
</div>
</section>
<section id="join_functions" class="level3">
<h3 class="anchored" data-anchor-id="join_functions">Joining relational tables</h3>
<p>For many research purposes you will want to join metadata from one of the relational tables to the core traits table. There are a series of <code>{austraits}</code> functions that facilitate this - or you can use <code>dplyr::left_join()</code> to join tables.</p>
<section id="joining-locations" class="level4">
<h4 class="anchored" data-anchor-id="joining-locations">Joining locations</h4>
<p>The locations table includes information on all location properties measured, including the actual location (latitude/longitude), climatic data, soil properties, fire history, vegetation history, and geologic history.</p>
<p>To just add location name, latitude, and longitude:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>traits_with_lat_long <span class="ot">&lt;-</span> </span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>traits_with_lat_long<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(dataset_id, taxon_name, trait_name, value, location_name, location_id, <span class="st">`</span><span class="at">latitude (deg)</span><span class="st">`</span>, <span class="st">`</span><span class="at">longitude (deg)</span><span class="st">`</span>) <span class="sc">%&gt;%</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(<span class="st">`</span><span class="at">latitude (deg)</span><span class="st">`</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 8
   dataset_id  taxon_name          trait_name    value location_name location_id
   &lt;chr&gt;       &lt;chr&gt;               &lt;chr&gt;         &lt;chr&gt; &lt;chr&gt;         &lt;chr&gt;      
 1 Ahrens_2019 Corymbia calophylla leaf_N_per_d… 9     BOO           01         
 2 Ahrens_2019 Corymbia calophylla leaf_area     1288… BOO           01         
 3 Ahrens_2019 Corymbia calophylla leaf_delta13C -27.… BOO           01         
 4 Ahrens_2019 Corymbia calophylla leaf_mass_pe… 185.… BOO           01         
 5 Ahrens_2019 Corymbia calophylla leaf_photoch… 0.02… BOO           01         
 6 Ahrens_2019 Corymbia calophylla leaf_water_b… 1.06… BOO           01         
 7 Ahrens_2019 Corymbia calophylla modified_NDVI 0.71… BOO           01         
 8 Ahrens_2019 Corymbia calophylla wood_density  0.60… BOO           01         
 9 Ahrens_2019 Corymbia calophylla leaf_N_per_d… 16    BOO           01         
10 Ahrens_2019 Corymbia calophylla leaf_area     9647… BOO           01         
11 Ahrens_2019 Corymbia calophylla leaf_delta13C -27.… BOO           01         
12 Ahrens_2019 Corymbia calophylla leaf_mass_pe… 138.… BOO           01         
13 Ahrens_2019 Corymbia calophylla leaf_photoch… 0.03… BOO           01         
14 Ahrens_2019 Corymbia calophylla leaf_water_b… 1.05… BOO           01         
15 Ahrens_2019 Corymbia calophylla modified_NDVI 0.77… BOO           01         
16 Ahrens_2019 Corymbia calophylla wood_density  0.48… BOO           01         
17 Ahrens_2019 Corymbia calophylla leaf_N_per_d… 10    BOO           01         
18 Ahrens_2019 Corymbia calophylla leaf_area     1250… BOO           01         
19 Ahrens_2019 Corymbia calophylla leaf_delta13C -27.… BOO           01         
20 Ahrens_2019 Corymbia calophylla leaf_mass_pe… 169.… BOO           01         
# ℹ 2 more variables: `latitude (deg)` &lt;chr&gt;, `longitude (deg)` &lt;chr&gt;</code></pre>
</div>
</div>
<p>However, the <code>{austraits}</code> function <code>join_locations</code> only joins latitude and longitude data from the locations tables.</p>
<p>To join all location properties you need to use <code>tidyr::pivot_wider</code> and <code>dplyr::left_join</code>, using the code in steps (2) and (3) below.</p>
<p>Moreover, as there are currently 202 location properties, you will want to first filter the locations properties, to tailor your join request to the exact information relevant to your study.</p>
<ol type="1">
<li>Select specific location properties to merge onto the traits table:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>locations <span class="sc">%&gt;%</span> </span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(location_property)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 202 × 1
   location_property      
   &lt;chr&gt;                  
 1 latitude (deg)         
 2 longitude (deg)        
 3 aridity index (MAP/PET)
 4 precipitation, MAP (mm)
 5 temperature, max (C)   
 6 locality               
 7 recorded by            
 8 description            
 9 notes                  
10 soil type              
# ℹ 192 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>locations <span class="ot">&lt;-</span> </span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>locations <span class="sc">%&gt;%</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(location_property <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"latitude (deg)"</span>, <span class="st">"longitude (deg)"</span>, <span class="st">"description"</span>, <span class="st">"aridity index (MAP/PET)"</span>, <span class="st">"soil type"</span>, <span class="st">"fire history"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>The data from the locations table must be pivoted into a wide-format table to join to the traits table.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>locations <span class="ot">&lt;-</span> </span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  locations <span class="sc">%&gt;%</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> location_property, <span class="at">values_from =</span> value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="3" type="1">
<li>You can now now join this table of location properties onto the traits table.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>traits_with_locations <span class="ot">&lt;-</span> </span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(locations, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"dataset_id"</span>, <span class="st">"location_id"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="joining-contexts" class="level4">
<h4 class="anchored" data-anchor-id="joining-contexts">Joining contexts</h4>
<p>The context tables documents additional context properties/ancillary data may be useful for interpreting trait values. Context properties are divided into 5 categories, <code>treatment context</code>, <code>plot context</code>, <code>entity context</code>, <code>temporal context</code>, and <code>method context</code>.</p>
<table class="table">
<colgroup>
<col style="width: 20%">
<col style="width: 79%">
</colgroup>
<thead>
<tr class="header">
<th>context category</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>treatment context</td>
<td>Context property that is an experimental manipulation that might affect the trait values measured on an individual, population or species-level entity.</td>
</tr>
<tr class="even">
<td>plot context</td>
<td>Context property that is a feature of a plot (subset of a location) that might affect the trait values measured on an individual, population or species-level entity.</td>
</tr>
<tr class="odd">
<td>entity context</td>
<td>Context property that is information about an organismal entity (individual, population or taxon) that does not comprise a trait-centered observation but might affect the trait values measured on the entity.</td>
</tr>
<tr class="even">
<td>temporal context</td>
<td>Context property that is a feature of a “point in time” that might affect the trait values measured on an individual, population or species-level entity.</td>
</tr>
<tr class="odd">
<td>method context</td>
<td>Context property that records specific information about a measurement method that is modified between measurements.</td>
</tr>
</tbody>
</table>
<p>Joining the context table to the traits table is best done using the <code>{austraits}</code> function, as there are 5 identifiers in the traits table (<code>entity_context_id</code>, <code>method_context_id</code>, <code>temporal_context_id</code>, <code>plot_context_id</code> and <code>treatment_context_id</code>) that link to two columns in the contexts table (<code>link_vals</code> and <code>link_id</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>traits_with_contexts <span class="ot">&lt;-</span> </span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_contexts</span>()</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>traits_with_contexts<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>dplyr<span class="sc">::</span><span class="fu">all_of</span>(<span class="fu">c</span>(<span class="st">"temporal_context_id"</span>, <span class="st">"method_context_id"</span>, <span class="st">"treatment_context_id"</span>, <span class="st">"entity_context_id"</span>, <span class="st">"plot_context_id"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,813,897 × 103
   dataset_id taxon_name       observation_id trait_name value unit  entity_type
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 ABRS_1981  Acanthocarpus c… 0001           leaf_comp… simp… &lt;NA&gt;  species    
 2 ABRS_1981  Acanthocarpus c… 0001           leaf_leng… 3     mm    species    
 3 ABRS_1981  Acanthocarpus c… 0001           leaf_leng… 15    mm    species    
 4 ABRS_1981  Acanthocarpus c… 0001           seed_heig… 3     mm    species    
 5 ABRS_1981  Acanthocarpus c… 0001           seed_leng… 3     mm    species    
 6 ABRS_1981  Acanthocarpus c… 0001           seed_width 3     mm    species    
 7 ABRS_1981  Acanthocarpus h… 0002           leaf_comp… simp… &lt;NA&gt;  species    
 8 ABRS_1981  Acanthocarpus h… 0002           leaf_leng… 4     mm    species    
 9 ABRS_1981  Acanthocarpus h… 0002           leaf_leng… 12    mm    species    
10 ABRS_1981  Acanthocarpus h… 0002           seed_leng… 4     mm    species    
# ℹ 1,813,887 more rows
# ℹ 96 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;,
#   replicates &lt;chr&gt;, basis_of_record &lt;chr&gt;, life_stage &lt;chr&gt;,
#   population_id &lt;chr&gt;, individual_id &lt;chr&gt;, repeat_measurements_id &lt;chr&gt;,
#   source_id &lt;chr&gt;, location_id &lt;chr&gt;, collection_date &lt;chr&gt;,
#   measurement_remarks &lt;chr&gt;, method_id &lt;chr&gt;, original_name &lt;chr&gt;,
#   `trait scoring method` &lt;chr&gt;, entity_measured &lt;chr&gt;, …</code></pre>
</div>
</div>
<p>This function adds columns to the traits table for all context properties. You may want to instead include fewer contexts.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>contexts <span class="sc">%&gt;%</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(context_property, category)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 83 × 2
   context_property       category         
   &lt;chr&gt;                  &lt;chr&gt;            
 1 trait scoring method   method_context   
 2 entity_measured        method_context   
 3 replicate observations temporal_context 
 4 sampling season        temporal_context 
 5 measurement CO2        method_context   
 6 seed provenance        treatment_context
 7 measurement irradiance method_context   
 8 repeat observations    temporal_context 
 9 habitat flammability   plot_context     
10 time to flowering type entity_context   
# ℹ 73 more rows</code></pre>
</div>
</div>
<p>Select all austraits$traits columns (26 currently) together with the context properties you want to keep.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>traits_with_contexts<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(austraits<span class="sc">$</span>traits),  <span class="fu">c</span>(<span class="st">"sampling season"</span>, <span class="st">"plant sex"</span>, <span class="st">"leaf surface"</span>, <span class="st">"leaf age"</span>, <span class="st">"fire intensity"</span>, <span class="st">"slope position"</span>, <span class="st">"fire season"</span>, <span class="st">"drought treatment"</span>, <span class="st">"temperature treatment"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,813,897 × 35
   dataset_id taxon_name       observation_id trait_name value unit  entity_type
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 ABRS_1981  Acanthocarpus c… 0001           leaf_comp… simp… &lt;NA&gt;  species    
 2 ABRS_1981  Acanthocarpus c… 0001           leaf_leng… 3     mm    species    
 3 ABRS_1981  Acanthocarpus c… 0001           leaf_leng… 15    mm    species    
 4 ABRS_1981  Acanthocarpus c… 0001           seed_heig… 3     mm    species    
 5 ABRS_1981  Acanthocarpus c… 0001           seed_leng… 3     mm    species    
 6 ABRS_1981  Acanthocarpus c… 0001           seed_width 3     mm    species    
 7 ABRS_1981  Acanthocarpus h… 0002           leaf_comp… simp… &lt;NA&gt;  species    
 8 ABRS_1981  Acanthocarpus h… 0002           leaf_leng… 4     mm    species    
 9 ABRS_1981  Acanthocarpus h… 0002           leaf_leng… 12    mm    species    
10 ABRS_1981  Acanthocarpus h… 0002           seed_leng… 4     mm    species    
# ℹ 1,813,887 more rows
# ℹ 28 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;,
#   replicates &lt;chr&gt;, basis_of_record &lt;chr&gt;, life_stage &lt;chr&gt;,
#   population_id &lt;chr&gt;, individual_id &lt;chr&gt;, repeat_measurements_id &lt;chr&gt;,
#   temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;, location_id &lt;chr&gt;,
#   entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;, treatment_context_id &lt;chr&gt;,
#   collection_date &lt;chr&gt;, measurement_remarks &lt;chr&gt;, method_id &lt;chr&gt;, …</code></pre>
</div>
</div>
</section>
<section id="joining-methods" class="level4">
<h4 class="anchored" data-anchor-id="joining-methods">Joining methods</h4>
<p>The methods table documents a selection of metadata recorded about the entire dataset and methods used for individual trait measurements. There is a single row of data per <code>dataset_id</code> x <code>trait_name</code> x <code>method_id</code> combination. <code>Method_id</code> is used to distinguish between instances where a single trait is measured twice using two separate protocols and is separate to <code>method_context_id</code> which documents specific components of a methods that are modified between measurements.</p>
<p>The <code>{austraits}</code> function offers a function to just add trait methods to the traits table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>traits_with_methods <span class="ot">&lt;-</span> </span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_methods</span>()</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>traits_with_methods<span class="sc">$</span>traits</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,813,897 × 27
   dataset_id taxon_name       observation_id trait_name value unit  entity_type
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 ABRS_1981  Acanthocarpus c… 0001           leaf_comp… simp… &lt;NA&gt;  species    
 2 ABRS_1981  Acanthocarpus c… 0001           leaf_leng… 3     mm    species    
 3 ABRS_1981  Acanthocarpus c… 0001           leaf_leng… 15    mm    species    
 4 ABRS_1981  Acanthocarpus c… 0001           seed_heig… 3     mm    species    
 5 ABRS_1981  Acanthocarpus c… 0001           seed_leng… 3     mm    species    
 6 ABRS_1981  Acanthocarpus c… 0001           seed_width 3     mm    species    
 7 ABRS_1981  Acanthocarpus h… 0002           leaf_comp… simp… &lt;NA&gt;  species    
 8 ABRS_1981  Acanthocarpus h… 0002           leaf_leng… 4     mm    species    
 9 ABRS_1981  Acanthocarpus h… 0002           leaf_leng… 12    mm    species    
10 ABRS_1981  Acanthocarpus h… 0002           seed_leng… 4     mm    species    
# ℹ 1,813,887 more rows
# ℹ 20 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;,
#   replicates &lt;chr&gt;, basis_of_record &lt;chr&gt;, life_stage &lt;chr&gt;,
#   population_id &lt;chr&gt;, individual_id &lt;chr&gt;, repeat_measurements_id &lt;chr&gt;,
#   temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;, location_id &lt;chr&gt;,
#   entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;, treatment_context_id &lt;chr&gt;,
#   collection_date &lt;chr&gt;, measurement_remarks &lt;chr&gt;, method_id &lt;chr&gt;, …</code></pre>
</div>
</div>
<p>To add all information within the methods table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>traits_with_all_methods <span class="ot">&lt;-</span> </span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(austraits<span class="sc">$</span>methods, </span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"dataset_id"</span>, <span class="st">"trait_name"</span>, <span class="st">"method_id"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As there are quite a few columns in the method table with lengthy text fields, it is worth first viewing the methods table and confirming you want information like <code>sampling_strategy</code> merged into the traits table.</p>
<p>Check the schema file embedded within AusTraits to see what information is documented in each column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>schema<span class="sc">$</span>austraits<span class="sc">$</span>elements<span class="sc">$</span>methods<span class="sc">$</span>elements <span class="sc">%&gt;%</span> </span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>  traits.build<span class="sc">::</span><span class="fu">util_list_to_df1</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 15 × 2
   key                              value                                       
   &lt;chr&gt;                            &lt;chr&gt;                                       
 1 dataset_id                       Primary identifier for each study contribut…
 2 trait_name                       Name of the trait sampled. Allowable values…
 3 methods                          A textual description of the methods used t…
 4 method_id                        A unique integer identifier to distinguish …
 5 description                      A 1-2 sentence description of the purpose o…
 6 sampling_strategy                A written description of how study location…
 7 source_primary_key               Citation key for the primary source in `sou…
 8 source_primary_citation          Citation for the primary source. This detai…
 9 source_secondary_key             Citation key for the secondary source in `s…
10 source_secondary_citation        Citations for the secondary source. This de…
11 source_original_dataset_key      Citation key for the original dataset_id in…
12 source_original_dataset_citation Citations for the original dataset_id in so…
13 data_collectors                  The person (people) leading data collection…
14 assistants                       Names of additional people who played a mor…
15 dataset_curators                 Names of AusTraits team member(s) who conta…</code></pre>
</div>
</div>
<p>Then just join the columns of interest:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>traits_with_all_methods <span class="ot">&lt;-</span> </span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(austraits<span class="sc">$</span>methods <span class="sc">%&gt;%</span> </span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>              dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="st">"dataset_id"</span>, <span class="st">"trait_name"</span>, <span class="st">"methods"</span>, <span class="st">"method_id"</span>, <span class="st">"description"</span>,  <span class="st">"source_primary_citation"</span>, <span class="st">"data_collectors"</span>),</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"dataset_id"</span>, <span class="st">"trait_name"</span>, <span class="st">"method_id"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="joining-taxa" class="level4">
<h4 class="anchored" data-anchor-id="joining-taxa">Joining taxa</h4>
<p>The taxon table documents a collection of names and identifiers for each taxon. Within AusTraits <code>names</code> submitted as identifiers within a dataset might be resolved to species, an infraspecific taxon, or sometimes just to genus- or family-level resolution; this information is recorded as the <code>taxon_rank</code>. The <code>taxon_rank</code> determines which information is filled in in the taxon table.</p>
<p>The <code>{austraits}</code> function makes it possible to quickly merge in family, genus, taxon_rank and establishment_means:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>austraits_with_taxa <span class="ot">&lt;-</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> austraits<span class="sc">::</span><span class="fu">join_taxonomy</span>()</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>austraits_with_taxa<span class="sc">$</span>traits</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,813,897 × 30
   dataset_id taxon_name       observation_id trait_name value unit  entity_type
   &lt;chr&gt;      &lt;chr&gt;            &lt;chr&gt;          &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
 1 ABRS_1981  Acanthocarpus c… 0001           leaf_comp… simp… &lt;NA&gt;  species    
 2 ABRS_1981  Acanthocarpus c… 0001           leaf_leng… 3     mm    species    
 3 ABRS_1981  Acanthocarpus c… 0001           leaf_leng… 15    mm    species    
 4 ABRS_1981  Acanthocarpus c… 0001           seed_heig… 3     mm    species    
 5 ABRS_1981  Acanthocarpus c… 0001           seed_leng… 3     mm    species    
 6 ABRS_1981  Acanthocarpus c… 0001           seed_width 3     mm    species    
 7 ABRS_1981  Acanthocarpus h… 0002           leaf_comp… simp… &lt;NA&gt;  species    
 8 ABRS_1981  Acanthocarpus h… 0002           leaf_leng… 4     mm    species    
 9 ABRS_1981  Acanthocarpus h… 0002           leaf_leng… 12    mm    species    
10 ABRS_1981  Acanthocarpus h… 0002           seed_leng… 4     mm    species    
# ℹ 1,813,887 more rows
# ℹ 23 more variables: value_type &lt;chr&gt;, basis_of_value &lt;chr&gt;,
#   replicates &lt;chr&gt;, basis_of_record &lt;chr&gt;, life_stage &lt;chr&gt;,
#   population_id &lt;chr&gt;, individual_id &lt;chr&gt;, repeat_measurements_id &lt;chr&gt;,
#   temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;, location_id &lt;chr&gt;,
#   entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;, treatment_context_id &lt;chr&gt;,
#   collection_date &lt;chr&gt;, measurement_remarks &lt;chr&gt;, method_id &lt;chr&gt;, …</code></pre>
</div>
</div>
<p>To add all columns from austraits$taxa:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>austraits_with_taxa <span class="ot">&lt;-</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">left_join</span>(austraits<span class="sc">$</span>taxa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
</section>
<section id="summarising-data-trait-means-modes-etc." class="level2" data-number="29.6">
<h2 data-number="29.6" class="anchored" data-anchor-id="summarising-data-trait-means-modes-etc."><span class="header-section-number">29.6</span> Summarising data: trait means, modes, etc.</h2>
<p>If you’ve sourced the file <code>extra_functions.R</code>, there are a few functions that allow you to summarise trait values.</p>
<section id="categorical-traits" class="level3">
<h3 class="anchored" data-anchor-id="categorical-traits">Categorical traits</h3>
<p>For instance, <code>categorical_summary</code> indicates how many times a specific trait value is reported for a given taxa (across all datasets):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>cat_summary <span class="ot">&lt;-</span> <span class="fu">categorical_summary</span>(austraits, <span class="st">"resprouting_capacity"</span>)</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>cat_summary</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12,817 × 3
   taxon_name                             trait_name           value_summary  
   &lt;chr&gt;                                  &lt;chr&gt;                &lt;chr&gt;          
 1 Abelmoschus ficulneus                  resprouting_capacity fire_killed (4)
 2 Abelmoschus manihot                    resprouting_capacity fire_killed (3)
 3 Abelmoschus moschatus                  resprouting_capacity resprouts (1)  
 4 Abelmoschus moschatus subsp. tuberosus resprouting_capacity resprouts (3)  
 5 Abrodictyum caudatum                   resprouting_capacity fire_killed (2)
 6 Abrodictyum obscurum                   resprouting_capacity fire_killed (2)
 7 Abrotanella forsteroides               resprouting_capacity resprouts (1)  
 8 Abrotanella nivigena                   resprouting_capacity resprouts (1)  
 9 Abrotanella sp. [White_2020]           resprouting_capacity resprouts (1)  
10 Abrus precatorius                      resprouting_capacity resprouts (1)  
# ℹ 12,807 more rows</code></pre>
</div>
</div>
<p>Alternatively, create a wider matrix with possible trait values as columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>categorical_summary_wider <span class="ot">&lt;-</span> </span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">categorical_summary_by_value</span>(austraits, <span class="st">"resprouting_capacity"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> value_tmp, <span class="at">values_from =</span> replicates)</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>categorical_summary_wider</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12,817 × 5
   taxon_name               trait_name fire_killed resprouts partial_resprouting
   &lt;chr&gt;                    &lt;chr&gt;            &lt;dbl&gt;     &lt;dbl&gt;               &lt;dbl&gt;
 1 Abelmoschus ficulneus    resprouti…           4        NA                  NA
 2 Abelmoschus manihot      resprouti…           3        NA                  NA
 3 Abelmoschus moschatus    resprouti…          NA         1                  NA
 4 Abelmoschus moschatus s… resprouti…          NA         3                  NA
 5 Abrodictyum caudatum     resprouti…           2        NA                  NA
 6 Abrodictyum obscurum     resprouti…           2        NA                  NA
 7 Abrotanella forsteroides resprouti…          NA         1                  NA
 8 Abrotanella nivigena     resprouti…          NA         1                  NA
 9 Abrotanella sp. [White_… resprouti…          NA         1                  NA
10 Abrus precatorius        resprouti…          NA         1                  NA
# ℹ 12,807 more rows</code></pre>
</div>
</div>
</section>
<section id="numeric-traits" class="level3">
<h3 class="anchored" data-anchor-id="numeric-traits">Numeric traits</h3>
<p>We’re hesitant to write functions to summarise numeric traits, as this quickly requires statistical assumptions.</p>
<p>The datasets that comprise AusTraits were collected by different people, with different number of replicates, different entity types are reported - one dataset might include 20 measurements on individuals for a trait and another dataset might have submitted a single population-level mean derived from 5 measurements.</p>
<p>How do you take the mean of these trait values?</p>
<p>Do you want to include both data from experiments and plants growing under natural conditions? This information is recorded in the <code>basis_of_record</code> column.</p>
<p>The function <code>austraits::summarise_trait_means()</code> outputs <code>location</code> x <code>taxon_name</code> x <code>dataset_id</code> means, but is not (<em>currently</em>) appropriately updating/removing columns such as <code>observation_id</code>, <code>individual_id</code>, and <code>replicates</code>.</p>
<p>An alternative function we’re developing is:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>weighted <span class="ot">&lt;-</span> <span class="fu">austraits_weighted_means</span>(austraits, <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>, <span class="st">"leaf_length"</span>))</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>weighted</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 24,920 × 10
   taxon_name       trait_name  mean   min   max median geom_mean all_replicates
   &lt;chr&gt;            &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;     &lt;dbl&gt;          &lt;dbl&gt;
 1 Abelia x grandi… leaf_leng…   35     20    50     NA      35                1
 2 Abelmoschus fic… leaf_leng…  105     50   160     NA     105                1
 3 Abelmoschus man… leaf_leng…  232.   150   380     NA     230.               2
 4 Abelmoschus man… leaf_leng…  200    200   200     NA     200                1
 5 Abelmoschus mos… leaf_leng…  103.    27   150     NA      97.6              3
 6 Abelmoschus mos… leaf_leng…  150    150   150     NA     150                1
 7 Abildgaardia ov… leaf_leng…  230     60   400     NA     230                1
 8 Abrodictyum bra… leaf_leng…  143.    40   250     NA     137.               2
 9 Abrodictyum cau… leaf_leng…  134.    35   200     NA     129.               4
10 Abrodictyum obs… leaf_leng…  203.    20   400     NA     175.               6
# ℹ 24,910 more rows
# ℹ 2 more variables: location_replicates &lt;dbl&gt;, flora_replicates &lt;dbl&gt;</code></pre>
</div>
</div>
<p>This function first groups values at the site level, then at the taxon level. For trait data sourced from floras, where trait values are documented as a minimum and maximum value, it simply takes the mean of these as an additional mean.</p>
</section>
</section>
<section id="plotting_functions" class="level2" data-number="29.7">
<h2 data-number="29.7" class="anchored" data-anchor-id="plotting_functions"><span class="header-section-number">29.7</span> Plotting data</h2>
<section id="plotting-trait-distributions" class="level3">
<h3 class="anchored" data-anchor-id="plotting-trait-distributions">Plotting trait distributions</h3>
<p>Another way to summarise AusTraits data by trait and determine if AusTraits offers sufficient data coverage for a trait of choice, is to plot the distribution of trait values in AusTraits.</p>
<p>As seen in <a href="#exploring"><code>A first look at data</code></a>, the function <code>austraits::plot_trait_distribution_beeswarm()</code> plots trait data by dataset_id:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(trait_name <span class="sc">==</span> <span class="st">"leaf_N_per_dry_mass"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name) <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2230</code></pre>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_trait_distribution_beeswarm</span>(austraits, <span class="st">"leaf_N_per_dry_mass"</span>, <span class="at">y_axis_category =</span> <span class="st">"dataset_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Orientation inferred to be along y-axis; override with
`position_quasirandom(orientation = 'x')`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="plotting-data-distribution-by-location" class="level3">
<h3 class="anchored" data-anchor-id="plotting-data-distribution-by-location">Plotting data distribution by location</h3>
<p>To plot locations, you need to begin by merging on the latitude &amp; longitude data from austraits$locations.</p>
<p>The <code>plot_locations</code> function plots the selected data, separating data into a series of plots based on the variable name selected. You can separate data based on the values of <strong>any</strong> column within the traits table - including <code>basis_of_record</code>, <code>life_stage</code>, <code>value_type</code> - or higher taxon categories (<code>genus</code>, <code>family</code>).</p>
<p>For instance, <code>austraits::plot_locations("trait_name")</code> will output a separate plot for each <code>trait_name</code> within the selected data.</p>
<p>A warning: <code>austraits::plot_locations()</code> WILL BE VERY SLOW if you request more than ~20 plots. For instance, do not attempt to generate plots for all traits simultaneously. Always first use extract/filter to just select a narrow range of traits, datasets, or taxa.</p>
<section id="plot-locations-by-trait" class="level4">
<h4 class="anchored" data-anchor-id="plot-locations-by-trait">Plot locations by trait</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span> </span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_names =</span> <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(<span class="st">`</span><span class="at">latitude (deg)</span><span class="st">`</span>)) <span class="sc">%&gt;%</span>        <span class="co"># reduce the size of the dataframe by only retaining values with geographic coordinates</span></span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"trait_name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># or</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_trait</span>(<span class="at">trait_names =</span> <span class="fu">c</span>(<span class="st">"leaf_mass_per_area"</span>, <span class="st">"plant_height"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(<span class="st">`</span><span class="at">latitude (deg)</span><span class="st">`</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"trait_name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-34-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="plot-locations-by-study" class="level4">
<h4 class="anchored" data-anchor-id="plot-locations-by-study">Plot locations by study</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="st">"Westoby_2014"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb78-7"><a href="#cb78-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(<span class="st">`</span><span class="at">latitude (deg)</span><span class="st">`</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb78-8"><a href="#cb78-8" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"dataset_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="st">"Schulze_2014"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>     austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(<span class="st">`</span><span class="at">latitude (deg)</span><span class="st">`</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb79-8"><a href="#cb79-8" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">plot_locations</span>(<span class="st">"dataset_id"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-35-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="plot-locations-by-taxon" class="level4">
<h4 class="anchored" data-anchor-id="plot-locations-by-taxon">Plot locations by taxon</h4>
<p><code>plot_locations</code> lets you plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">extract_taxa</span>(<span class="at">taxon_name =</span> <span class="st">"Acacia aneura"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>  austraits<span class="sc">::</span><span class="fu">join_locations</span>()</span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>traits <span class="ot">&lt;-</span> data<span class="sc">$</span>traits <span class="sc">%&gt;%</span> </span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(<span class="st">`</span><span class="at">latitude (deg)</span><span class="st">`</span>)) </span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_locations</span>(data, <span class="st">"taxon_name"</span>)     <span class="co"># 1 taxon, so 1 plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">::</span><span class="fu">plot_locations</span>(data, <span class="st">"dataset_id"</span>)     <span class="co"># 1 plot for each dataset_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="AusTraits_tutorial_files/figure-html/unnamed-chunk-36-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="more-complex-workflows---some-examples" class="level3">
<h3 class="anchored" data-anchor-id="more-complex-workflows---some-examples">More complex workflows - some examples</h3>
<section id="an-example-looking-at-trait-climate-gradients" class="level4">
<h4 class="anchored" data-anchor-id="an-example-looking-at-trait-climate-gradients">An example looking at trait-climate gradients</h4>
<p>A simple workflow allows one to look at <a href="traits_and_climate_example.html">trait values across a climate gradient</a></p>
</section>
<section id="an-example-incorporating-ala-distribution-data" class="level4">
<h4 class="anchored" data-anchor-id="an-example-incorporating-ala-distribution-data">An example incorporating ALA distribution data</h4>
<p>A recent tutorial posted by ALA shows how one can combine AusTraits trait data and ALA spatial occurrence data:</p>
<p>https://labs.ala.org.au/posts/2023-08-28_alternatives-to-box-plots/post.html</p>
<p>We’ve adopted it <a href="spatial_data_example.html">here</a>.</p>
</section>
</section>
</section>
<section id="a-complexity-pivoting-datasets" class="level2" data-number="29.8">
<h2 data-number="29.8" class="anchored" data-anchor-id="a-complexity-pivoting-datasets"><span class="header-section-number">29.8</span> A complexity: pivoting datasets</h2>
<p>The AusTraits tables are all in <code>long</code> format with an individual row for each trait measurement. This is the most compact way to store data and offers the flexibility of documenting diverse metadata for each trait measurement.</p>
<p>However, for many research uses, it may be more useful to view data in a <code>wide</code> format, with the multiple traits that comprise a single observation displayed as consecutive columns.</p>
<p>The <code>{austraits}</code> function <code>trait_pivot_wider</code> allows AusTraits datasets to be pivoted from <code>long</code> to <code>wide</code> format.</p>
<p>It is recommended to only use this function on individual datasets - or perhaps a small selection of datasets - as each dataset includes a different collection of traits and pivoting wider otherwise creates a very “holey” dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>Farrell_2017_values <span class="ot">&lt;-</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>    austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="st">"Farrell_2017"</span>)</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>Farrell_2017_pivoted <span class="ot">&lt;-</span> </span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>  Farrell_2017_values<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a>    austraits<span class="sc">::</span><span class="fu">trait_pivot_wider</span>()</span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a>Farrell_2017_pivoted</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 177 × 25
   dataset_id   taxon_name observation_id entity_type basis_of_record life_stage
   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;          &lt;chr&gt;       &lt;chr&gt;           &lt;chr&gt;     
 1 Farrell_2017 Arthropod… 001            individual  lab             adult     
 2 Farrell_2017 Arthropod… 002            individual  lab             adult     
 3 Farrell_2017 Arthropod… 003            individual  lab             adult     
 4 Farrell_2017 Arthropod… 004            individual  lab             adult     
 5 Farrell_2017 Arthropod… 005            individual  lab             adult     
 6 Farrell_2017 Arthropod… 006            individual  lab             adult     
 7 Farrell_2017 Arthropod… 007            individual  lab             adult     
 8 Farrell_2017 Arthropod… 008            individual  lab             adult     
 9 Farrell_2017 Arthropod… 009            individual  lab             adult     
10 Farrell_2017 Chrysocep… 010            individual  lab             adult     
# ℹ 167 more rows
# ℹ 19 more variables: population_id &lt;chr&gt;, individual_id &lt;chr&gt;,
#   repeat_measurements_id &lt;chr&gt;, temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;,
#   location_id &lt;chr&gt;, entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;,
#   treatment_context_id &lt;chr&gt;, collection_date &lt;chr&gt;, method_id &lt;chr&gt;,
#   method_context_id &lt;chr&gt;, original_name &lt;chr&gt;,
#   bulk_modulus_of_elasticity &lt;chr&gt;, …</code></pre>
</div>
</div>
<p>This example pivots “nicely” as all observations have <code>entity_type = individual</code>.</p>
<p>Compare this first example to the dataset <code>Edwards_2000</code> which includes individual-level, population-level, and species-level observations:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>Edwards_2000_values <span class="ot">&lt;-</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>  austraits <span class="sc">%&gt;%</span></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>    austraits<span class="sc">::</span><span class="fu">extract_dataset</span>(<span class="st">"Edwards_2000"</span>)</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>Edwards_2000_pivoted <span class="ot">&lt;-</span> </span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a>  Edwards_2000_values<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a>    austraits<span class="sc">::</span><span class="fu">trait_pivot_wider</span>()</span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-9"><a href="#cb84-9" aria-hidden="true" tabindex="-1"></a>Edwards_2000_pivoted</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 139 × 35
   dataset_id   taxon_name observation_id entity_type basis_of_record life_stage
   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;          &lt;chr&gt;       &lt;chr&gt;           &lt;chr&gt;     
 1 Edwards_2000 Acacia me… 001            individual  field           adult     
 2 Edwards_2000 Acacia me… 002            individual  field           adult     
 3 Edwards_2000 Acacia me… 003            individual  field           adult     
 4 Edwards_2000 Acacia me… 004            individual  field           adult     
 5 Edwards_2000 Acacia me… 005            individual  field           adult     
 6 Edwards_2000 Acacia me… 006            population  field           adult     
 7 Edwards_2000 Acacia me… 007            species     field           adult     
 8 Edwards_2000 Acacia su… 008            individual  field           adult     
 9 Edwards_2000 Acacia su… 009            individual  field           adult     
10 Edwards_2000 Acacia su… 010            individual  field           adult     
# ℹ 129 more rows
# ℹ 29 more variables: population_id &lt;chr&gt;, individual_id &lt;chr&gt;,
#   repeat_measurements_id &lt;chr&gt;, temporal_context_id &lt;chr&gt;, source_id &lt;chr&gt;,
#   location_id &lt;chr&gt;, entity_context_id &lt;chr&gt;, plot_context_id &lt;chr&gt;,
#   treatment_context_id &lt;chr&gt;, collection_date &lt;chr&gt;, method_id &lt;chr&gt;,
#   method_context_id &lt;chr&gt;, original_name &lt;chr&gt;, leaf_N_per_area &lt;chr&gt;,
#   leaf_N_per_dry_mass &lt;chr&gt;, leaf_area &lt;chr&gt;, leaf_mass_per_area &lt;chr&gt;, …</code></pre>
</div>
</div>
<p>The values at the individual, population and species level do not collapse together, because traits measured on different <code>entity_types</code> have separate <code>observation_id</code>’s.</p>
<p>One of the core identifiers assigned to data points is the <code>observation_id</code>. An observation is a collection of measurements made on a specific entity at a single point in time.</p>
<p>Observation_id’s are therefore unique combinations of:</p>
<ul>
<li>dataset_id</li>
<li>source_id</li>
<li>entity_type</li>
<li>taxon_name</li>
<li>population_id (location_id, plot_context_id, treatment_context_id)</li>
<li>individual_id</li>
<li>basis_of_record</li>
<li>entity_context_id</li>
<li>life_stage</li>
<li>temporal_context_id</li>
<li>collection_date</li>
<li>original_name</li>
</ul>
<p>If a single dataset includes traits that are attributed to different entity types, they are assigned separate <code>observation_id</code>’s. For instance, many datasets are comprised of individual-level physiological trait data and a column <code>growth_form</code>, documenting the growth form (i.e.&nbsp;tree, shrub, herb, etc.) of each <em>species</em>.</p>
<p>We’re developing a function that collapses the pivoted data into a more condensed table, but this loses some of the metadata:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>Edwards_2000_pivoted_merged <span class="ot">&lt;-</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">merge_entity_types</span>(<span class="st">"Edwards_2000"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><p>This function will duplicate any “higher-entity” trait values (e.g.&nbsp;A single species-level value is filled in for all individuals or populations)</p></li>
<li><p>Metadata fields like <code>entity_type</code>, <code>value_type</code> are only retained if it is identical for all measurements</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>Westoby_2014_pivoted_merged <span class="ot">&lt;-</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">merge_entity_types</span>(<span class="st">"Westoby_2014"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="intepreting-trait-names-taxon-names" class="level2" data-number="29.9">
<h2 data-number="29.9" class="anchored" data-anchor-id="intepreting-trait-names-taxon-names"><span class="header-section-number">29.9</span> Intepreting trait names, taxon names</h2>
<section id="trait-dictionary" class="level3">
<h3 class="anchored" data-anchor-id="trait-dictionary">Trait dictionary</h3>
<p>The AusTraits pipeline requires a trait dictionary that documents 4 pieces of information about each trait:</p>
<ul>
<li>trait name&nbsp;</li>
<li>trait type (categorical vs numeric)&nbsp;</li>
<li>allowable trait values (for categorical traits)&nbsp;</li>
<li>allowable trait range and units&nbsp;</li>
</ul>
<p>The trait dictionary embedded within AusTraits also has:</p>
<ul>
<li>trait labels&nbsp;</li>
<li>trait definitions&nbsp;</li>
<li>definitions for all categorical trait values&nbsp;</li>
</ul>
<p>Together these clarify each “trait concept”, which we define as: “a circumscribed set of trait measurements”. Much like a taxon concept delimits a collection of organisms, a trait concept delimits a collection of trait values pertaining to a distinct characteristic of a specific part of an organism (cell, tissue, organ, or whole organism).</p>
<p>The <a href="http:///w3id.org/APD">AusTraits Plant Dictionary (APD)</a> offers detailed descriptions for all trait concepts included in AusTraits. With the APD, each trait is given a unique, resolvable identifier, allowing trait definitions to be reused and shared.</p>
<p>The trait dictionary also includes:&nbsp;</p>
<ul>
<li>keywords&nbsp;</li>
<li>plant structure measured&nbsp;</li>
<li>characteristic measured&nbsp;</li>
<li>references&nbsp;</li>
<li>links to the same (or similar) trait concepts in other databases and dictionaries&nbsp;</li>
</ul>
</section>
<section id="understanding-taxon-names" class="level3">
<h3 class="anchored" data-anchor-id="understanding-taxon-names">Understanding taxon names</h3>
<p>AusTraits uses the taxon names in the Australian Plant Census (APC) and the Australian Plant Names Index (APNI).</p>
<p>The R package <a href="https://github.com/traitecoevo/APCalign"><code>{APCalign}</code></a> is used to align and update taxon names submitted to AusTraits with those in the APC/APNI.</p>
<p>APC align can also be installed directly from CRAN</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("APCalign")</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(APCalign)     <span class="co"># Australian plant taxon alignment function, available on CRAN </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are two key components to the workflow:&nbsp;</p>
<ol type="1">
<li>aligning names&nbsp;</li>
</ol>
<ul>
<li><p>syntax is standardised, including for phrase names&nbsp;</p></li>
<li><p>most spelling mistakes are corrected&nbsp;</p></li>
<li><p>names that indicate the plant can only be identified to genus are reformatted to <code>genus sp. [available notes; dataset_id]</code>&nbsp;</p>
<ol type="1">
<li><p>they are linked to an APC-accepted genus but not to an APC-accepted binomial.&nbsp;</p></li>
<li><p>they include the dataset_id so people don’t mistakenly group all <code>Eucalyptus sp.</code> as a single “species”&nbsp;</p></li>
</ol></li>
</ul>
<ol start="2" type="1">
<li>updating names&nbsp;</li>
</ol>
<ul>
<li>all aligned names that are in the APC, but that have a <code>taxonomic status</code> other than <code>accepted</code> are updated to their currently accepted name.&nbsp;</li>
</ul>
<p>Examples:&nbsp;</p>
<p>Identical <code>genus sp.</code> inputs from disparate datasets are given unique “names”:&nbsp;</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(stringr<span class="sc">::</span><span class="fu">str_detect</span>(original_name, <span class="st">"Eucalyptus sp</span><span class="sc">\\</span><span class="st">.$"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(dataset_id, taxon_name, original_name) <span class="sc">%&gt;%</span></span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(original_name <span class="sc">!=</span> taxon_name) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 3
  dataset_id     taxon_name                      original_name 
  &lt;chr&gt;          &lt;chr&gt;                           &lt;chr&gt;         
1 Clarke_2015    Eucalyptus sp. [Clarke_2015]    Eucalyptus sp.
2 Nolan_2022     Eucalyptus sp. [Nolan_2022]     Eucalyptus sp.
3 Soper_2014     Eucalyptus sp. [Soper_2014]     Eucalyptus sp.
4 Zieminska_2015 Eucalyptus sp. [Zieminska_2015] Eucalyptus sp.</code></pre>
</div>
</div>
<p>Outdated names are updated:&nbsp;</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(stringr<span class="sc">::</span><span class="fu">str_detect</span>(original_name, <span class="st">"Dryandra"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name, original_name) <span class="sc">%&gt;%</span></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(original_name <span class="sc">!=</span> taxon_name) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 15 × 2
   taxon_name                                 original_name                     
   &lt;chr&gt;                                      &lt;chr&gt;                             
 1 Banksia acanthopoda                        Dryandra acanthopoda              
 2 Banksia anatona                            Dryandra anatona                  
 3 Banksia arborea                            Dryandra arborea                  
 4 Banksia arctotidis                         Dryandra arctotidis               
 5 Banksia armata var. armata                 Dryandra armata var. armata       
 6 Banksia armata var. ignicida               Dryandra armata var. ignicida     
 7 Banksia armata                             Dryandra armata                   
 8 Banksia aurantia                           Dryandra aurantia                 
 9 Banksia biterax                            Dryandra baxteri                  
10 Banksia bipinnatifida subsp. bipinnatifida Dryandra bipinnatifida subsp. bip…
11 Banksia bipinnatifida subsp. multifida     Dryandra bipinnatifida subsp. mul…
12 Banksia bipinnatifida                      Dryandra bipinnatifida            
13 Banksia pellaeifolia                       Dryandra blechnifolia             
14 Banksia borealis subsp. borealis           Dryandra borealis subsp. borealis 
15 Banksia borealis subsp. elatior            Dryandra borealis subsp. elatior  </code></pre>
</div>
</div>
<p>Phrase name syntax across datasets is aligned:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>austraits<span class="sc">$</span>traits <span class="sc">%&gt;%</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(stringr<span class="sc">::</span><span class="fu">str_detect</span>(taxon_name, <span class="st">"Argyrodendron sp. Whyanbeel"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>(taxon_name, original_name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  taxon_name                                       original_name                
  &lt;chr&gt;                                            &lt;chr&gt;                        
1 Argyrodendron sp. Whyanbeel (B.P.Hyland RFK1106) Argyrodendron (Whyanbeel)    
2 Argyrodendron sp. Whyanbeel (B.P.Hyland RFK1106) Argyrodendron ssp. (Whyanbee…
3 Argyrodendron sp. Whyanbeel (B.P.Hyland RFK1106) Argyrodendron Whyanbeel      
4 Argyrodendron sp. Whyanbeel (B.P.Hyland RFK1106) Argyrodendron sp. (Whyanbeel…
5 Argyrodendron sp. Whyanbeel (B.P.Hyland RFK1106) Argyrodendron sp. Whyanbeel …
6 Argyrodendron sp. Whyanbeel (B.P.Hyland RFK1106) Argyrodendron sp. Whyanbeel …</code></pre>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./austraits_database.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">AusTraits, Australia’s Plant Trait Database</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./traits_and_climate_example.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">Combining trait values and climate data</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">How to get help: <a href="https://traitecoevo.github.io/traits.build-book/help.html" class="uri">https://traitecoevo.github.io/traits.build-book/help.html</a><br>Copyright 2023, Daniel Falster and Elizabeth Wenk</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>