<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>The {traits.build} data standard, R package, and workflow - 7&nbsp; Data structure</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./database_metadata.html" rel="next">
<link href="./database_standard.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./long_wide.html">Data structure and standard</a></li><li class="breadcrumb-item"><a href="./database_structure.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Data structure</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">The {traits.build} data standard, R package, and workflow</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Introduction</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">About</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./motivation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Motivation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./usage_examples.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Usage examples</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Data structure and standard</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./long_wide.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Long vs Wide data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_standard.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data standard</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_structure.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Data structure</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database_metadata.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Overview of traits.build ontology</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dictionary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Trait dictionary</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Creating with <code>traits.build</code></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./traits_build.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">The package</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_compilation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Tutorial: Example compilation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./file_organisation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">File organisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./create_dictionary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Creating a dictionary</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./adding_data_brief.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Adding datasets (brief)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./custom_R_code.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Adding custom_R_code</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./publishing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Publishing</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Guide to adding data</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./adding_data_long.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Adding datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_common_issues.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Common issues</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Tutorial: Adding datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Tutorial 1: Adding a simple dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Tutorial 2: Adding a more complex dataset</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Tutorial 3: Adding contexts and complex units</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Tutorial 4: Additional complexities</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Tutorial 5: Multiple columns for a trait</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_6.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">Tutorial 6: Data with repeat measurements</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial_dataset_7.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">Tutorial 7: Adding long format dataset</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Using outputs of <code>traits.build</code></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./austraits_database.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">AusTraits, Australia’s Plant Trait Database</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./austraits_package.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">The <code>austraits</code> package</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./AusTraits_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">AusTraits tutorial</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./traits_and_climate_example.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">Combining trait values and climate data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spatial_data_example.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">Tutorial: Using AusTraits with spatial data</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Getting help</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./help.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">Getting help</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./debugging.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">Debugging</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./csv.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">CSV files</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./yaml.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Yaml files</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#components" id="toc-components" class="nav-link active" data-scroll-target="#components"><span class="header-section-number">8</span> Components</a>
  <ul class="collapse">
  <li><a href="#traits" id="toc-traits" class="nav-link" data-scroll-target="#traits"><span class="header-section-number">8.1</span> traits</a>
  <ul class="collapse">
  <li><a href="#entity-type" id="toc-entity-type" class="nav-link" data-scroll-target="#entity-type">Entity type</a></li>
  <li><a href="#identifiers" id="toc-identifiers" class="nav-link" data-scroll-target="#identifiers">Identifiers</a></li>
  <li><a href="#life-stage-basis-of-record" id="toc-life-stage-basis-of-record" class="nav-link" data-scroll-target="#life-stage-basis-of-record">Life stage, basis of record</a></li>
  <li><a href="#value_types" id="toc-value_types" class="nav-link" data-scroll-target="#value_types">Values, Value types, basis of value</a></li>
  </ul></li>
  <li><a href="#locations" id="toc-locations" class="nav-link" data-scroll-target="#locations"><span class="header-section-number">8.2</span> locations</a></li>
  <li><a href="#contexts" id="toc-contexts" class="nav-link" data-scroll-target="#contexts"><span class="header-section-number">8.3</span> contexts</a></li>
  <li><a href="#methods" id="toc-methods" class="nav-link" data-scroll-target="#methods"><span class="header-section-number">8.4</span> methods</a></li>
  <li><a href="#excluded_data" id="toc-excluded_data" class="nav-link" data-scroll-target="#excluded_data"><span class="header-section-number">8.5</span> Excluded_data</a></li>
  <li><a href="#taxa" id="toc-taxa" class="nav-link" data-scroll-target="#taxa"><span class="header-section-number">8.6</span> taxa</a></li>
  <li><a href="#taxonomic_updates" id="toc-taxonomic_updates" class="nav-link" data-scroll-target="#taxonomic_updates"><span class="header-section-number">8.7</span> taxonomic_updates</a></li>
  <li><a href="#definitions" id="toc-definitions" class="nav-link" data-scroll-target="#definitions"><span class="header-section-number">8.8</span> definitions</a></li>
  <li><a href="#contributors" id="toc-contributors" class="nav-link" data-scroll-target="#contributors"><span class="header-section-number">8.9</span> contributors</a></li>
  <li><a href="#sources" id="toc-sources" class="nav-link" data-scroll-target="#sources"><span class="header-section-number">8.10</span> sources</a></li>
  <li><a href="#metadata" id="toc-metadata" class="nav-link" data-scroll-target="#metadata"><span class="header-section-number">8.11</span> metadata</a></li>
  <li><a href="#build_info" id="toc-build_info" class="nav-link" data-scroll-target="#build_info"><span class="header-section-number">8.12</span> build_info</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Data structure</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>This chapter describes the structure of the output of a <code>traits.build</code> compilation.</p>
<p>Note that the information provided below is based on the information provided within the file <code>traits.build_schema.yml</code>, which can be accessed by running <code>get_schema</code> or <code>system.file("support", "traits.build_schema.yml", package = "traits.build")</code>.</p>
<p>A <code>traits.build</code> compilation results in a series of linked components, which cross link against each other::</p>
<pre><code>austraits
├── traits
├── locations
├── contexts
├── methods
├── excluded_data
├── taxonomic_updates
├── taxa
├── contributors
├── sources
├── definitions
├── schema
├── metadata
└── build_info</code></pre>
<p>These include all the data and contextual information submitted with each contributed dataset.</p>
<section id="components" class="level1" data-number="8">
<h1 data-number="8"><span class="header-section-number">8</span> Components</h1>
<p>The core components are defined as follows.</p>
<section id="traits" class="level2" data-number="8.1">
<h2 data-number="8.1" class="anchored" data-anchor-id="traits"><span class="header-section-number">8.1</span> traits</h2>
<p><strong>Description:</strong> A table containing measurements of traits.</p>
<p><strong>Content:</strong></p>

<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead>
<tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often these are scientific papers, books, or online resources. By default this should be the name of the first author and year of publication, e.g.&nbsp;<code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_name
</td>
<td style="text-align:left;">
Scientific name of the taxon on which traits were sampled, without authorship. When possible, this is the currently accepted (botanical) or valid (zoological) scientific name, but might also be a higher taxonomic level.
</td>
</tr>
<tr>
<td style="text-align:left;">
observation_id
</td>
<td style="text-align:left;">
A unique integral identifier for the observation, where an observation is all measurements made on an individual at a single point in time. It is important for joining traits coming from the same <code>observation_id</code>. Within each dataset, observation_id’s are unique combinations of <code>taxon_name</code>, <code>population_id</code>, <code>individual_id</code>, and <code>temporal_context_id</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
trait_name
</td>
<td style="text-align:left;">
Name of the trait sampled. Allowable values specified in the table <code>definitions</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
value
</td>
<td style="text-align:left;">
The measured value of a trait, location property or context property.
</td>
</tr>
<tr>
<td style="text-align:left;">
unit
</td>
<td style="text-align:left;">
Units of the sampled trait value after aligning with AusTraits standards.
</td>
</tr>
<tr>
<td style="text-align:left;">
entity_type
</td>
<td style="text-align:left;">
A categorical variable specifying the entity corresponding to the trait values recorded.
</td>
</tr>
<tr>
<td style="text-align:left;">
value_type
</td>
<td style="text-align:left;">
A categorical variable describing the statistical nature of the trait value recorded.
</td>
</tr>
<tr>
<td style="text-align:left;">
basis_of_value
</td>
<td style="text-align:left;">
A categorical variable describing how the trait value was obtained.
</td>
</tr>
<tr>
<td style="text-align:left;">
replicates
</td>
<td style="text-align:left;">
Number of replicate measurements that comprise a recorded trait measurement. A numeric value (or range) is ideal and appropriate if the value type is a <code>mean</code>, <code>median</code>, <code>min</code> or <code>max</code>. For these value types, if replication is unknown the entry should be <code>unknown</code>. If the value type is <code>raw_value</code> the replicate value should be 1. If the trait is categorical or the value indicates a measurement for an entire species (or other taxon) replicate value should be <code>.na</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
basis_of_record
</td>
<td style="text-align:left;">
A categorical variable specifying from which kind of specimen traits were recorded.
</td>
</tr>
<tr>
<td style="text-align:left;">
life_stage
</td>
<td style="text-align:left;">
A field to indicate the life stage or age class of the entity measured. Standard values are <code>adult</code>, <code>sapling</code>, <code>seedling</code> and <code>juvenile</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
population_id
</td>
<td style="text-align:left;">
A unique integer identifier for a population, where a population is defined as individuals growing in the same location (location_id /location_name) and plot (plot_context_id, a context category) and being subjected to the same treatment (treatment_context_id, a context category).
</td>
</tr>
<tr>
<td style="text-align:left;">
individual_id
</td>
<td style="text-align:left;">
A unique integer identifier for an individual, with individuals numbered sequentially within each dataset by taxon by population grouping. Most often each row of data represents an individual, but in some datasets trait data collected on a single individual is presented across multiple rows of data, such as if the same trait is measured using different methods or the same individual is measured repeatedly across time.
</td>
</tr>
<tr>
<td style="text-align:left;">
repeat_measurements_id
</td>
<td style="text-align:left;">
A unique integer identifier for repeat measurements of a trait that comprise a single observation, such as a response curve.
</td>
</tr>
<tr>
<td style="text-align:left;">
temporal_context_id
</td>
<td style="text-align:left;">
A unique integer identifier assigned where repeat observations are made on the same individual (or population, or taxon) across time. The identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_id
</td>
<td style="text-align:left;">
For datasets that are compilations, an identifier for the original data source.
</td>
</tr>
<tr>
<td style="text-align:left;">
location_id
</td>
<td style="text-align:left;">
A unique integer identifier for a location, with locations numbered sequentially within a dataset. The identifier links to specific information in the location table.
</td>
</tr>
<tr>
<td style="text-align:left;">
entity_context_id
</td>
<td style="text-align:left;">
A unique integer identifier indicating specific contextual properties of an individual, possibly including the individual’s sex or caste (for social insects).
</td>
</tr>
<tr>
<td style="text-align:left;">
plot_context_id
</td>
<td style="text-align:left;">
A unique integer identifier for a plot, where a plot is a distinct collection of organisms within a single geographic location, such as plants growing on different aspects or blocks in an experiment. The identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
treatment_context_id
</td>
<td style="text-align:left;">
A unique integer identifier for a treatment, where a treatment is any experimental manipulation to an organism’s growing/living conditions. The identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
collection_date
</td>
<td style="text-align:left;">
Date sample was taken, in the format <code>yyyy-mm-dd</code>, <code>yyyy-mm</code> or <code>yyyy</code>, depending on the resoluton specified. Alternatively an overall range for the study can be indicating, with the starting and ending sample date sepatated by a <code>/</code>, as in 2010-10/2011-03.
</td>
</tr>
<tr>
<td style="text-align:left;">
measurement_remarks
</td>
<td style="text-align:left;">
Brief comments or notes accompanying the trait measurement.
</td>
</tr>
<tr>
<td style="text-align:left;">
method_id
</td>
<td style="text-align:left;">
A unique integer identifier to distinguish between multiple sets of methods used to measure a single trait within the same dataset. The identifier links to specific information in the methods table.
</td>
</tr>
<tr>
<td style="text-align:left;">
method_context_id
</td>
<td style="text-align:left;">
A unique integer identifier indicating a trait is measured multiple times on the same entity, with different methods used for each entry. This field is only used if a single trait is measured using multiple methods within the same dataset. The identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
original_name
</td>
<td style="text-align:left;">
Name given to taxon in the original data supplied by the authors.
</td>
</tr>
</tbody>

</table>
<section id="entity-type" class="level3">
<h3 class="anchored" data-anchor-id="entity-type">Entity type</h3>
<p>An entity is the <code>feature of interest</code>, indicating what a trait value applies to. While an entity can be just a component of an organism, within the scope of AusTraits, an <code>individual</code> is the finest scale entity that can be documented. The same study might measure some traits at a population-level (<code>entity = population</code>) and others at an individual-level (<code>entity = individual</code>).</p>
<p>In detail:</p>
<ul>
<li><code>entity_type</code> is a categorical variable specifying the entity corresponding to the trait values recorded. Possible values are:</li>
</ul>

<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead>
<tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
individual
</td>
<td style="text-align:left;">
Value comes from a single individual.
</td>
</tr>
<tr>
<td style="text-align:left;">
population
</td>
<td style="text-align:left;">
Value represents a summary statistic from multiple individuals at a single location.
</td>
</tr>
<tr>
<td style="text-align:left;">
metapopulation
</td>
<td style="text-align:left;">
Value represents a summary statistic from individuals of the taxon across multiple locations.
</td>
</tr>
<tr>
<td style="text-align:left;">
species
</td>
<td style="text-align:left;">
Value represents a summary statistic for a species or infraspecific taxon across its range or as estimated by an expert based on their knowledge of the taxon. Data fitting this category include estimates from reference books that represent a taxon’s entire range and values for categorical variables obtained from a reference book or identified by an expert.
</td>
</tr>
<tr>
<td style="text-align:left;">
genus
</td>
<td style="text-align:left;">
Value represents a summary statistic or expert score for an entire genus.
</td>
</tr>
<tr>
<td style="text-align:left;">
family
</td>
<td style="text-align:left;">
Value represents a summary statistic or expert score for an entire family.
</td>
</tr>
<tr>
<td style="text-align:left;">
order
</td>
<td style="text-align:left;">
Value represents a summary statistic or expert score for an entire order.
</td>
</tr>
</tbody>

</table>
</section>
<section id="identifiers" class="level3">
<h3 class="anchored" data-anchor-id="identifiers">Identifiers</h3>
<p>The traits table includes 12 identifiers, <code>dataset_id</code>, <code>observation_id</code>, <code>taxon_name</code>, <code>population_id</code>, <code>individual_id</code>, <code>temporal_context_id</code>, <code>source_id</code>, <code>location_id</code>, <code>entity_context_id</code>, <code>plot_context_id</code>, <code>treatment_context_id</code>, and <code>method_context_id</code>.</p>
<p><code>dataset_id</code>, <code>source_id</code> and <code>taxon_name</code> have easy-to-interpret values. The others are simply integral identifiers that link groups of measurements and are automatically generated through the AusTraits workflow (<code>individual_id</code> can be assigned in the metadata file or automatically generated.)</p>
<p>To expand on the definitions provided above,</p>
<ul>
<li><p><code>observation_id</code> links measurements made on the same entity (individual, population, or species) at a single point in time.</p></li>
<li><p><code>population_id</code> indicates entites that share a common <code>location_id</code>, <code>plot_context_id</code>, and <code>treatment_context_id</code>. It is used to align measurements and observation_id’s for <code>individuals</code> versus <code>populations</code> (i.e.&nbsp;distinct <code>entity_types</code>) that share a common <code>population_id</code>. It is numbered sequentially within a dataset.</p></li>
<li><p><code>individual_id</code> indicates a unique organisms. It is numbered sequentially within a dataset by population. Multiple observations on the same organism across time (with distinct <code>observation_id</code> values), share a common <code>individual_id</code>.</p></li>
<li><p><code>temporal_context_id</code> indicates a distinct point in time and is used only if there are repeat measurements on a population or individual across time. The identifier links to context properties (&amp; their associated information) in the <code>contexts</code> table for context properties of type <code>temporal</code>.</p></li>
<li><p><code>source_id</code> is applied if not all data within a single dataset (<code>dataset_id</code>) is from the same source, such as when a dataset represents a compilation for a meta-analysis.</p></li>
<li><p><code>location_id</code> links to a distinct <code>location_name</code> and associated <code>location_properties</code> in the <code>location</code> table.</p></li>
<li><p><code>entity_context_id</code> links to information in the <code>contexts</code> table for context properties (&amp; associated values/descriptions) with category <code>entity_context</code>. <code>Entity_contexts</code> include organism sex, organism caste and any other features of an entity that needs to be documented.</p></li>
<li><p><code>plot_context_id</code> links to information in the <code>contexts</code> table for context properties (&amp; associated values/descriptions) with category <code>plot</code>. <code>Plot contexts</code> include both blocks/plots within an experimental design as well as any stratified variation within a location that needs to be documented (e.g.&nbsp;slope position).</p></li>
<li><p><code>treatment_context_id</code>links to information in the <code>contexts</code> table for context properties (&amp; associated values/descriptions) with category <code>treatment</code>. <code>Treatment contexts</code> are experimental manipulations applied to groups of individuals.</p></li>
<li><p><code>method_context_id</code>links to information in the <code>contexts</code> table for context properties (&amp; associated values/descriptions) with category <code>method</code>. A <code>method context</code> indicates that the same trait was measured on or across individuals using different methods.</p></li>
</ul>
<p>As well, <code>measurement_remarks</code> is used to document brief comments or notes accompanying the trait measurement.</p>
</section>
<section id="life-stage-basis-of-record" class="level3">
<h3 class="anchored" data-anchor-id="life-stage-basis-of-record">Life stage, basis of record</h3>
<ul>
<li><p><code>life_stage</code> is a field to indicate the life stage or age class of the entity measured. standard values are <code>adult</code>, <code>sapling</code>, <code>seedling</code> and <code>juvenile</code>..</p></li>
<li><p><code>basis_of_record</code> is a categorical variable specifying from which kind of specimen traits were recorded. Possible values are:</p></li>
</ul>

<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead>
<tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
field
</td>
<td style="text-align:left;">
Traits were recorded on entities living naturally in the field.
</td>
</tr>
<tr>
<td style="text-align:left;">
field_experiment
</td>
<td style="text-align:left;">
Traits were recorded on entities living under experimentally manipulated conditions in the field.
</td>
</tr>
<tr>
<td style="text-align:left;">
captive_cultivated
</td>
<td style="text-align:left;">
Traits were recorded on entities living in a common garden, arboretum, or botanical or zoological garden.
</td>
</tr>
<tr>
<td style="text-align:left;">
lab
</td>
<td style="text-align:left;">
Traits were recorded on entities growing in a lab, glasshouse or growth chamber.
</td>
</tr>
<tr>
<td style="text-align:left;">
preserved_specimen
</td>
<td style="text-align:left;">
Traits were recorded from specimens preserved in a collection, eg. herbarium or museum.
</td>
</tr>
<tr>
<td style="text-align:left;">
literature
</td>
<td style="text-align:left;">
Traits were sourced from values reported in the literature, and where the basis of record is not otherwise known.
</td>
</tr>
</tbody>

</table>
</section>
<section id="value_types" class="level3">
<h3 class="anchored" data-anchor-id="value_types">Values, Value types, basis of value</h3>
<p>Each record in the table of trait data has an associated <code>value</code>, <code>value_type</code>, and <code>basis_of_value</code>.</p>
<p><code>Value type:</code> A trait’s <code>value_type</code> is either <code>numeric</code> or <code>categorical</code>. - For traits with numerical values, the recorded value has been converted into standardised units and the AusTraits workflow has confirmed the value can be converted into a number and lies within the allowable range. - For categorical variables, records have been aligned through substitutions to values listed as allowable values (terms) in a trait’s definition. * we use <code>_</code> for multi-word terms, e.g.&nbsp;<code>semi_deciduous</code><br>
* we use a space for situations where two values co-occur for the same entity. For instance, a flora might indicate that a plant species can be either annual or biennial, in which case the trait is scored as <code>annual biennial</code>.</p>
<p>Each trait measurement has an associated <code>value_type</code>, which is a categorical variable describing the statistical nature of the trait value recorded. Possible values are:</p>

<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead>
<tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
raw
</td>
<td style="text-align:left;">
Value recorded for an entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
minimum
</td>
<td style="text-align:left;">
Value is the minimum of values recorded for an entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
mean
</td>
<td style="text-align:left;">
Value is the mean of values recorded for an entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
median
</td>
<td style="text-align:left;">
Value is the median of values recorded for an entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
maximum
</td>
<td style="text-align:left;">
Value is the maximum of values recorded for an entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
mode
</td>
<td style="text-align:left;">
Value is the mode of values recorded for an entity. This is the appropriate value type for a categorical trait value.
</td>
</tr>
<tr>
<td style="text-align:left;">
range
</td>
<td style="text-align:left;">
Value is a range of values recorded for an entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
bin
</td>
<td style="text-align:left;">
Value for an entity falls within specified limits.
</td>
</tr>
<tr>
<td style="text-align:left;">
unknown
</td>
<td style="text-align:left;">
Not currently known.
</td>
</tr>
</tbody>

</table>
<p>Each trait measurement also has an associated <code>basis_of_value</code>, which is a categorical variable describing how the trait value was obtained. Possible values are:</p>

<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead>
<tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
measurement
</td>
<td style="text-align:left;">
Value is the result of a measurement(s) made on a specimen(s).
</td>
</tr>
<tr>
<td style="text-align:left;">
expert_score
</td>
<td style="text-align:left;">
Value has been estimated by an expert based on their knowledge of the entity.
</td>
</tr>
<tr>
<td style="text-align:left;">
model_derived
</td>
<td style="text-align:left;">
Value is derived from a statistical model, for example via gap-filling.
</td>
</tr>
<tr>
<td style="text-align:left;">
unknown
</td>
<td style="text-align:left;">
Not currently known.
</td>
</tr>
</tbody>

</table>
</section>
</section>
<section id="locations" class="level2" data-number="8.2">
<h2 data-number="8.2" class="anchored" data-anchor-id="locations"><span class="header-section-number">8.2</span> locations</h2>
<p><strong>Description:</strong> A table containing observations of location/site characteristics associated with information in <code>traits</code>. Cross referencing between the two dataframes is possible using combinations of the variables <code>dataset_id</code>, <code>location_name</code>.</p>
<p><strong>Content:</strong></p>

<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead>
<tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often these are scientific papers, books, or online resources. By default this should be the name of the first author and year of publication, e.g.&nbsp;<code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
location_id
</td>
<td style="text-align:left;">
A unique integer identifier for a location, with locations numbered sequentially within a dataset. The identifier links to specific information in the location table.
</td>
</tr>
<tr>
<td style="text-align:left;">
location_name
</td>
<td style="text-align:left;">
The location name.
</td>
</tr>
<tr>
<td style="text-align:left;">
location_property
</td>
<td style="text-align:left;">
The location characteristic being recorded. The name should include units of measurement, e.g.&nbsp;<code>MAT (C)</code>. Ideally we have at least the following variables for each location, <code>longitude (deg)</code>, <code>latitude (deg)</code>, <code>description</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
value
</td>
<td style="text-align:left;">
The measured value of a location property.
</td>
</tr>
</tbody>

</table>
</section>
<section id="contexts" class="level2" data-number="8.3">
<h2 data-number="8.3" class="anchored" data-anchor-id="contexts"><span class="header-section-number">8.3</span> contexts</h2>
<p><strong>Description:</strong> A table containing observations of contextual characteristics associated with information in <code>traits</code>. Cross referencing between the two dataframes is possible using combinations of the variables <code>dataset_id</code>, <code>link_id</code>, and <code>link_vals</code>.</p>
<p><strong>Content:</strong></p>

<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead>
<tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often these are scientific papers, books, or online resources. By default this should be the name of the first author and year of publication, e.g.&nbsp;<code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
context_property
</td>
<td style="text-align:left;">
The contextual characteristic being recorded. If applicable, name should include units of measurement, e.g.&nbsp;<code>CO2 concentration (ppm)</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
category
</td>
<td style="text-align:left;">
The category of context property, with options being <code>plot</code>, <code>treatment</code>, <code>individual_context</code>, <code>temporal</code> and <code>method</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
value
</td>
<td style="text-align:left;">
The measured value of a context property.
</td>
</tr>
<tr>
<td style="text-align:left;">
description
</td>
<td style="text-align:left;">
Description of a specific context property value.
</td>
</tr>
<tr>
<td style="text-align:left;">
link_id
</td>
<td style="text-align:left;">
Variable indicating which identifier column in the traits table contains the specified <code>link_vals</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
link_vals
</td>
<td style="text-align:left;">
Unique integer identifiers that link between identifier columns in the <code>traits</code> table and the contextual properties/values in the <code>contexts</code> table.
</td>
</tr>
</tbody>

</table>
</section>
<section id="methods" class="level2" data-number="8.4">
<h2 data-number="8.4" class="anchored" data-anchor-id="methods"><span class="header-section-number">8.4</span> methods</h2>
<p><strong>Description:</strong> A table containing details on methods with which data were collected, including time frame and source. Cross referencing with the <code>traits</code> table is possible using combinations of the variables <code>dataset_id</code>, <code>trait_name</code>.</p>
<p><strong>Content:</strong></p>

<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead>
<tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often these are scientific papers, books, or online resources. By default this should be the name of the first author and year of publication, e.g.&nbsp;<code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
trait_name
</td>
<td style="text-align:left;">
Name of the trait sampled. Allowable values specified in the table <code>definitions</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
methods
</td>
<td style="text-align:left;">
A textual description of the methods used to collect the trait data. Whenever available, methods are taken near-verbatim from the referenced source. Methods can include descriptions such as ‘measured on botanical collections’, ‘data from the literature’, or a detailed description of the field or lab methods used to collect the data.
</td>
</tr>
<tr>
<td style="text-align:left;">
method_id
</td>
<td style="text-align:left;">
A unique integer identifier to distinguish between multiple sets of methods used to measure a single trait within the same dataset. The identifier links to specific information in the methods table.
</td>
</tr>
<tr>
<td style="text-align:left;">
description
</td>
<td style="text-align:left;">
A 1-2 sentence description of the purpose of the study.
</td>
</tr>
<tr>
<td style="text-align:left;">
sampling_strategy
</td>
<td style="text-align:left;">
A written description of how study locations were selected and how study individuals were selected. When available, this information is lifted verbatim from a published manuscript. For preserved specimens, this field ideally indicates which records were ‘sampled’ to measure a specific trait.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_primary_key
</td>
<td style="text-align:left;">
Citation key for the primary source in <code>sources</code>. The key is typically formatted as <code>Surname_year</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_primary_citation
</td>
<td style="text-align:left;">
Citation for the primary source. This detail is generated from the primary source in the metadata.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_secondary_key
</td>
<td style="text-align:left;">
Citation key for the secondary source in <code>sources</code>. The key is typically formatted as <code>Surname_year</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_secondary_citation
</td>
<td style="text-align:left;">
Citations for the secondary source. This detail is generated from the secondary source in the metadata.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_original_dataset_key
</td>
<td style="text-align:left;">
Citation key for the original dataset_id in sources; for compilations. The key is typically formatted as <code>Surname_year</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_original_dataset_citation
</td>
<td style="text-align:left;">
Citations for the original dataset_id in sources; for compilationse. This detail is generated from the original source in the metadata.
</td>
</tr>
<tr>
<td style="text-align:left;">
data_collectors
</td>
<td style="text-align:left;">
The person (people) leading data collection for this study.
</td>
</tr>
<tr>
<td style="text-align:left;">
assistants
</td>
<td style="text-align:left;">
Names of additional people who played a more minor role in data collection for the study.
</td>
</tr>
<tr>
<td style="text-align:left;">
dataset_curators
</td>
<td style="text-align:left;">
Names of AusTraits team member(s) who contacted the data collectors and added the study to the AusTraits repository.
</td>
</tr>
</tbody>

</table>
</section>
<section id="excluded_data" class="level2" data-number="8.5">
<h2 data-number="8.5" class="anchored" data-anchor-id="excluded_data"><span class="header-section-number">8.5</span> Excluded_data</h2>
<p><strong>Description:</strong> A table of data that did not pass quality tests and so were excluded from the master dataset. The structure is identical to that presented in the <code>traits</code> table, only with an extra column called <code>error</code> indicating why the record was excluded. Common reasons are missing_unit_conversions, missing_value, and unsupported_trait_value.</p>
<p><strong>Content:</strong></p>

<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead>
<tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
error
</td>
<td style="text-align:left;">
Indicating why the record was excluded. Common reasons are missing_unit_conversions, missing_value, and unsupported_trait_value.
</td>
</tr>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often these are scientific papers, books, or online resources. By default this should be the name of the first author and year of publication, e.g.&nbsp;<code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_name
</td>
<td style="text-align:left;">
Scientific name of the taxon on which traits were sampled, without authorship. When possible, this is the currently accepted (botanical) or valid (zoological) scientific name, but might also be a higher taxonomic level.
</td>
</tr>
<tr>
<td style="text-align:left;">
observation_id
</td>
<td style="text-align:left;">
A unique integral identifier for the observation, where an observation is all measurements made on an individual at a single point in time. It is important for joining traits coming from the same <code>observation_id</code>. Within each dataset, observation_id’s are unique combinations of <code>taxon_name</code>, <code>population_id</code>, <code>individual_id</code>, and <code>temporal_context_id</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
trait_name
</td>
<td style="text-align:left;">
Name of the trait sampled. Allowable values specified in the table <code>definitions</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
value
</td>
<td style="text-align:left;">
The measured value of a trait.
</td>
</tr>
<tr>
<td style="text-align:left;">
unit
</td>
<td style="text-align:left;">
Units of the sampled trait value after aligning with AusTraits standards.
</td>
</tr>
<tr>
<td style="text-align:left;">
entity_type
</td>
<td style="text-align:left;">
A categorical variable specifying the entity corresponding to the trait values recorded.
</td>
</tr>
<tr>
<td style="text-align:left;">
value_type
</td>
<td style="text-align:left;">
A categorical variable describing the statistical nature of the trait value recorded.
</td>
</tr>
<tr>
<td style="text-align:left;">
basis_of_value
</td>
<td style="text-align:left;">
A categorical variable describing how the trait value was obtained.
</td>
</tr>
<tr>
<td style="text-align:left;">
replicates
</td>
<td style="text-align:left;">
Number of replicate measurements that comprise a recorded trait measurement. A numeric value (or range) is ideal and appropriate if the value type is a <code>mean</code>, <code>median</code>, <code>min</code> or <code>max</code>. For these value types, if replication is unknown the entry should be <code>unknown</code>. If the value type is <code>raw_value</code> the replicate value should be 1. If the trait is categorical or the value indicates a measurement for an entire species (or other taxon) replicate value should be <code>.na</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
basis_of_record
</td>
<td style="text-align:left;">
A categorical variable specifying from which kind of specimen traits were recorded.
</td>
</tr>
<tr>
<td style="text-align:left;">
life_stage
</td>
<td style="text-align:left;">
A field to indicate the life stage or age class of the entity measured. Standard values are <code>adult</code>, <code>sapling</code>, <code>seedling</code> and <code>juvenile</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
population_id
</td>
<td style="text-align:left;">
A unique integer identifier for a population, where a population is defined as individuals growing in the same location (location_id /location_name) and plot (plot_context_id, a context category) and being subjected to the same treatment (treatment_context_id, a context category).
</td>
</tr>
<tr>
<td style="text-align:left;">
individual_id
</td>
<td style="text-align:left;">
A unique integer identifier for an individual, with individuals numbered sequentially within each dataset by taxon by population grouping. Most often each row of data represents an individual, but in some datasets trait data collected on a single individual is presented across multiple rows of data, such as if the same trait is measured using different methods or the same individual is measured repeatedly across time.
</td>
</tr>
<tr>
<td style="text-align:left;">
repeat_measurements_id
</td>
<td style="text-align:left;">
A unique integer identifier for repeat measurements of a trait that comprise a single observation, such as a response curve.
</td>
</tr>
<tr>
<td style="text-align:left;">
temporal_context_id
</td>
<td style="text-align:left;">
A unique integer identifier assigned where repeat observations are made on the same individual (or population, or taxon) across time. The identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
source_id
</td>
<td style="text-align:left;">
For datasets that are compilations, an identifier for the original data source.
</td>
</tr>
<tr>
<td style="text-align:left;">
location_id
</td>
<td style="text-align:left;">
A unique integer identifier for a location, with locations numbered sequentially within a dataset. The identifier links to specific information in the location table.
</td>
</tr>
<tr>
<td style="text-align:left;">
entity_context_id
</td>
<td style="text-align:left;">
A unique integer identifier indicating specific contextual properties of an individual, possibly including the individual’s sex or caste (for social insects).
</td>
</tr>
<tr>
<td style="text-align:left;">
plot_context_id
</td>
<td style="text-align:left;">
A unique integer identifier for a plot, where a plot is a distinct collection of organisms within a single geographic location, such as plants growing on different aspects or blocks in an experiment. The identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
treatment_context_id
</td>
<td style="text-align:left;">
A unique integer identifier for a treatment, where a treatment is any experimental manipulation to an organism’s growing/living conditions. The identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
collection_date
</td>
<td style="text-align:left;">
Date sample was taken, in the format <code>yyyy-mm-dd</code>, <code>yyyy-mm</code> or <code>yyyy</code>, depending on the resoluton specified. Alternatively an overall range for the study can be indicating, with the starting and ending sample date sepatated by a <code>/</code>, as in 2010-10/2011-03.
</td>
</tr>
<tr>
<td style="text-align:left;">
measurement_remarks
</td>
<td style="text-align:left;">
Brief comments or notes accompanying the trait measurement.
</td>
</tr>
<tr>
<td style="text-align:left;">
method_id
</td>
<td style="text-align:left;">
A unique integer identifier to distinguish between multiple sets of methods used to measure a single trait within the same dataset. The identifier links to specific information in the methods table.
</td>
</tr>
<tr>
<td style="text-align:left;">
method_context_id
</td>
<td style="text-align:left;">
A unique integer identifier indicating a trait is measured multiple times on the same entity, with different methods used for each entry. This field is only used if a single trait is measured using multiple methods within the same dataset. The identifier links to specific information in the context table.
</td>
</tr>
<tr>
<td style="text-align:left;">
original_name
</td>
<td style="text-align:left;">
Name given to taxon in the original data supplied by the authors.
</td>
</tr>
</tbody>

</table>
</section>
<section id="taxa" class="level2" data-number="8.6">
<h2 data-number="8.6" class="anchored" data-anchor-id="taxa"><span class="header-section-number">8.6</span> taxa</h2>
<p><strong>Description:</strong> A table containing details on taxa that are included in the table <a href="#traits"><code>traits</code></a>. We have attempted to align species names with known taxonomic units in the <a href="https://biodiversity.org.au/nsl/services/apc"><code>Australian Plant Census</code> (APC)</a> and/or the <a href="https://biodiversity.org.au/nsl/services/APNI"><code>Australian Plant Names Index</code> (APNI)</a>; the sourced information is released under a CC-BY3 license.</p>
<p>Version 0.1.0 of AusTraits contains records for 1245 different taxa.</p>
<p><strong>Content:</strong></p>

<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead>
<tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
taxon_name
</td>
<td style="text-align:left;">
Scientific name of the taxon on which traits were sampled, without authorship. When possible, this is the currently accepted (botanical) or valid (zoological) scientific name, but might also be a higher taxonomic level.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxonomic_dataset
</td>
<td style="text-align:left;">
Name of the taxonomy (tree) that contains this concept. ie. APC, AusMoss etc.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_rank
</td>
<td style="text-align:left;">
The taxonomic rank of the most specific name in the scientific name.
</td>
</tr>
<tr>
<td style="text-align:left;">
trinomial
</td>
<td style="text-align:left;">
The infraspecific taxon name match for an original name. This column is assigned <code>na</code> for taxon name that are at a broader taxonomic_resolution.
</td>
</tr>
<tr>
<td style="text-align:left;">
binomial
</td>
<td style="text-align:left;">
The species-level taxon name match for an original name. This column is assigned <code>na</code> for taxon name that are at a broader taxonomic_resolution.
</td>
</tr>
<tr>
<td style="text-align:left;">
genus
</td>
<td style="text-align:left;">
Genus of the taxon without authorship.
</td>
</tr>
<tr>
<td style="text-align:left;">
family
</td>
<td style="text-align:left;">
Family of the taxon.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_distribution
</td>
<td style="text-align:left;">
Known distribution of the taxon, by Australian state.
</td>
</tr>
<tr>
<td style="text-align:left;">
establishment_means
</td>
<td style="text-align:left;">
Statement about whether an organism or organisms have been introduced to a given place and time through the direct or indirect activity of modern humans.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxonomic_status
</td>
<td style="text-align:left;">
The status of the use of the scientificName as a label for the taxon in regard to the ‘accepted (or valid) taxonomy’. The assigned taxonomic status must be linked to a specific taxonomic reference that defines the concept.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_id
</td>
<td style="text-align:left;">
An identifier for the set of taxon information (data associated with the taxon class). May be a global unique identifier or an identifier specific to the data set. Must be resolvable within this dataset.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_id_genus
</td>
<td style="text-align:left;">
An identifier for the set of taxon information (data associated with the taxon class) for the genus associated with a taxon name. May be a global unique identifier or an identifier specific to the data set. Must be resolvable within this dataset.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_id_family
</td>
<td style="text-align:left;">
An identifier for the set of taxon information (data associated with the taxon class) for the family associated with a taxon name. May be a global unique identifier or an identifier specific to the data set. Must be resolvable within this dataset.
</td>
</tr>
<tr>
<td style="text-align:left;">
scientific_name
</td>
<td style="text-align:left;">
The full scientific name, with authorship and date information if known.
</td>
</tr>
<tr>
<td style="text-align:left;">
scientific_name_id
</td>
<td style="text-align:left;">
An identifier for the set of taxon information (data associated with the taxon class). May be a global unique identifier or an identifier specific to the data set. Must be resolvable within this dataset.
</td>
</tr>
</tbody>

</table>
</section>
<section id="taxonomic_updates" class="level2" data-number="8.7">
<h2 data-number="8.7" class="anchored" data-anchor-id="taxonomic_updates"><span class="header-section-number">8.7</span> taxonomic_updates</h2>
<p><strong>Description:</strong> A table of all taxonomic changes implemented in the construction of AusTraits. Changes are determined by comparing the originally submitted taxon name against the taxonomic names listed in the taxonomic reference files, best placed in a subfolder in the <code>config</code> folder . Cross referencing with the <code>traits</code> table is possible using combinations of the variables <code>dataset_id</code> and <code>taxon_name</code>.</p>
<p><strong>Content:</strong></p>

<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead>
<tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often these are scientific papers, books, or online resources. By default this should be the name of the first author and year of publication, e.g.&nbsp;<code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
original_name
</td>
<td style="text-align:left;">
Name given to taxon in the original data supplied by the authors.
</td>
</tr>
<tr>
<td style="text-align:left;">
aligned_name
</td>
<td style="text-align:left;">
The taxon name without authorship after implementing automated syntax standardisation and spelling changes as well as manually encoded syntax alignments for this taxon in the metadata file for the corresponding <code>dataset_id</code>. This name has not yet been matched to the currently accepted (botanical) or valid (zoological) taxon name in cases where there are taxonomic synonyms, isonyms, orthographic variants, etc.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxonomic_resolution
</td>
<td style="text-align:left;">
The rank of the most specific taxon name (or scientific name) to which a submitted orignal name resolves.
</td>
</tr>
<tr>
<td style="text-align:left;">
taxon_name
</td>
<td style="text-align:left;">
Scientific name of the taxon on which traits were sampled, without authorship. When possible, this is the currently accepted (botanical) or valid (zoological) scientific name, but might also be a higher taxonomic level.
</td>
</tr>
<tr>
<td style="text-align:left;">
aligned_name_taxon_id
</td>
<td style="text-align:left;">
An identifier for the aligned name before it is updated to the currently accepted name usage. This may be a global unique identifier or an identifier specific to the data set. Must be resolvable within this dataset.
</td>
</tr>
<tr>
<td style="text-align:left;">
aligned_name_taxonomic_status
</td>
<td style="text-align:left;">
The status of the use of the <code>aligned_name</code> as a label for a taxon. Requires taxonomic opinion to define the scope of a taxon. Rules of priority then are used to define the taxonomic status of the nomenclature contained in that scope, combined with the experts opinion. It must be linked to a specific taxonomic reference that defines the concept.
</td>
</tr>
</tbody>

</table>
<p>Both the original and the updated taxon names are included in the <a href="#traits"><code>traits</code></a> table.</p>
</section>
<section id="definitions" class="level2" data-number="8.8">
<h2 data-number="8.8" class="anchored" data-anchor-id="definitions"><span class="header-section-number">8.8</span> definitions</h2>
<p><strong>Description:</strong> A copy of the definitions for all tables and terms. Information included here was used to process data and generate any documentation for the study.</p>
<p><strong>Details on trait definitions:</strong> The allowable trait names and trait values are defined in the definitions file. Each trait is labelled as either <code>numeric</code> or <code>categorical</code>. An example of each type is as follows. For an example, see the the <a href="http://traitecoevo.github.io/austraits.build/articles/trait_definitions.html">Trait definitions for AusTraits</a>.</p>
<p><strong>leaf_mass_per_area</strong></p>
<ul>
<li>number of records: 0</li>
<li>number of studies: 0</li>
</ul>
<p><strong>woodiness</strong></p>
<ul>
<li>number of records: 0</li>
<li>number of studies: 0</li>
</ul>
</section>
<section id="contributors" class="level2" data-number="8.9">
<h2 data-number="8.9" class="anchored" data-anchor-id="contributors"><span class="header-section-number">8.9</span> contributors</h2>
<p><strong>Description:</strong> A table of people contributing to each study.</p>
<p><strong>Content:</strong></p>

<table class="table table-striped table-hover table-condensed table-responsive" style="margin-left:30px; width: auto !important; ">
<thead>
<tr>
<th style="text-align:left;">
key
</th>
<th style="text-align:left;">
value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
dataset_id
</td>
<td style="text-align:left;">
Primary identifier for each study contributed to AusTraits; most often these are scientific papers, books, or online resources. By default this should be the name of the first author and year of publication, e.g.&nbsp;<code>Falster_2005</code>.
</td>
</tr>
<tr>
<td style="text-align:left;">
last_name
</td>
<td style="text-align:left;">
Last name of the data collector.
</td>
</tr>
<tr>
<td style="text-align:left;">
given_name
</td>
<td style="text-align:left;">
Given names of the data collector.
</td>
</tr>
<tr>
<td style="text-align:left;">
ORCID
</td>
<td style="text-align:left;">
ORCID of the data collector.
</td>
</tr>
<tr>
<td style="text-align:left;">
affiliation
</td>
<td style="text-align:left;">
Last known institution or affiliation.
</td>
</tr>
<tr>
<td style="text-align:left;">
additional_role
</td>
<td style="text-align:left;">
Additional roles of data collector, mostly contact person.
</td>
</tr>
</tbody>

</table>
</section>
<section id="sources" class="level2" data-number="8.10">
<h2 data-number="8.10" class="anchored" data-anchor-id="sources"><span class="header-section-number">8.10</span> sources</h2>
<p>For each dataset in the compilation there is the option to list primary and secondary citations. The primary citation is defined as, <code>The secondary citation is defined as,</code></p>
<p>The element <code>sources</code> includes bibtex versions of all sources which can be imported into your reference library:</p>
<pre><code>RefManageR::WriteBib(austraits$sources, "refs.bib") # write all sources to file
RefManageR::WriteBib(austraits$sources["Falster_2005_1"], "refs.bib") # write a single reference to a file</code></pre>
<p>Or individually viewed:</p>
<pre><code>austraits$sources["Falster_2005_1"]</code></pre>
<p>A formatted version of the sources also exists within the table <a href="#methods">methods</a>.</p>
</section>
<section id="metadata" class="level2" data-number="8.11">
<h2 data-number="8.11" class="anchored" data-anchor-id="metadata"><span class="header-section-number">8.11</span> metadata</h2>
<p><strong>Description:</strong> Metadata associated with the dataset, including title, creators, license, subject, funding sources.</p>
</section>
<section id="build_info" class="level2" data-number="8.12">
<h2 data-number="8.12" class="anchored" data-anchor-id="build_info"><span class="header-section-number">8.12</span> build_info</h2>
<p><strong>Description:</strong> A description of the computing environment used to create this version of the dataset, including version number, git commit and R session_info.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./database_standard.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data standard</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./database_metadata.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Overview of traits.build ontology</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">How to get help: <a href="https://traitecoevo.github.io/traits.build-book/help.html" class="uri">https://traitecoevo.github.io/traits.build-book/help.html</a><br>Copyright 2023, Daniel Falster and Elizabeth Wenk</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>